// @cemex/cmx-table-v7/esm2015/lib/components/cmx-datepicker-filter/cmx-datepicker-filter.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Host } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { Filter } from '../../classes/filter.class';
import { WatcherService } from '../../services/watcher.service';
import { CmxTableComponent } from '../cmx-table/cmx-table.component';
export class CmxDatepickerFilterComponent {
    /**
     * @param {?} watcher
     * @param {?} parent
     */
    constructor(watcher, parent) {
        this.watcher = watcher;
        this.parent = parent;
        this.isRtl = false;
        this.activeFilter = new BehaviorSubject(false);
        this.isRtl = this.parent.rtl;
    }
    /**
     * @return {?}
     */
    listenToActiveFilter() {
        return this.activeFilter.asObservable();
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    activateFilter($event) {
        this.filter.value = $event.value;
        this.filter.active = true;
        /** @type {?} */
        const $aux = [this.filter];
        this.activeFilter.next(true);
        this.watcher.emitActiveFilters(this.parent.assignedId, $aux);
    }
    /**
     * @return {?}
     */
    removeFilter() {
        this.filter.active = false;
        /** @type {?} */
        const $aux = [this.filter];
        this.activeFilter.next(false);
        this.watcher.emitActiveFilters(this.parent.assignedId, $aux);
    }
}
CmxDatepickerFilterComponent.decorators = [
    { type: Component, args: [{
                selector: 'cmx-datepicker-filter',
                template: "<cmx-datepicker (select)=\"activateFilter( $event )\" (cancel)=\"removeFilter()\" [rtl]=\"isRtl\"></cmx-datepicker>",
                styles: [""]
            }] }
];
/** @nocollapse */
CmxDatepickerFilterComponent.ctorParameters = () => [
    { type: WatcherService },
    { type: CmxTableComponent, decorators: [{ type: Host }] }
];
CmxDatepickerFilterComponent.propDecorators = {
    filter: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    CmxDatepickerFilterComponent.prototype.filter;
    /** @type {?} */
    CmxDatepickerFilterComponent.prototype.isRtl;
    /**
     * @type {?}
     * @private
     */
    CmxDatepickerFilterComponent.prototype.activeFilter;
    /**
     * @type {?}
     * @private
     */
    CmxDatepickerFilterComponent.prototype.watcher;
    /**
     * @type {?}
     * @private
     */
    CmxDatepickerFilterComponent.prototype.parent;
}
