// @cemex/cmx-table-v7/esm5/lib/components/cmx-datepicker-filter/cmx-datepicker-filter.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Host } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { Filter } from '../../classes/filter.class';
import { WatcherService } from '../../services/watcher.service';
import { CmxTableComponent } from '../cmx-table/cmx-table.component';
var CmxDatepickerFilterComponent = /** @class */ (function () {
    function CmxDatepickerFilterComponent(watcher, parent) {
        this.watcher = watcher;
        this.parent = parent;
        this.isRtl = false;
        this.activeFilter = new BehaviorSubject(false);
        this.isRtl = this.parent.rtl;
    }
    /**
     * @return {?}
     */
    CmxDatepickerFilterComponent.prototype.listenToActiveFilter = /**
     * @return {?}
     */
    function () {
        return this.activeFilter.asObservable();
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    CmxDatepickerFilterComponent.prototype.activateFilter = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        this.filter.value = $event.value;
        this.filter.active = true;
        /** @type {?} */
        var $aux = [this.filter];
        this.activeFilter.next(true);
        this.watcher.emitActiveFilters(this.parent.assignedId, $aux);
    };
    /**
     * @return {?}
     */
    CmxDatepickerFilterComponent.prototype.removeFilter = /**
     * @return {?}
     */
    function () {
        this.filter.active = false;
        /** @type {?} */
        var $aux = [this.filter];
        this.activeFilter.next(false);
        this.watcher.emitActiveFilters(this.parent.assignedId, $aux);
    };
    CmxDatepickerFilterComponent.decorators = [
        { type: Component, args: [{
                    selector: 'cmx-datepicker-filter',
                    template: "<cmx-datepicker (select)=\"activateFilter( $event )\" (cancel)=\"removeFilter()\" [rtl]=\"isRtl\"></cmx-datepicker>",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    CmxDatepickerFilterComponent.ctorParameters = function () { return [
        { type: WatcherService },
        { type: CmxTableComponent, decorators: [{ type: Host }] }
    ]; };
    CmxDatepickerFilterComponent.propDecorators = {
        filter: [{ type: Input }]
    };
    return CmxDatepickerFilterComponent;
}());
export { CmxDatepickerFilterComponent };
if (false) {
    /** @type {?} */
    CmxDatepickerFilterComponent.prototype.filter;
    /** @type {?} */
    CmxDatepickerFilterComponent.prototype.isRtl;
    /**
     * @type {?}
     * @private
     */
    CmxDatepickerFilterComponent.prototype.activeFilter;
    /**
     * @type {?}
     * @private
     */
    CmxDatepickerFilterComponent.prototype.watcher;
    /**
     * @type {?}
     * @private
     */
    CmxDatepickerFilterComponent.prototype.parent;
}
