// @cemex-core/types-v1/documentFile.dto.ts
export interface Reference{
    cemexReference: string;
    documentId: number;
    checked?: boolean;
}


export interface SearchReferenceDTO{
    references: Array<Reference>;
    //
    documentTypeCode?: string;
    customerBlobReqId?: number;
}

export interface FileReference{
    uploadTrigger?: boolean;
    uploadDuration?: number;
    //
    blobDocumentID?: string;
    cemexReference?: string;
    cemexReferenceType?: string;
    customerBlobReqId?: number;
    customerReference?: string;
    documentTypeCode?: string;
    referenceDate?: string;
    referenceDateObj?: Date;
    referenceStatus?: boolean;
    fileName?: string;
    fileContent?: string;
    customerCode: string;
    countryCode: string;
}

export interface DocumentFile{
    uploadDate: string;
    uploadDateObj?: Date;
    documentTypeCode: string;
    documentTypeDesc: string;
    customerDocsBlobId: number;
    cemexReference: string;
    blobDocumentID: string;
    customerReference: string;
    customerDocsStatus: boolean;
    documentSearchCode: string;
    documentSearchDesc: string;
}

export interface DocumentFileDTO{
    documents: Array<DocumentFile>;
}
