// @cemex-core/types-v1/document.dto.ts
export class Document{
    // folioDocument: number;
    amount: number;
    applyDiscount: boolean;
    balance: number;
    checked?: boolean;
    companyCode: string;
    countryCode: string;
    contract: string;
    createDate: string;
    createDateObj?: Date;
    createDateTimestamp?: number;
    currencyCode: string;
    customerCode: string;
    customerJobsiteNumber: string;
    daysOverDue: number;
    daysToDue: number;
    debitFlag: boolean;
    discount: number;
    dispute: boolean;
    disputeStatusCode: string;
    disputeStatusDesc: string;
    disputeCompromiseDate: string;
    disputeCompromiseDateObj?: Date;
    disputeCode: number;
    divisionCode: string;
    docStatusCode: string;
    docStatusDesc: string;
    documentCode: string;
    documentId: number;
    documentItem: string;
    documentTypeCode: string;
    documentTypeDesc: string;
    dueDate: string;
    dueDateObj?: Date;
    dueDateTimestamp?: number;
    index?: number;
    invoiceReviewed: boolean;
    jobNumber: string; 
    jobSiteCode: string;
    jobSiteDesc: string;
    paymentFlag: boolean;
    paymentProcessFlag: boolean;
    purchaseOrder: string;
    referenceNumber: string;
    relatedDocument: string;
    salesOrgCode: string;
    totalContracts: number;
    totalPO: number;
    payment?: number;
    expectedPayment?: number;
    // isCheked: Boolean;
    // InvoiceDisputesStatus: string;
}
