// @cemex-core/types-v1/drivers.order.dto.ts
import { DriverPlantDto } from './drivers.plant.dto';

export const STATUS_ALL: string = "All";
export const STATUS_TICKETED: string = "Ticketed";
export const STATUS_ON_SITE: string = "On_Site";
export const STATUS_TO_JOB: string = "To_Job";
export const STATUS_TO_PLANT: string = "To_Plant";

export interface IOrderOptions {
    orderId: string;
    orderCode: string;
    ticketId: string;
    ticketCode: string;
    totalQuantity: number;
    um: string;
    volume: number;
    podName: string;
    clientName: string;
}

export interface IOrderTruck {
    trucks: any[];
}

export interface DictStatusTruck {
    [status: string]: any[];
}

export class OrderStatusTruck implements IOrderTruck {
    statusCode: string;
    trucks: any[];

    constructor(status?: string) {
        this.statusCode = status;
        this.trucks = [];
    }

    get statusDesc(): string {
        let desc: string = this.statusCode.split("_").join(" ");
        return `${desc} (${this.trucks.length})`;
    }

}

export class DriverOrder {
    orderId: string;
    orderCode: string;
    ticketId: string;
    ticketCode: string;
    totalQuantity: number;
    um: string;
    volume: number;
    podName: string;
    clientName: string;

    latitude: number = null;
    longitude: number = null;
    jobsiteId: string = null;
    jobsiteName: string = null;

    plants: DriverPlantDto[];
    statusTrucks: OrderStatusTruck[];
    dictStatusTruck: DictStatusTruck = {};

    constructor(options?: IOrderOptions) {
        this.orderId = options.orderId;
        this.orderCode = options.orderCode;
        this.ticketId = options.ticketId;
        this.ticketCode = options.ticketCode;
        this.totalQuantity = options.totalQuantity;
        this.um = options.um;
        this.volume = options.volume;
        this.podName = options.podName;
        this.clientName = options.clientName;

        this.plants = [];
        this.statusTrucks = [];

        this.dictStatusTruck[STATUS_TICKETED] = [];
        this.dictStatusTruck[STATUS_ON_SITE] = [];
        this.dictStatusTruck[STATUS_TO_JOB] = [];
        this.dictStatusTruck[STATUS_TO_PLANT] = [];

        this.statusTrucks[0] = new OrderStatusTruck(STATUS_ALL);
        this.statusTrucks[1] = new OrderStatusTruck(STATUS_TICKETED);
        this.statusTrucks[2] = new OrderStatusTruck(STATUS_TO_JOB);
        this.statusTrucks[3] = new OrderStatusTruck(STATUS_ON_SITE);
        this.statusTrucks[4] = new OrderStatusTruck(STATUS_TO_PLANT);
    }

}
