// @cemex/cmx-alert-v7/esm5/lib/services/alert.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { AlertType } from '../models/alert';
import { AlertRef } from './alert-ref';
import { BehaviorSubject } from 'rxjs';
var AlertService = /** @class */ (function () {
    function AlertService() {
        this.alerts$ = new BehaviorSubject([]);
        this.alerts = [];
    }
    /**
     * @param {?} alertMessage
     * @param {?} actionMessage
     * @param {?=} duration
     * @param {?=} alertType
     * @return {?}
     */
    AlertService.prototype.openSuccess = /**
     * @param {?} alertMessage
     * @param {?} actionMessage
     * @param {?=} duration
     * @param {?=} alertType
     * @return {?}
     */
    function (alertMessage, actionMessage, duration, alertType) {
        if (duration === void 0) { duration = 5000; }
        if (alertType === void 0) { alertType = AlertType.Success; }
        return this.open(alertMessage, actionMessage, duration, alertType);
    };
    /**
     * @param {?} alertMessage
     * @param {?} actionMessage
     * @param {?=} duration
     * @param {?=} alertType
     * @return {?}
     */
    AlertService.prototype.openError = /**
     * @param {?} alertMessage
     * @param {?} actionMessage
     * @param {?=} duration
     * @param {?=} alertType
     * @return {?}
     */
    function (alertMessage, actionMessage, duration, alertType) {
        if (duration === void 0) { duration = 5000; }
        if (alertType === void 0) { alertType = AlertType.Error; }
        return this.open(alertMessage, actionMessage, duration, alertType);
    };
    /**
     * @param {?} alertMessage
     * @param {?} actionMessage
     * @param {?=} duration
     * @param {?=} alertType
     * @return {?}
     */
    AlertService.prototype.openWarning = /**
     * @param {?} alertMessage
     * @param {?} actionMessage
     * @param {?=} duration
     * @param {?=} alertType
     * @return {?}
     */
    function (alertMessage, actionMessage, duration, alertType) {
        if (duration === void 0) { duration = 5000; }
        if (alertType === void 0) { alertType = AlertType.Warning; }
        return this.open(alertMessage, actionMessage, duration, alertType);
    };
    /**
     * @param {?} alertMessage
     * @param {?} actionMessage
     * @param {?=} duration
     * @param {?=} alertType
     * @return {?}
     */
    AlertService.prototype.openInfo = /**
     * @param {?} alertMessage
     * @param {?} actionMessage
     * @param {?=} duration
     * @param {?=} alertType
     * @return {?}
     */
    function (alertMessage, actionMessage, duration, alertType) {
        if (duration === void 0) { duration = 5000; }
        if (alertType === void 0) { alertType = AlertType.Info; }
        return this.open(alertMessage, actionMessage, duration, alertType);
    };
    /**
     * @private
     * @param {?} alertMessage
     * @param {?} actionMessage
     * @param {?} duration
     * @param {?} alertType
     * @return {?}
     */
    AlertService.prototype.open = /**
     * @private
     * @param {?} alertMessage
     * @param {?} actionMessage
     * @param {?} duration
     * @param {?} alertType
     * @return {?}
     */
    function (alertMessage, actionMessage, duration, alertType) {
        /** @type {?} */
        var alertRef = new AlertRef(alertMessage, actionMessage, duration, alertType);
        this.alerts.push(alertRef);
        this.alerts$.next(this.alerts);
        if (alertRef.duration > 0) {
            this._setTimeout(alertRef);
        }
        return alertRef;
    };
    /**
     * @private
     * @param {?} alert
     * @return {?}
     */
    AlertService.prototype._setTimeout = /**
     * @private
     * @param {?} alert
     * @return {?}
     */
    function (alert) {
        var _this = this;
        window.setTimeout((/**
         * @return {?}
         */
        function () {
            _this.removeAlert(alert);
        }), alert.duration);
    };
    /**
     * @param {?} alert
     * @return {?}
     */
    AlertService.prototype.removeAlert = /**
     * @param {?} alert
     * @return {?}
     */
    function (alert) {
        this.alerts = this.alerts.filter((/**
         * @param {?} x
         * @return {?}
         */
        function (x) { return x !== alert; }));
        this.alerts$.next(this.alerts);
    };
    AlertService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    AlertService.ctorParameters = function () { return []; };
    return AlertService;
}());
export { AlertService };
if (false) {
    /** @type {?} */
    AlertService.prototype.alerts$;
    /** @type {?} */
    AlertService.prototype.alerts;
}
