// @cemex/cmx-alert-v7/esm5/lib/services/alert-ref.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Subject } from 'rxjs';
var AlertRef = /** @class */ (function () {
    function AlertRef(alertMessage, actionMessage, duration, alertType) {
        this.alertMessage = alertMessage;
        this.actionMessage = actionMessage;
        this.duration = duration;
        this.alertType = alertType;
        this._onAction = new Subject();
        this._onClose = new Subject();
    }
    /**
     * @return {?}
     */
    AlertRef.prototype.onAction = /**
     * @return {?}
     */
    function () {
        this._onAction.next();
    };
    /**
     * @return {?}
     */
    AlertRef.prototype.onActionClicked = /**
     * @return {?}
     */
    function () {
        return this._onAction.asObservable();
    };
    /**
     * @return {?}
     */
    AlertRef.prototype.onClose = /**
     * @return {?}
     */
    function () {
        this._onClose.next();
    };
    /**
     * @return {?}
     */
    AlertRef.prototype.onCloseClicked = /**
     * @return {?}
     */
    function () {
        return this._onClose.asObservable();
    };
    return AlertRef;
}());
export { AlertRef };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AlertRef.prototype._onAction;
    /**
     * @type {?}
     * @private
     */
    AlertRef.prototype._onClose;
    /** @type {?} */
    AlertRef.prototype.alertMessage;
    /** @type {?} */
    AlertRef.prototype.actionMessage;
    /** @type {?} */
    AlertRef.prototype.duration;
    /** @type {?} */
    AlertRef.prototype.alertType;
}
