// @cemex/cmx-table-v7/esm2015/lib/components/cxm-filter-sorter/cmx-filter-sorter.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
export class CmxFilterSorterComponent {
    constructor() {
        this._searchValue = '';
        this._isAscending = false;
        this._isDescending = false;
        this.specificProp = null;
        this.enableSearch = true;
        this.onSelectedFilterItem = new EventEmitter();
        this.isfiltered = new EventEmitter();
        this.inputSelected = new EventEmitter();
        this.checkboxSelected = new EventEmitter();
        this._filteredCollection = [];
        this._compareCustomFn = null;
        this._searchFilterFn = null;
        this._rtl = false;
    }
    /**
     * @return {?}
     */
    get collection() {
        return this._filteredCollection;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set collection(value) {
        this._filteredCollection = value;
        if (value && value !== undefined) {
            this._originalCollection = value.slice(0, value.length);
        }
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set searchValue(value) {
        this._searchValue = value;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set sorterCustomFn(value) {
        this._compareCustomFn = value;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set searchFilterFn(value) {
        this._searchFilterFn = value;
    }
    /**
     * @return {?}
     */
    get rtl() {
        return this._rtl;
    }
    /**
     * @param {?} $value
     * @return {?}
     */
    set rtl($value) {
        if ($value !== undefined && $value !== undefined) {
            this._rtl = $value;
        }
    }
    /**
     * @return {?}
     */
    get dirRtl() {
        return (this._rtl) ? 'rtl' : 'ltr';
    }
    /**
     * @param {?} event
     * @return {?}
     */
    clickInside(event) {
        // event.stopPropagation();
    }
    /**
     * @return {?}
     */
    reset() {
        this._searchValue = '';
        this._isAscending = false;
        this._isDescending = false;
        this._filteredCollection = this._originalCollection.slice(0, this._originalCollection.length);
        this.onSelectedFilterItem.emit(this._filteredCollection);
        this.isfiltered.emit(false);
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    ascendingItem($event) {
        this.checkboxSelected.emit({ order: 'asc', value: $event });
        this._isAscending = $event;
        this._isDescending = false;
        this.orderAZ();
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    descendingItem($event) {
        this.checkboxSelected.emit({ order: 'des', value: $event });
        this._isAscending = false;
        this._isDescending = $event;
        this.orderAZ();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onSearch(event) {
        /** @type {?} */
        const isAZ = this._isAscending || this._isDescending;
        this._filteredCollection = this._searchFilterFn(this._originalCollection, event.currentTarget.value, this.specificProp);
        if (isAZ) {
            this.orderAZ();
        }
        else {
            this.onSelectedFilterItem.emit(this._filteredCollection);
        }
        this.isfiltered.emit(this._searchValue !== '');
    }
    /**
     * @return {?}
     */
    resetInput() {
        /** @type {?} */
        const validate = this._isAscending || this._isDescending;
        if ((this._searchValue === '' || !this._searchValue) && !validate) {
            this.inputSelected.emit(this.id);
        }
    }
    // tslint:disable-next-line:typedef
    /**
     * @private
     * @return {?}
     */
    orderAZ() {
        if (!this._compareCustomFn) {
            return;
        }
        /** @type {?} */
        const isAZ = this._isAscending || this._isDescending;
        if (isAZ) {
            this._filteredCollection = this.orderAscDesc();
        }
        else {
            this._filteredCollection = this._searchValue !== '' ? this._filteredCollection
                .slice(0, this._originalCollection.length) :
                this._originalCollection.slice(0, this._originalCollection.length);
        }
        this.onSelectedFilterItem.emit(this._filteredCollection);
        this.isfiltered.emit(true);
    }
    /**
     * @private
     * @return {?}
     */
    orderAscDesc() {
        /** @type {?} */
        let sortCollection;
        if (this.specificProp) {
            sortCollection = this._filteredCollection.sort(this._compareCustomFn(this.specificProp));
            return this._isAscending ? sortCollection : sortCollection.reverse();
        }
        else {
            sortCollection = this._filteredCollection.sort(this._compareCustomFn);
            return this._isAscending ? sortCollection : sortCollection.reverse();
        }
    }
}
CmxFilterSorterComponent.decorators = [
    { type: Component, args: [{
                selector: 'cmx-filter-sorter',
                template: "<!-- this component must be inside a div with dropdown bootstrao class \r\n    this component must be after a button to fire bootstrap dropdown\r\n    this component must have dropdown-menu bootstrap class\r\n-->\r\n\r\n<div \r\n    class=\"dropdown-menu-container\" \r\n    aria-labelledby=\"\" \r\n    (click)=\"clickInside($event)\" id=\"filtro\">\r\n    <div id=\"cmx-filter-sorter-search-input\" *ngIf=\"enableSearch\" dir=\"{{dirRtl}}\" class=\"search-input\">\r\n        <span class=\"cmx-icon-magnifier-glass icon\" aria-hidden=\"true\"></span>\r\n        <input (keyup.enter)=\"onSearch($event)\" [(ngModel)]=\"_searchValue\" type=\"text\" placeholder=\"{{ placeholder }}\" (focus)=\"resetInput()\">\r\n        <span class=\"cmx-icon-close icon-clear\" (click)=\"reset()\" aria-hidden=\"true\"></span>\r\n    </div>\r\n    <div class=\"cmx-checkbox__container\">\r\n        <cmx-checkbox [rtl]=\"rtl\" class=\"sorted__box\" [ngModel]=\"_isAscending\" (ngModelChange)=\"ascendingItem($event)\">{{keyOrderAsc}}</cmx-checkbox>\r\n        <br/>\r\n        <cmx-checkbox [rtl]=\"rtl\" class=\"sorted__box\" [ngModel]=\"_isDescending\" (ngModelChange)=\"descendingItem($event)\">{{keyOrderDesc}}</cmx-checkbox>\r\n    </div>\r\n</div>",
                styles: [".roboto-bold{font-weight:700}.roboto-medium{font-weight:500}.roboto-regular{font-weight:400}.roboto-light{font-weight:300}.roboto-thin{font-weight:100}.h1-role,h1{font-family:Roboto,sans-serif;font-size:3rem;line-height:3rem;font-weight:700;color:#001b3a}.h2-role,.h3-role,h2,h3{font-family:Roboto,sans-serif;font-size:1.625rem;line-height:2rem;font-weight:400;color:#001b3a}.h4-role,h4{font-family:Roboto,sans-serif;font-size:2rem;line-height:2rem;font-weight:500;color:#001b3a}.h5-role,h5{font-family:Roboto,sans-serif;font-size:1.25rem;line-height:1.5rem;font-weight:400;color:#001b3a}.h6-role,h6{font-family:Roboto,sans-serif;font-size:.875rem;line-height:1rem;font-weight:700;color:#001b3a;text-transform:uppercase}.table-entry{font-family:Roboto,sans-serif;font-size:.875rem;line-height:1rem}.table-entry-strong{font-family:Roboto,sans-serif;font-size:.875rem;line-height:1rem;color:#333;font-weight:700}.link,.table-entry,a{text-decoration:none;color:#3fa9f5}a:hover{text-decoration:underline}.input-placeholder{font-family:Roboto,sans-serif;font-size:.875rem;line-height:1rem;color:#666}.input-text{font-family:Roboto,sans-serif;font-size:.875rem;line-height:1rem;color:#333}.size-15{font-size:.9375rem}.size-24{font-size:1.5rem}.text-left{text-align:left!important}.text-right{text-align:right!important}.text-center{text-align:center!important}.flex-first{order:-1}.flex-last{order:1}.flex-unordered{order:0}.flex-row{flex-direction:row!important}.flex-column{flex-direction:column!important}.flex-row-reverse{flex-direction:row-reverse!important}.flex-column-reverse{flex-direction:column-reverse!important}.flex-wrap{flex-wrap:wrap!important}.flex-nowrap{flex-wrap:nowrap!important}.flex-wrap-reverse{flex-wrap:wrap-reverse!important}.justify-content-start{justify-content:flex-start!important}.justify-content-end{justify-content:flex-end!important}.justify-content-center{justify-content:center!important}.justify-content-between{justify-content:space-between!important}.justify-content-around{justify-content:space-around!important}.align-items-start{align-items:flex-start!important}.align-items-end{align-items:flex-end!important}.align-items-center{align-items:center!important}.align-items-baseline{align-items:baseline!important}.align-items-stretch{align-items:stretch!important}.align-content-start{align-content:flex-start!important}.align-content-end{align-content:flex-end!important}.align-content-center{align-content:center!important}.align-content-between{align-content:space-between!important}.align-content-around{align-content:space-around!important}.align-content-stretch{align-content:stretch!important}.align-self-auto{-ms-grid-row-align:auto!important;align-self:auto!important}.align-self-start{align-self:flex-start!important}.align-self-end{align-self:flex-end!important}.align-self-center{-ms-grid-row-align:center!important;align-self:center!important}.align-self-baseline{align-self:baseline!important}.align-self-stretch{-ms-grid-row-align:stretch!important;align-self:stretch!important}@media (min-width:576px){.text-xs-left{text-align:left!important}.text-xs-right{text-align:right!important}.text-xs-center{text-align:center!important}.flex-xs-first{order:-1}.flex-xs-last{order:1}.flex-xs-unordered{order:0}.flex-xs-row{flex-direction:row!important}.flex-xs-column{flex-direction:column!important}.flex-xs-row-reverse{flex-direction:row-reverse!important}.flex-xs-column-reverse{flex-direction:column-reverse!important}.flex-xs-wrap{flex-wrap:wrap!important}.flex-xs-nowrap{flex-wrap:nowrap!important}.flex-xs-wrap-reverse{flex-wrap:wrap-reverse!important}.justify-content-xs-start{justify-content:flex-start!important}.justify-content-xs-end{justify-content:flex-end!important}.justify-content-xs-center{justify-content:center!important}.justify-content-xs-between{justify-content:space-between!important}.justify-content-xs-around{justify-content:space-around!important}.align-items-xs-start{align-items:flex-start!important}.align-items-xs-end{align-items:flex-end!important}.align-items-xs-center{align-items:center!important}.align-items-xs-baseline{align-items:baseline!important}.align-items-xs-stretch{align-items:stretch!important}.align-content-xs-start{align-content:flex-start!important}.align-content-xs-end{align-content:flex-end!important}.align-content-xs-center{align-content:center!important}.align-content-xs-between{align-content:space-between!important}.align-content-xs-around{align-content:space-around!important}.align-content-xs-stretch{align-content:stretch!important}.align-self-xs-auto{-ms-grid-row-align:auto!important;align-self:auto!important}.align-self-xs-start{align-self:flex-start!important}.align-self-xs-end{align-self:flex-end!important}.align-self-xs-center{-ms-grid-row-align:center!important;align-self:center!important}.align-self-xs-baseline{align-self:baseline!important}.align-self-xs-stretch{-ms-grid-row-align:stretch!important;align-self:stretch!important}}@media (min-width:768px){.text-sm-left{text-align:left!important}.text-sm-right{text-align:right!important}.text-sm-center{text-align:center!important}.flex-sm-first{order:-1}.flex-sm-last{order:1}.flex-sm-unordered{order:0}.flex-sm-row{flex-direction:row!important}.flex-sm-column{flex-direction:column!important}.flex-sm-row-reverse{flex-direction:row-reverse!important}.flex-sm-column-reverse{flex-direction:column-reverse!important}.flex-sm-wrap{flex-wrap:wrap!important}.flex-sm-nowrap{flex-wrap:nowrap!important}.flex-sm-wrap-reverse{flex-wrap:wrap-reverse!important}.justify-content-sm-start{justify-content:flex-start!important}.justify-content-sm-end{justify-content:flex-end!important}.justify-content-sm-center{justify-content:center!important}.justify-content-sm-between{justify-content:space-between!important}.justify-content-sm-around{justify-content:space-around!important}.align-items-sm-start{align-items:flex-start!important}.align-items-sm-end{align-items:flex-end!important}.align-items-sm-center{align-items:center!important}.align-items-sm-baseline{align-items:baseline!important}.align-items-sm-stretch{align-items:stretch!important}.align-content-sm-start{align-content:flex-start!important}.align-content-sm-end{align-content:flex-end!important}.align-content-sm-center{align-content:center!important}.align-content-sm-between{align-content:space-between!important}.align-content-sm-around{align-content:space-around!important}.align-content-sm-stretch{align-content:stretch!important}.align-self-sm-auto{-ms-grid-row-align:auto!important;align-self:auto!important}.align-self-sm-start{align-self:flex-start!important}.align-self-sm-end{align-self:flex-end!important}.align-self-sm-center{-ms-grid-row-align:center!important;align-self:center!important}.align-self-sm-baseline{align-self:baseline!important}.align-self-sm-stretch{-ms-grid-row-align:stretch!important;align-self:stretch!important}}@media (min-width:1080px){.text-md-left{text-align:left!important}.text-md-right{text-align:right!important}.text-md-center{text-align:center!important}.flex-md-first{order:-1}.flex-md-last{order:1}.flex-md-unordered{order:0}.flex-md-row{flex-direction:row!important}.flex-md-column{flex-direction:column!important}.flex-md-row-reverse{flex-direction:row-reverse!important}.flex-md-column-reverse{flex-direction:column-reverse!important}.flex-md-wrap{flex-wrap:wrap!important}.flex-md-nowrap{flex-wrap:nowrap!important}.flex-md-wrap-reverse{flex-wrap:wrap-reverse!important}.justify-content-md-start{justify-content:flex-start!important}.justify-content-md-end{justify-content:flex-end!important}.justify-content-md-center{justify-content:center!important}.justify-content-md-between{justify-content:space-between!important}.justify-content-md-around{justify-content:space-around!important}.align-items-md-start{align-items:flex-start!important}.align-items-md-end{align-items:flex-end!important}.align-items-md-center{align-items:center!important}.align-items-md-baseline{align-items:baseline!important}.align-items-md-stretch{align-items:stretch!important}.align-content-md-start{align-content:flex-start!important}.align-content-md-end{align-content:flex-end!important}.align-content-md-center{align-content:center!important}.align-content-md-between{align-content:space-between!important}.align-content-md-around{align-content:space-around!important}.align-content-md-stretch{align-content:stretch!important}.align-self-md-auto{-ms-grid-row-align:auto!important;align-self:auto!important}.align-self-md-start{align-self:flex-start!important}.align-self-md-end{align-self:flex-end!important}.align-self-md-center{-ms-grid-row-align:center!important;align-self:center!important}.align-self-md-baseline{align-self:baseline!important}.align-self-md-stretch{-ms-grid-row-align:stretch!important;align-self:stretch!important}}@media (min-width:1320px){.text-lg-left{text-align:left!important}.text-lg-right{text-align:right!important}.text-lg-center{text-align:center!important}.flex-lg-first{order:-1}.flex-lg-last{order:1}.flex-lg-unordered{order:0}.flex-lg-row{flex-direction:row!important}.flex-lg-column{flex-direction:column!important}.flex-lg-row-reverse{flex-direction:row-reverse!important}.flex-lg-column-reverse{flex-direction:column-reverse!important}.flex-lg-wrap{flex-wrap:wrap!important}.flex-lg-nowrap{flex-wrap:nowrap!important}.flex-lg-wrap-reverse{flex-wrap:wrap-reverse!important}.justify-content-lg-start{justify-content:flex-start!important}.justify-content-lg-end{justify-content:flex-end!important}.justify-content-lg-center{justify-content:center!important}.justify-content-lg-between{justify-content:space-between!important}.justify-content-lg-around{justify-content:space-around!important}.align-items-lg-start{align-items:flex-start!important}.align-items-lg-end{align-items:flex-end!important}.align-items-lg-center{align-items:center!important}.align-items-lg-baseline{align-items:baseline!important}.align-items-lg-stretch{align-items:stretch!important}.align-content-lg-start{align-content:flex-start!important}.align-content-lg-end{align-content:flex-end!important}.align-content-lg-center{align-content:center!important}.align-content-lg-between{align-content:space-between!important}.align-content-lg-around{align-content:space-around!important}.align-content-lg-stretch{align-content:stretch!important}.align-self-lg-auto{-ms-grid-row-align:auto!important;align-self:auto!important}.align-self-lg-start{align-self:flex-start!important}.align-self-lg-end{align-self:flex-end!important}.align-self-lg-center{-ms-grid-row-align:center!important;align-self:center!important}.align-self-lg-baseline{align-self:baseline!important}.align-self-lg-stretch{-ms-grid-row-align:stretch!important;align-self:stretch!important}}@media (min-width:1560px){.text-xlg-left{text-align:left!important}.text-xlg-right{text-align:right!important}.text-xlg-center{text-align:center!important}.flex-xlg-first{order:-1}.flex-xlg-last{order:1}.flex-xlg-unordered{order:0}.flex-xlg-row{flex-direction:row!important}.flex-xlg-column{flex-direction:column!important}.flex-xlg-row-reverse{flex-direction:row-reverse!important}.flex-xlg-column-reverse{flex-direction:column-reverse!important}.flex-xlg-wrap{flex-wrap:wrap!important}.flex-xlg-nowrap{flex-wrap:nowrap!important}.flex-xlg-wrap-reverse{flex-wrap:wrap-reverse!important}.justify-content-xlg-start{justify-content:flex-start!important}.justify-content-xlg-end{justify-content:flex-end!important}.justify-content-xlg-center{justify-content:center!important}.justify-content-xlg-between{justify-content:space-between!important}.justify-content-xlg-around{justify-content:space-around!important}.align-items-xlg-start{align-items:flex-start!important}.align-items-xlg-end{align-items:flex-end!important}.align-items-xlg-center{align-items:center!important}.align-items-xlg-baseline{align-items:baseline!important}.align-items-xlg-stretch{align-items:stretch!important}.align-content-xlg-start{align-content:flex-start!important}.align-content-xlg-end{align-content:flex-end!important}.align-content-xlg-center{align-content:center!important}.align-content-xlg-between{align-content:space-between!important}.align-content-xlg-around{align-content:space-around!important}.align-content-xlg-stretch{align-content:stretch!important}.align-self-xlg-auto{-ms-grid-row-align:auto!important;align-self:auto!important}.align-self-xlg-start{align-self:flex-start!important}.align-self-xlg-end{align-self:flex-end!important}.align-self-xlg-center{-ms-grid-row-align:center!important;align-self:center!important}.align-self-xlg-baseline{align-self:baseline!important}.align-self-xlg-stretch{-ms-grid-row-align:stretch!important;align-self:stretch!important}}:host{display:block;max-height:196px;border-top:3px solid #3fa9f5;cursor:default}.dropdown-menu-container{width:325px;max-height:204px}.dropdown-menu-container .search-input{padding:0 10px}.dropdown-menu-container .search-input .icon{top:8%;position:absolute}.dropdown-menu-container .search-input .icon-magnifier-glass{width:15px;height:15px;position:absolute}.dropdown-menu-container .search-input input{width:100%;border:0;border-bottom:1px solid #d6d6d6;background-color:transparent;text-indent:20px;padding-bottom:5px}.dropdown-menu-container .search-input input:focus{outline:0}.dropdown-menu-container .cmx-checkbox__container{margin-top:10px}.search-input[dir=rtl] .icon-clear{right:90%}.sorted__box{font-size:14px;padding:10px 9px}.icon-clear{position:absolute;top:8%;right:5%;width:4.3%}.icon-clear .icon-close-black{width:12px}.content-buttons-action{flex-direction:row-reverse;display:flex}.content-buttons-action a{padding:5px}"]
            }] }
];
/** @nocollapse */
CmxFilterSorterComponent.ctorParameters = () => [];
CmxFilterSorterComponent.propDecorators = {
    collection: [{ type: Input }],
    sorterCustomFn: [{ type: Input }],
    searchFilterFn: [{ type: Input }],
    rtl: [{ type: Input }],
    specificProp: [{ type: Input }],
    id: [{ type: Input }],
    placeholder: [{ type: Input }],
    keyTitle: [{ type: Input }],
    keySubtitle: [{ type: Input }],
    keyOrderAsc: [{ type: Input }],
    keyOrderDesc: [{ type: Input }],
    enableSearch: [{ type: Input }],
    onSelectedFilterItem: [{ type: Output }],
    isfiltered: [{ type: Output }],
    inputSelected: [{ type: Output }],
    checkboxSelected: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    CmxFilterSorterComponent.prototype._searchValue;
    /** @type {?} */
    CmxFilterSorterComponent.prototype._isAscending;
    /** @type {?} */
    CmxFilterSorterComponent.prototype._isDescending;
    /** @type {?} */
    CmxFilterSorterComponent.prototype.specificProp;
    /** @type {?} */
    CmxFilterSorterComponent.prototype.id;
    /** @type {?} */
    CmxFilterSorterComponent.prototype.placeholder;
    /** @type {?} */
    CmxFilterSorterComponent.prototype.keyTitle;
    /** @type {?} */
    CmxFilterSorterComponent.prototype.keySubtitle;
    /** @type {?} */
    CmxFilterSorterComponent.prototype.keyOrderAsc;
    /** @type {?} */
    CmxFilterSorterComponent.prototype.keyOrderDesc;
    /** @type {?} */
    CmxFilterSorterComponent.prototype.enableSearch;
    /** @type {?} */
    CmxFilterSorterComponent.prototype.onSelectedFilterItem;
    /** @type {?} */
    CmxFilterSorterComponent.prototype.isfiltered;
    /** @type {?} */
    CmxFilterSorterComponent.prototype.inputSelected;
    /** @type {?} */
    CmxFilterSorterComponent.prototype.checkboxSelected;
    /**
     * @type {?}
     * @private
     */
    CmxFilterSorterComponent.prototype._filteredCollection;
    /**
     * @type {?}
     * @private
     */
    CmxFilterSorterComponent.prototype._originalCollection;
    /**
     * @type {?}
     * @private
     */
    CmxFilterSorterComponent.prototype._compareCustomFn;
    /**
     * @type {?}
     * @private
     */
    CmxFilterSorterComponent.prototype._searchFilterFn;
    /**
     * @type {?}
     * @private
     */
    CmxFilterSorterComponent.prototype._rtl;
}
