// @cemex-core/angular-localization-v7/esm2015/lib/pipes/cmx-number.pipe.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Pipe } from '@angular/core';
import { FormatterService } from './../services/formatter.service';
import { TranslationService } from './../services/translation.service';
export class CmxNumberPipe {
    /**
     * @param {?} translationService
     * @param {?} formatterService
     */
    constructor(translationService, formatterService) {
        this.translationService = translationService;
        this.formatterService = formatterService;
    }
    /**
     * @param {?} numberEntry
     * @return {?}
     */
    transform(numberEntry) {
        return this.formatterService.formatNumberTS(numberEntry, this.translationService.selectedLanguage.languageISO.replace('_', '-'), this.translationService.selectedLanguage.decimalNumbers);
    }
}
CmxNumberPipe.decorators = [
    { type: Pipe, args: [{
                name: 'cmxNumber',
                pure: false
            },] }
];
/** @nocollapse */
CmxNumberPipe.ctorParameters = () => [
    { type: TranslationService },
    { type: FormatterService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    CmxNumberPipe.prototype.translationService;
    /**
     * @type {?}
     * @private
     */
    CmxNumberPipe.prototype.formatterService;
}
