// @cemex-core/angular-localization-v7/esm2015/lib/pipes/cmx-date-time-moment.pipe.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Pipe } from '@angular/core';
import { DateTimeActions, FormatterService } from './../services/formatter.service';
import { TranslationService } from './../services/translation.service';
export class CmxDateTimeMomentPipe {
    /**
     * @param {?} translationService
     * @param {?} formatterService
     */
    constructor(translationService, formatterService) {
        this.translationService = translationService;
        this.formatterService = formatterService;
    }
    /**
     * @param {?} dateTime
     * @param {?=} action
     * @return {?}
     */
    transform(dateTime, action = DateTimeActions.ConvertToLocal) {
        return this.formatterService.formatDateTimeMoment(dateTime, this.translationService.selectedLanguage.languageISO.replace('_', '-'), this.translationService.selectedLanguage.formatDate +
            ' ' +
            this.translationService.selectedLanguage.formatTime, action);
    }
}
CmxDateTimeMomentPipe.decorators = [
    { type: Pipe, args: [{
                name: 'cmxDateTimeMoment',
                pure: false
            },] }
];
/** @nocollapse */
CmxDateTimeMomentPipe.ctorParameters = () => [
    { type: TranslationService },
    { type: FormatterService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    CmxDateTimeMomentPipe.prototype.translationService;
    /**
     * @type {?}
     * @private
     */
    CmxDateTimeMomentPipe.prototype.formatterService;
}
