// @cemex-core/angular-localization-v7/esm2015/lib/pipes/cmx-currency.pipe.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { isPlatformBrowser } from '@angular/common';
import { Inject, Pipe, PLATFORM_ID } from '@angular/core';
import { FormatterService } from './../services/formatter.service';
import { TranslationService } from './../services/translation.service';
export class CmxCurrencyPipe {
    /**
     * @param {?} formatter
     * @param {?} translation
     * @param {?} platformId
     */
    constructor(formatter, translation, platformId) {
        this.formatter = formatter;
        this.translation = translation;
        this.platformId = platformId;
        if (isPlatformBrowser(platformId)) {
            this.window = window;
        }
    }
    /**
     * @param {?} value
     * @param {?=} countryCode
     * @return {?}
     */
    transform(value, countryCode) {
        if (value === '' || value == null || isNaN((/** @type {?} */ (value)))) {
            console.warn('The value is not a number');
            return '';
        }
        countryCode = countryCode || this.window.sessionStorage.getItem('country');
        // tslint:disable-next-line:triple-equals
        if (countryCode == undefined) {
            console.warn(`Can't retrieve countryCode`);
            return '';
        }
        /** @type {?} */
        const lang = this.translation.getLanguageByCountryCode(countryCode);
        /** @type {?} */
        const symbolPosition = (lang.currencySymbolFloat && lang.currencySymbolFloat.toLowerCase()) ||
            'left';
        /** @type {?} */
        const currencySymbol = lang.currencySymbol || '$';
        /** @type {?} */
        const formattedNumber = this.formatter.formatNumber(value, lang && lang.decimalNumbersMoney, lang && lang.thousandSeparator, lang && lang.decimalSeparator);
        /** @type {?} */
        const numberWithCurrency = symbolPosition === 'left'
            ? `${currencySymbol}${formattedNumber}`
            : `${formattedNumber}${currencySymbol}`;
        return '&lrm;' + numberWithCurrency;
    }
}
CmxCurrencyPipe.decorators = [
    { type: Pipe, args: [{
                name: 'cmxCurrency',
                pure: true
            },] }
];
/** @nocollapse */
CmxCurrencyPipe.ctorParameters = () => [
    { type: FormatterService },
    { type: TranslationService },
    { type: undefined, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
];
if (false) {
    /** @type {?} */
    CmxCurrencyPipe.prototype.window;
    /**
     * @type {?}
     * @private
     */
    CmxCurrencyPipe.prototype.formatter;
    /**
     * @type {?}
     * @private
     */
    CmxCurrencyPipe.prototype.translation;
    /**
     * @type {?}
     * @private
     */
    CmxCurrencyPipe.prototype.platformId;
}
