// @cemex/cmx-dialog-v7/esm2015/lib/services/cmx-dialog.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class CmxDialogService {
    constructor() {
        this.cmxDialogs = new Array();
    }
    /**
     * @param {?} cmxDialog
     * @return {?}
     */
    addDialog(cmxDialog) {
        this.hideLastDialog();
        this.cmxDialogs.push(cmxDialog);
    }
    /**
     * @param {?} cmxDialog
     * @param {?} showLastDialog
     * @return {?}
     */
    removeDialog(cmxDialog, showLastDialog) {
        this.cmxDialogs.splice(this.cmxDialogs.length - 1, 1);
        if (showLastDialog) {
            this.showLastDialog();
        }
    }
    /**
     * @return {?}
     */
    closeAllDialogs() {
        for (let i = this.cmxDialogs.length - 1; i >= 0; i--) {
            this.cmxDialogs[i].close(false);
        }
    }
    /**
     * @private
     * @return {?}
     */
    showLastDialog() {
        if (this.cmxDialogs.length > 0) {
            this.cmxDialogs[this.cmxDialogs.length - 1].show();
        }
    }
    /**
     * @private
     * @return {?}
     */
    hideLastDialog() {
        if (this.cmxDialogs.length > 0) {
            this.cmxDialogs[this.cmxDialogs.length - 1].hide();
        }
    }
}
CmxDialogService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
/** @nocollapse */
CmxDialogService.ctorParameters = () => [];
/** @nocollapse */ CmxDialogService.ngInjectableDef = i0.defineInjectable({ factory: function CmxDialogService_Factory() { return new CmxDialogService(); }, token: CmxDialogService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    CmxDialogService.prototype.cmxDialogs;
}
