// @cemex/cmx-panel-card-v1/dist/cmx-panel-card.component.js
import { Component, Input } from '@angular/core';
var CmxPanelCardComponent = /** @class */ (function () {
    function CmxPanelCardComponent() {
        if (this.masterHeight === undefined) {
            this.masterHeight = '220px';
        }
        if (this.minHeight === undefined) {
            this.minHeight = '0';
        }
        this.class = 'simple-card-border';
    }
    CmxPanelCardComponent.decorators = [
        { type: Component, args: [{
                    selector: 'cmx-panel-card',
                    template: "\n      <div style=\"display: block\" class=\"{{class}}\" [style.height]=\"masterHeight\" [style.min-height]=\"minHeight\">\n          <div class=\"simple-card-title\">\n              <span class=\"simple-card-span\">{{title}}</span>\n          </div>\n          <ng-content></ng-content>\n      </div>\n    ",
                    styles: ["\n      .simple-card-border{width:100%;background-color:#fff;min-height:220px;box-shadow:0 2px 6px 3px rgba(0,0,0,0.03);color:#001B3A;border-radius:3px}.simple-card-title{background-color:#001B3A;height:32px;text-align:left;width:100%;font-weight:bold;line-height:32px}.simple-card-span{padding-left:12px;color:#fff;font-size:16px;font-family:'Roboto', sans-serif}\n    "]
                },] },
    ];
    /** @nocollapse */
    CmxPanelCardComponent.ctorParameters = function () { return []; };
    CmxPanelCardComponent.propDecorators = {
        'title': [{ type: Input },],
        'masterHeight': [{ type: Input },],
        'minHeight': [{ type: Input },],
        'class': [{ type: Input },],
    };
    return CmxPanelCardComponent;
}());
export { CmxPanelCardComponent };
