// @cemex/cmx-alert-v7/esm5/lib/alert.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AlertComponent } from './components/alert.component';
// SERVICES
import { AlertService } from './services/alert.service';
var AlertModule = /** @class */ (function () {
    function AlertModule() {
    }
    /**
     * @return {?}
     */
    AlertModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: AlertModule,
            providers: [
                AlertService
            ]
        };
    };
    AlertModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule],
                    declarations: [
                        AlertComponent,
                    ],
                    exports: [
                        AlertComponent,
                    ]
                },] }
    ];
    return AlertModule;
}());
export { AlertModule };
