// @cemex/cmx-table-v7/esm2015/lib/services/watcher.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
/** @type {?} */
let $tableId = 0;
/**
 * @record
 */
function WatchSubject() { }
if (false) {
    /** @type {?} */
    WatchSubject.prototype.id;
    /** @type {?} */
    WatchSubject.prototype.activeFilters;
    /** @type {?} */
    WatchSubject.prototype.activeFiltersSubject;
}
export class WatcherService {
    constructor() {
        this.watchSubjects = [];
    }
    /**
     * @return {?}
     */
    clearActiveFilters() {
        this.watchSubjects.forEach((/**
         * @param {?} sub
         * @return {?}
         */
        (sub) => {
            sub.activeFiltersSubject.next([]);
        }));
    }
    /**
     * @return {?}
     */
    createSubject() {
        $tableId++;
        /** @type {?} */
        const $subject = {
            id: $tableId,
            activeFilters: [],
            activeFiltersSubject: new BehaviorSubject([]),
        };
        this.watchSubjects.push($subject);
        return $tableId;
    }
    /**
     * @param {?} $id
     * @return {?}
     */
    listenToActiveFilters($id) {
        /** @type {?} */
        const $aux = this.watchSubjects.find((/**
         * @param {?} $element
         * @return {?}
         */
        ($element) => $element.id === $id));
        return $aux.activeFiltersSubject.asObservable();
    }
    /**
     * @param {?} $id
     * @param {?} $filtersValue
     * @return {?}
     */
    emitActiveFilters($id, $filtersValue) {
        /** @type {?} */
        const $aux = this.watchSubjects.find((/**
         * @param {?} $element
         * @return {?}
         */
        ($element) => $element.id === $id));
        $filtersValue.forEach((/**
         * @param {?} $filter
         * @return {?}
         */
        ($filter) => {
            /** @type {?} */
            const $foundIndex = $aux.activeFilters.findIndex((/**
             * @param {?} $element
             * @return {?}
             */
            ($element) => $element.name === $filter.name));
            if ($foundIndex === -1 && $filter.active === true) {
                // new filter to push
                $aux.activeFilters.push($filter);
            }
            else if ($foundIndex > -1 && $filter.active === false) {
                // filter was in the array but now has been deactivaded
                $aux.activeFilters.splice($foundIndex, 1);
            }
        }));
        $aux.activeFiltersSubject.next($aux.activeFilters);
    }
}
WatcherService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
WatcherService.ctorParameters = () => [];
if (false) {
    /**
     * @type {?}
     * @private
     */
    WatcherService.prototype.watchSubjects;
}
