// @cemex/cmx-table-v7/esm5/lib/services/watcher.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
/** @type {?} */
var $tableId = 0;
/**
 * @record
 */
function WatchSubject() { }
if (false) {
    /** @type {?} */
    WatchSubject.prototype.id;
    /** @type {?} */
    WatchSubject.prototype.activeFilters;
    /** @type {?} */
    WatchSubject.prototype.activeFiltersSubject;
}
var WatcherService = /** @class */ (function () {
    function WatcherService() {
        this.watchSubjects = [];
    }
    /**
     * @return {?}
     */
    WatcherService.prototype.clearActiveFilters = /**
     * @return {?}
     */
    function () {
        this.watchSubjects.forEach((/**
         * @param {?} sub
         * @return {?}
         */
        function (sub) {
            sub.activeFiltersSubject.next([]);
        }));
    };
    /**
     * @return {?}
     */
    WatcherService.prototype.createSubject = /**
     * @return {?}
     */
    function () {
        $tableId++;
        /** @type {?} */
        var $subject = {
            id: $tableId,
            activeFilters: [],
            activeFiltersSubject: new BehaviorSubject([]),
        };
        this.watchSubjects.push($subject);
        return $tableId;
    };
    /**
     * @param {?} $id
     * @return {?}
     */
    WatcherService.prototype.listenToActiveFilters = /**
     * @param {?} $id
     * @return {?}
     */
    function ($id) {
        /** @type {?} */
        var $aux = this.watchSubjects.find((/**
         * @param {?} $element
         * @return {?}
         */
        function ($element) { return $element.id === $id; }));
        return $aux.activeFiltersSubject.asObservable();
    };
    /**
     * @param {?} $id
     * @param {?} $filtersValue
     * @return {?}
     */
    WatcherService.prototype.emitActiveFilters = /**
     * @param {?} $id
     * @param {?} $filtersValue
     * @return {?}
     */
    function ($id, $filtersValue) {
        /** @type {?} */
        var $aux = this.watchSubjects.find((/**
         * @param {?} $element
         * @return {?}
         */
        function ($element) { return $element.id === $id; }));
        $filtersValue.forEach((/**
         * @param {?} $filter
         * @return {?}
         */
        function ($filter) {
            /** @type {?} */
            var $foundIndex = $aux.activeFilters.findIndex((/**
             * @param {?} $element
             * @return {?}
             */
            function ($element) { return $element.name === $filter.name; }));
            if ($foundIndex === -1 && $filter.active === true) {
                // new filter to push
                $aux.activeFilters.push($filter);
            }
            else if ($foundIndex > -1 && $filter.active === false) {
                // filter was in the array but now has been deactivaded
                $aux.activeFilters.splice($foundIndex, 1);
            }
        }));
        $aux.activeFiltersSubject.next($aux.activeFilters);
    };
    WatcherService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    WatcherService.ctorParameters = function () { return []; };
    return WatcherService;
}());
export { WatcherService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    WatcherService.prototype.watchSubjects;
}
