// @cemex-core/angular-services-v7/esm2015/lib/invoices-payments/cashSales.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { map } from 'rxjs/operators';
import { EndpointsConstants } from '@cemex-core/constants-v2/dist';
import { HttpCemex } from '../http.service';
export class CashSalesService {
    /**
     * @param {?} http
     * @param {?} endpoints
     */
    constructor(http, endpoints) {
        this.http = http;
        this.endpoints = endpoints;
        // TODO: MOVE TO PATH CONSTANTS
        this.apiCashDocuments = 'cashdocuments';
    }
    /**
     * @param {?} companyCode
     * @param {?} customerCode
     * @param {?} jobSiteCode
     * @return {?}
     */
    getCreditDocs(companyCode, customerCode, jobSiteCode) {
        /** @type {?} */
        const apiQuery = `companyCode=${companyCode}&customerCode=${customerCode}&jobSiteCode=${jobSiteCode}`;
        /** @type {?} */
        const endpoint = `${this.apiCashDocuments}?${apiQuery}`;
        return this.http
            .get(this.getUrl(endpoint))
            .pipe(map((/**
         * @param {?} response
         * @return {?}
         */
        response => {
            /** @type {?} */
            const aux = response.json();
            return this.handleCashDocumentsData(aux);
        })))
            .toPromise()
            .then((/**
         * @param {?} data
         * @return {?}
         */
        data => data))
            .catch((/**
         * @param {?} error
         * @return {?}
         */
        error => error));
    }
    /**
     * @private
     * @param {?} aux
     * @return {?}
     */
    handleCashDocumentsData(aux) {
        aux.forEach((/**
         * @param {?} document
         * @param {?} i
         * @return {?}
         */
        (document, i) => {
            document.index = i;
            document.flagInvalidAmnt = false;
            document.errMsg = '';
            document.toCommit = 0;
            // document.createDate = '2017-01-01T00:00:00';
            if (document.createDate) {
                /** @type {?} */
                const d1 = document.createDate.split('T')[0].split('-');
                /** @type {?} */
                const d2 = document.createDate.split('T')[1];
                document.createDate = d1[2] + '-' + d1[1] + '-' + d1[0] + 'T' + d2;
                // document.createDate = this.getDateObject(document.createDate).toString();
            }
            document.amount = !document.amount ? 0 : document.amount;
            document.reconciledAmount = !document.reconciledAmount
                ? 0
                : document.reconciledAmount;
            document.commitedAmount = !document.commitedAmount
                ? 0
                : document.commitedAmount;
        }));
        return aux;
    }
    /**
     * @private
     * @param {?} api
     * @return {?}
     */
    getUrl(api) {
        /** @type {?} */
        const url = this.http.generateEndpoint(this.endpoints.V1 + this.endpoints.peCapability + api);
        return url;
    }
}
CashSalesService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
CashSalesService.ctorParameters = () => [
    { type: HttpCemex },
    { type: EndpointsConstants }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    CashSalesService.prototype.apiCashDocuments;
    /**
     * @type {?}
     * @private
     */
    CashSalesService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    CashSalesService.prototype.endpoints;
}
