// @cemex-core/events-v7/esm5/lib/calendar.event.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
var CalendarEvents = /** @class */ (function () {
    function CalendarEvents() {
        this.datesSubject = new BehaviorSubject({
            start: this.moment(),
            end: this.moment()
        });
        // Observable items streams
        this.dates$ = this.datesSubject.asObservable();
    }
    /**
     * @param {?} dates
     * @return {?}
     */
    CalendarEvents.prototype.setDates = /**
     * @param {?} dates
     * @return {?}
     */
    function (dates) {
        ((/** @type {?} */ (window))).localStorage.setItem('dates', JSON.stringify(dates));
        /** @type {?} */
        var currentDates = this.getCurrentValue();
        if (dates.start.diff(currentDates.start, 'days') !== 0 || dates.end.diff(currentDates.end, 'days') !== 0) {
            this.datesSubject.next(dates);
        }
    };
    /**
     * @return {?}
     */
    CalendarEvents.prototype.getDates = /**
     * @return {?}
     */
    function () {
        return this.dates$;
    };
    /**
     * @return {?}
     */
    CalendarEvents.prototype.getCurrentValue = /**
     * @return {?}
     */
    function () {
        return this.datesSubject.getValue();
    };
    CalendarEvents.decorators = [
        { type: Injectable }
    ];
    return CalendarEvents;
}());
export { CalendarEvents };
if (false) {
    /** @type {?} */
    CalendarEvents.prototype.moment;
    /**
     * @type {?}
     * @private
     */
    CalendarEvents.prototype.datesSubject;
    /**
     * @type {?}
     * @private
     */
    CalendarEvents.prototype.dates$;
}
