// @cemex-core/events-v7/esm5/lib/order.event.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
var OrderEvents = /** @class */ (function () {
    function OrderEvents() {
        // Observable items sources
        this.orderSelectedSubject = new BehaviorSubject(null);
        // Observable items streams
        this.order$ = this.orderSelectedSubject.asObservable();
    }
    /**
     * @param {?} order
     * @return {?}
     */
    OrderEvents.prototype.selectOrder = /**
     * @param {?} order
     * @return {?}
     */
    function (order) {
        /** @type {?} */
        var currentOrder = this.orderSelectedSubject.getValue();
        // Fire only if customer differs from the selected one
        if (currentOrder !== order) {
            this.orderSelectedSubject.next(order);
        }
    };
    /**
     * @return {?}
     */
    OrderEvents.prototype.getOrderSelected = /**
     * @return {?}
     */
    function () {
        return this.order$;
    };
    /**
     * @return {?}
     */
    OrderEvents.prototype.currentOrderSelected = /**
     * @return {?}
     */
    function () {
        return this.orderSelectedSubject.getValue();
    };
    OrderEvents.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    OrderEvents.ctorParameters = function () { return []; };
    return OrderEvents;
}());
export { OrderEvents };
if (false) {
    /**
     * @type {?}
     * @private
     */
    OrderEvents.prototype.orderSelectedSubject;
    /**
     * @type {?}
     * @private
     */
    OrderEvents.prototype.order$;
}
