// @cemex-core/events-v7/esm5/lib/sidebar.event.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { BehaviorSubject } from 'rxjs';
var SidebarEvents = /** @class */ (function () {
    function SidebarEvents(router) {
        this.router = router;
        // Observable items sources
        this.jobsiteSubject = new BehaviorSubject(null);
        this.customerSubject = new BehaviorSubject(null);
        this.customersSubject = new BehaviorSubject(null);
        // Observable items streams
        this.jobsite$ = this.jobsiteSubject.asObservable();
        this.customer$ = this.customerSubject.asObservable();
    }
    /**
     * @param {?} jobsite
     * @param {?=} redirect
     * @param {?=} lock
     * @return {?}
     */
    SidebarEvents.prototype.selectJobsite = /**
     * @param {?} jobsite
     * @param {?=} redirect
     * @param {?=} lock
     * @return {?}
     */
    function (jobsite, redirect, lock) {
        if (redirect === void 0) { redirect = false; }
        if (lock === void 0) { lock = false; }
        /** @type {?} */
        var currentJobsite = this.jobsiteSubject.getValue();
        // Fire only if customer differs from the selected one
        if (currentJobsite !== jobsite) {
            if (!lock) {
                this.jobsiteSubject.next(jobsite);
            }
        }
        if (jobsite && redirect) {
            this.router.navigate(['app/jobsite', jobsite.jobsiteId]);
        }
    };
    /**
     * @param {?} customer
     * @param {?=} redirect
     * @return {?}
     */
    SidebarEvents.prototype.selectCustomer = /**
     * @param {?} customer
     * @param {?=} redirect
     * @return {?}
     */
    function (customer, redirect) {
        if (redirect === void 0) { redirect = true; }
        /** @type {?} */
        var currentCustomer = this.customerSubject.getValue();
        // Fire only if customer differs from the selected one
        if (currentCustomer !== customer) {
            this.customerSubject.next(customer);
        }
        // If its not requesting an order directly
        if (redirect) {
            this.selectJobsite(null);
            this.router.navigate(['app/client-detail']);
        }
    };
    /**
     * @param {?} customers
     * @return {?}
     */
    SidebarEvents.prototype.setCustomers = /**
     * @param {?} customers
     * @return {?}
     */
    function (customers) {
        this.customersSubject.next(customers);
    };
    /**
     * @return {?}
     */
    SidebarEvents.prototype.getJobsiteSelected = /**
     * @return {?}
     */
    function () {
        return this.jobsite$;
    };
    /**
     * @return {?}
     */
    SidebarEvents.prototype.getCustomerSelected = /**
     * @return {?}
     */
    function () {
        return this.customer$;
    };
    /**
     * @param {?} code
     * @param {?=} redirect
     * @return {?}
     */
    SidebarEvents.prototype.selectCustomerByCode = /**
     * @param {?} code
     * @param {?=} redirect
     * @return {?}
     */
    function (code, redirect) {
        if (redirect === void 0) { redirect = false; }
        var e_1, _a;
        try {
            for (var _b = tslib_1.__values(this.customersSubject.getValue()), _c = _b.next(); !_c.done; _c = _b.next()) {
                var customer = _c.value;
                if (customer.customerCode === code) {
                    this.selectCustomer(customer, redirect);
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
    };
    /**
     * @return {?}
     */
    SidebarEvents.prototype.currentCustomer = /**
     * @return {?}
     */
    function () {
        return this.customerSubject.getValue();
    };
    /**
     * @return {?}
     */
    SidebarEvents.prototype.currentJobsite = /**
     * @return {?}
     */
    function () {
        return this.jobsiteSubject.getValue();
    };
    SidebarEvents.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    SidebarEvents.ctorParameters = function () { return [
        { type: Router }
    ]; };
    return SidebarEvents;
}());
export { SidebarEvents };
if (false) {
    /**
     * @type {?}
     * @private
     */
    SidebarEvents.prototype.jobsiteSubject;
    /**
     * @type {?}
     * @private
     */
    SidebarEvents.prototype.customerSubject;
    /**
     * @type {?}
     * @private
     */
    SidebarEvents.prototype.customersSubject;
    /**
     * @type {?}
     * @private
     */
    SidebarEvents.prototype.jobsite$;
    /** @type {?} */
    SidebarEvents.prototype.customer$;
    /**
     * @type {?}
     * @private
     */
    SidebarEvents.prototype.router;
}
