// @cemex/cmx-alert-v7/esm5/lib/components/alert.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { AlertService } from '../services/alert.service';
import { AlertType } from '../models/alert';
import { CmxFlugeeService } from '@cemex/cmx-flugee-v7';
var AlertComponent = /** @class */ (function () {
    function AlertComponent(alertService, flugeeService) {
        this.alertService = alertService;
        this.flugeeService = flugeeService;
        this.alerts$ = alertService.alerts$;
    }
    /**
     * @param {?} alert
     * @return {?}
     */
    AlertComponent.prototype.onLinkClick = /**
     * @param {?} alert
     * @return {?}
     */
    function (alert) {
        alert.onAction();
    };
    /**
     * @param {?} alert
     * @return {?}
     */
    AlertComponent.prototype.getCssClass = /**
     * @param {?} alert
     * @return {?}
     */
    function (alert) {
        return {
            'alert--success': AlertType.Success === alert.alertType,
            'alert--error': AlertType.Error === alert.alertType,
            'alert--warning': AlertType.Warning === alert.alertType,
            'alert--info': AlertType.Info === alert.alertType,
        };
    };
    /**
     * @param {?} alert
     * @return {?}
     */
    AlertComponent.prototype.onClose = /**
     * @param {?} alert
     * @return {?}
     */
    function (alert) {
        this.alertService.removeAlert(alert);
        alert.onClose();
    };
    /**
     * @return {?}
     */
    AlertComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.flugeeService.direction$.subscribe((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            _this.dirState = data;
        }));
    };
    AlertComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:component-selector
                    selector: 'cmx-alert',
                    template: "<div *ngFor=\"let alert of alerts$ | async\" class=\"alert\" [ngClass]=\"getCssClass(alert)\" dir={{dirState}}>\n  <div class=\"alert__text\">{{alert.alertMessage}} <span class=\"alert__link\" (click)=\"onLinkClick(alert)\">{{alert.actionMessage}}</span></div>\n  <div class=\"alert__close\" (click)=\"onClose(alert)\"><i class=\"alert__icon cmx-icon-close\"></i></div>\n</div>",
                    styles: ["@import url(https://www.cemexgo.com/cdn/cmx-dls-icons.css);@import url(https://www.cemexgo.com/cdn/cmx-fonts.css);.alert{width:100%;min-height:32px;color:#fff;font-size:.875rem;font-weight:700;line-height:1rem;position:relative;margin-bottom:10px;font-family:Roboto-Medium,sans-serif}.alert--error{background:#ee3d42}.alert--success{background:#7ac943}.alert--warning{background:#fbb03b}.alert--info{background:#3fa9f5}.alert__text{padding:8px 16px;font-size:.875rem;line-height:1rem;width:calc(100% - 70px);color:#fff}.alert__link{font-size:.875rem;font-weight:700;color:#fff;line-height:1rem;text-decoration:underline;cursor:pointer}.alert__close{position:absolute;right:10px;top:10px;cursor:pointer}.alert__icon{color:#fff}.alert--rtl,.alert[dir=rtl]{padding:8px 35px}"]
                }] }
    ];
    /** @nocollapse */
    AlertComponent.ctorParameters = function () { return [
        { type: AlertService },
        { type: CmxFlugeeService }
    ]; };
    return AlertComponent;
}());
export { AlertComponent };
if (false) {
    /** @type {?} */
    AlertComponent.prototype.alerts$;
    /** @type {?} */
    AlertComponent.prototype.dirState;
    /**
     * @type {?}
     * @private
     */
    AlertComponent.prototype.alertService;
    /**
     * @type {?}
     * @private
     */
    AlertComponent.prototype.flugeeService;
}
