// @cemex-core/angular-localization-v7/esm5/lib/services/direction.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import { TranslationService } from './translation.service';
var DirectionService = /** @class */ (function () {
    function DirectionService(translation) {
        var _this = this;
        this.translation = translation;
        this.direction$ = new Subject();
        this.rtl$ = new Subject();
        this.translation.getSelectedLanguage().subscribe((/**
         * @param {?} language
         * @return {?}
         */
        function (language) {
            _this.direction$.next(language.direction);
            _this.rtl$.next(language.rtl);
        }));
    }
    /**
     * @return {?}
     */
    DirectionService.prototype.getDirection = /**
     * @return {?}
     */
    function () {
        return this.direction$.asObservable();
    };
    /**
     * @return {?}
     */
    DirectionService.prototype.getRTL = /**
     * @return {?}
     */
    function () {
        return this.rtl$.asObservable();
    };
    /**
     * @param {?} languageCode
     * @return {?}
     */
    DirectionService.prototype.getDirectionFor = /**
     * @param {?} languageCode
     * @return {?}
     */
    function (languageCode) {
        /** @type {?} */
        var language = this.translation.getLanguage(languageCode.split(/[-_]/)[0]);
        this.direction$.next(language.dir);
    };
    /**
     * @param {?} languageCode
     * @return {?}
     */
    DirectionService.prototype.getRtlFor = /**
     * @param {?} languageCode
     * @return {?}
     */
    function (languageCode) {
        /** @type {?} */
        var language = this.translation.getLanguage(languageCode.split(/[-_]/)[0]);
        this.rtl$.next(language.rtl);
    };
    DirectionService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    DirectionService.ctorParameters = function () { return [
        { type: TranslationService }
    ]; };
    return DirectionService;
}());
export { DirectionService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    DirectionService.prototype.direction$;
    /**
     * @type {?}
     * @private
     */
    DirectionService.prototype.rtl$;
    /**
     * @type {?}
     * @private
     */
    DirectionService.prototype.translation;
}
