// @cemex-core/types-v1/dist/drivers.order.dto.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.STATUS_ALL = "All";
exports.STATUS_TICKETED = "Ticketed";
exports.STATUS_ON_SITE = "On_Site";
exports.STATUS_TO_JOB = "To_Job";
exports.STATUS_TO_PLANT = "To_Plant";
var OrderStatusTruck = /** @class */ (function () {
    function OrderStatusTruck(status) {
        this.statusCode = status;
        this.trucks = [];
    }
    Object.defineProperty(OrderStatusTruck.prototype, "statusDesc", {
        get: function () {
            var desc = this.statusCode.split("_").join(" ");
            return desc + " (" + this.trucks.length + ")";
        },
        enumerable: true,
        configurable: true
    });
    return OrderStatusTruck;
}());
exports.OrderStatusTruck = OrderStatusTruck;
var DriverOrder = /** @class */ (function () {
    function DriverOrder(options) {
        this.latitude = null;
        this.longitude = null;
        this.jobsiteId = null;
        this.jobsiteName = null;
        this.dictStatusTruck = {};
        this.orderId = options.orderId;
        this.orderCode = options.orderCode;
        this.ticketId = options.ticketId;
        this.ticketCode = options.ticketCode;
        this.totalQuantity = options.totalQuantity;
        this.um = options.um;
        this.volume = options.volume;
        this.podName = options.podName;
        this.clientName = options.clientName;
        this.plants = [];
        this.statusTrucks = [];
        this.dictStatusTruck[exports.STATUS_TICKETED] = [];
        this.dictStatusTruck[exports.STATUS_ON_SITE] = [];
        this.dictStatusTruck[exports.STATUS_TO_JOB] = [];
        this.dictStatusTruck[exports.STATUS_TO_PLANT] = [];
        this.statusTrucks[0] = new OrderStatusTruck(exports.STATUS_ALL);
        this.statusTrucks[1] = new OrderStatusTruck(exports.STATUS_TICKETED);
        this.statusTrucks[2] = new OrderStatusTruck(exports.STATUS_TO_JOB);
        this.statusTrucks[3] = new OrderStatusTruck(exports.STATUS_ON_SITE);
        this.statusTrucks[4] = new OrderStatusTruck(exports.STATUS_TO_PLANT);
    }
    return DriverOrder;
}());
exports.DriverOrder = DriverOrder;
