// @cemex-core/constants-v2/src/test/karma.conf.js
// Karma configuration file, see link for more information
// https://karma-runner.github.io/0.13/config/configuration-file.html

module.exports = function (config) {
    config.set({
        basePath: '.',
        frameworks: ['jasmine'],
        files: [
            '../../node_modules/zone.js/dist/zone.js',
            '../../node_modules/zone.js/dist/long-stack-trace-zone.js',
            '../../node_modules/zone.js/dist/proxy.js',
            '../../node_modules/zone.js/dist/sync-test.js',
            '../../node_modules/zone.js/dist/jasmine-patch.js',
            '../../node_modules/zone.js/dist/async-test.js',
            '../../node_modules/zone.js/dist/fake-async-test.js',
            'http://cdn.jsdelivr.net/jquery/1/jquery.min.js',
            'http://cdn.jsdelivr.net/momentjs/latest/moment.min.js',
            'boot-tests.ts',
        ],
        preprocessors: {
            'boot-tests.ts': ['webpack']
        },
        webpack: require('./webpack.config')({env: 'test'}),
        reporters: ['progress'],
        port: 9876,
        colors: true,
        logLevel: config.LOG_INFO,
        autoWatch: true,
        browsers: ['ChromeHeadlessNoSandbox'],
        customLaunchers: {
          ChromeHeadlessNoSandbox: {
            base: 'ChromeHeadless',
            flags: ['--no-sandbox']
          }
        },
        mime: { 'application/javascript': ['ts','tsx'] },
        singleRun: true
    });
};
