// @cemex/cmx-table-v7/esm5/lib/components/cmx-pagination/cmx-pagination.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, Host } from '@angular/core';
import { CmxTableComponent } from '../cmx-table/cmx-table.component';
var CmxPaginationComponent = /** @class */ (function () {
    function CmxPaginationComponent(parent) {
        var _this = this;
        this.parent = parent;
        this.pageChanged = new EventEmitter();
        this.beginning = [];
        this.end = [];
        this.state = 1;
        this.isRtl = false;
        this._currentPage = 1;
        this.range = 5;
        // strange shit that a delay is needed to set the value from the parent
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this.isRtl = _this.parent.rtl;
        }), 100);
    }
    Object.defineProperty(CmxPaginationComponent.prototype, "pages", {
        get: /**
         * @return {?}
         */
        function () {
            return this._pages;
        },
        set: /**
         * @param {?} $value
         * @return {?}
         */
        function ($value) {
            if ($value !== undefined && $value !== null) {
                this._pages = $value;
                this.organizePages();
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CmxPaginationComponent.prototype, "currentPage", {
        get: /**
         * @return {?}
         */
        function () {
            return this._currentPage;
        },
        set: /**
         * @param {?} $value
         * @return {?}
         */
        function ($value) {
            if ($value !== undefined && $value != null) {
                this._currentPage = $value;
                if (this._currentPage < this.range) {
                    this.state = 1;
                }
                else if (this._currentPage >= this.range &&
                    this._currentPage <= (this.pages - this.range)) {
                    this.state = 2;
                }
                else if (this._currentPage > (this.pages - this.range)) {
                    this.state = 3;
                }
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @description Goes one page down from the current one.
     */
    /**
     * \@description Goes one page down from the current one.
     * @return {?}
     */
    CmxPaginationComponent.prototype.previousPage = /**
     * \@description Goes one page down from the current one.
     * @return {?}
     */
    function () {
        /** @type {?} */
        var $calc = this.currentPage - 1;
        if ($calc > 0) {
            this.setPage($calc);
        }
    };
    /**
     * @description Goes one page up from the current one.
     */
    /**
     * \@description Goes one page up from the current one.
     * @return {?}
     */
    CmxPaginationComponent.prototype.nextPage = /**
     * \@description Goes one page up from the current one.
     * @return {?}
     */
    function () {
        /** @type {?} */
        var $calc = this.currentPage + 1;
        if ($calc <= this.pages) {
            this.setPage($calc);
        }
    };
    /**
     * @description Check if the last page shortcut is shown
     * whether the first page is present or not.
     * @returns boolean
     */
    /**
     * \@description Check if the last page shortcut is shown
     * whether the first page is present or not.
     * @return {?} boolean
     */
    CmxPaginationComponent.prototype.showFirstPageShortcut = /**
     * \@description Check if the last page shortcut is shown
     * whether the first page is present or not.
     * @return {?} boolean
     */
    function () {
        /** @type {?} */
        var $show = true;
        this.end.forEach((/**
         * @param {?} $page
         * @return {?}
         */
        function ($page) {
            if ($page === 1) {
                $show = false;
            }
        }));
        return $show && (this.state !== 1);
    };
    /**
     * @description Check if the last page shortcut is shown
     * whether the last page is present or not.
     * @returns boolean
     */
    /**
     * \@description Check if the last page shortcut is shown
     * whether the last page is present or not.
     * @return {?} boolean
     */
    CmxPaginationComponent.prototype.showLastPageShortcut = /**
     * \@description Check if the last page shortcut is shown
     * whether the last page is present or not.
     * @return {?} boolean
     */
    function () {
        var _this = this;
        /** @type {?} */
        var $show = true;
        this.beginning.forEach((/**
         * @param {?} $page
         * @return {?}
         */
        function ($page) {
            if ($page === _this.pages) {
                $show = false;
            }
        }));
        return $show && (this.state !== 3);
    };
    /**
     * @description Sets the current page from the user's click.
     * @param number $page The number of the page clicked.
     */
    /**
     * \@description Sets the current page from the user's click.
     * @param {?} $page
     * @return {?}
     */
    CmxPaginationComponent.prototype.setPage = /**
     * \@description Sets the current page from the user's click.
     * @param {?} $page
     * @return {?}
     */
    function ($page) {
        this.currentPage = $page;
        if (this.currentPage < this.range) {
            this.state = 1;
        }
        else if (this.currentPage >= this.range &&
            this.currentPage <= (this.pages - this.range)) {
            this.state = 2;
        }
        else if (this.currentPage > (this.pages - this.range)) {
            this.state = 3;
        }
        this.pageChanged.emit(this.currentPage);
    };
    /**
     * @private
     * @return {?}
     */
    CmxPaginationComponent.prototype.organizePages = /**
     * @private
     * @return {?}
     */
    function () {
        this.state = 1;
        this.currentPage = 1;
        this.beginning = [];
        this.end = [];
        /** @type {?} */
        var $i = 1;
        while ($i <= this.range && $i <= this.pages) {
            this.beginning.push($i);
            $i++;
        }
        $i = (this.pages - this.range);
        if ($i === 0) {
            $i++;
        }
        while ($i <= this.pages) {
            this.end.push($i);
            $i++;
        }
    };
    CmxPaginationComponent.decorators = [
        { type: Component, args: [{
                    selector: 'cmx-pagination',
                    template: "<ul class=\"pagination\" [dir]=\"isRtl ? 'rtl':''\">\r\n    <ng-container *ngIf=\"currentPage !== 1\">\r\n        <li class=\"pagination-prev\">\r\n            <a (click)=\"previousPage()\">\r\n                <ng-container *ngIf=\"isRtl === false\">\r\n                    <svg id=\"icon-straight-arrow-left\" viewBox=\"0 0 6 10\" width=\"100%\" height=\"100%\">\r\n                        <title>icon-straight-arrow-left</title>\r\n                        <polygon class=\"icon-straight-arrow-left\" points=\"4.86 0 6 1.18 2.29 5 6 8.83 4.86 10 0 5 4.86 0\"></polygon>\r\n                    </svg>\r\n                </ng-container>\r\n                <ng-container *ngIf=\"isRtl === true\">\r\n                    <svg id=\"icon-straight-arrow-right\" viewBox=\"0 0 6 10\" width=\"100%\" height=\"100%\">\r\n                        <title>icon-straight-arrow-right</title>\r\n                        <polygon class=\"icon-straight-arrow-right\" points=\"1.14 0 0 1.18 3.71 5 0 8.83 1.14 10 6 5 1.14 0\"></polygon>\r\n                    </svg>\r\n                </ng-container>\r\n            </a>\r\n        </li>\r\n    </ng-container>\r\n    <ng-container *ngIf=\"showFirstPageShortcut()\">\r\n        <li>\r\n            <a (click)=\"setPage(1)\">\r\n                1\r\n            </a>\r\n        </li>\r\n        <li>\r\n            ...\r\n        </li>\r\n    </ng-container>\r\n    <ng-container *ngIf=\"state === 1\">\r\n        <ng-container *ngFor=\"let $page of beginning;\">\r\n            <li>\r\n                <a [class.current]=\"$page === currentPage\" (click)=\"setPage($page)\">\r\n                    {{ $page }}\r\n                </a>\r\n            </li>    \r\n        </ng-container>\r\n    </ng-container>\r\n    <ng-container *ngIf=\"state === 2\">\r\n        <li>\r\n            <a (click)=\"setPage(currentPage - 1)\">\r\n                {{ currentPage - 1 }}\r\n            </a>\r\n        </li>\r\n        <li>\r\n            <a class=\"current\" (click)=\"setPage(currentPage)\">\r\n                {{ currentPage }}\r\n            </a>\r\n        </li>\r\n        <li>\r\n            <a (click)=\"setPage(currentPage + 1)\">\r\n                {{ currentPage + 1 }}\r\n            </a>\r\n        </li>   \r\n    </ng-container>\r\n    <ng-container *ngIf=\"state === 3\">\r\n        <ng-container *ngFor=\"let $page of end;\">\r\n            <li>\r\n                <a [class.current]=\"$page === currentPage\" (click)=\"setPage($page)\">\r\n                    {{ $page }}\r\n                </a>\r\n            </li>      \r\n        </ng-container>\r\n    </ng-container>\r\n    <ng-container *ngIf=\"showLastPageShortcut()\">\r\n        <li>\r\n            ...\r\n        </li>\r\n        <li>\r\n            <a (click)=\"setPage(pages)\">\r\n                {{ pages }}\r\n            </a>\r\n        </li>\r\n    </ng-container>\r\n    <ng-container *ngIf=\"currentPage !== pages\">\r\n        <li class=\"pagination-next\">\r\n            <a (click)=\"nextPage()\">\r\n                <ng-container *ngIf=\"isRtl === false\">\r\n                    <svg id=\"icon-straight-arrow-right\" viewBox=\"0 0 6 10\" width=\"100%\" height=\"100%\">\r\n                        <title>icon-straight-arrow-right</title>\r\n                        <polygon class=\"icon-straight-arrow-right\" points=\"1.14 0 0 1.18 3.71 5 0 8.83 1.14 10 6 5 1.14 0\"></polygon>\r\n                    </svg>\r\n                </ng-container>\r\n                <ng-container *ngIf=\"isRtl === true\">\r\n                    <svg id=\"icon-straight-arrow-left\" viewBox=\"0 0 6 10\" width=\"100%\" height=\"100%\">\r\n                        <title>icon-straight-arrow-left</title>\r\n                        <polygon class=\"icon-straight-arrow-left\" points=\"4.86 0 6 1.18 2.29 5 6 8.83 4.86 10 0 5 4.86 0\"></polygon>\r\n                    </svg>\r\n                </ng-container>\r\n            </a>\r\n        </li>\r\n    </ng-container>\r\n</ul>",
                    styles: [".roboto-bold{font-weight:700}.roboto-medium{font-weight:500}.roboto-regular{font-weight:400}.roboto-light{font-weight:300}.roboto-thin{font-weight:100}.h1-role,h1{font-family:Roboto,sans-serif;font-size:3rem;line-height:3rem;font-weight:700;color:#001b3a}.h2-role,.h3-role,h2,h3{font-family:Roboto,sans-serif;font-size:1.625rem;line-height:2rem;font-weight:400;color:#001b3a}.h4-role,h4{font-family:Roboto,sans-serif;font-size:2rem;line-height:2rem;font-weight:500;color:#001b3a}.h5-role,h5{font-family:Roboto,sans-serif;font-size:1.25rem;line-height:1.5rem;font-weight:400;color:#001b3a}.h6-role,h6{font-family:Roboto,sans-serif;font-size:.875rem;line-height:1rem;font-weight:700;color:#001b3a;text-transform:uppercase}.table-entry{font-family:Roboto,sans-serif;font-size:.875rem;line-height:1rem}.table-entry-strong{font-family:Roboto,sans-serif;font-size:.875rem;line-height:1rem;color:#333;font-weight:700}.link,.table-entry,a{text-decoration:none;color:#3fa9f5}a:hover{text-decoration:underline}.input-placeholder{font-family:Roboto,sans-serif;font-size:.875rem;line-height:1rem;color:#666}.input-text{font-family:Roboto,sans-serif;font-size:.875rem;line-height:1rem;color:#333}.size-15{font-size:.9375rem}.size-24{font-size:1.5rem}.text-left{text-align:left!important}.text-right{text-align:right!important}.text-center{text-align:center!important}.flex-first{order:-1}.flex-last{order:1}.flex-unordered{order:0}.flex-row{flex-direction:row!important}.flex-column{flex-direction:column!important}.flex-row-reverse{flex-direction:row-reverse!important}.flex-column-reverse{flex-direction:column-reverse!important}.flex-wrap{flex-wrap:wrap!important}.flex-nowrap{flex-wrap:nowrap!important}.flex-wrap-reverse{flex-wrap:wrap-reverse!important}.justify-content-start{justify-content:flex-start!important}.justify-content-end{justify-content:flex-end!important}.justify-content-center{justify-content:center!important}.justify-content-between{justify-content:space-between!important}.justify-content-around{justify-content:space-around!important}.align-items-start{align-items:flex-start!important}.align-items-end{align-items:flex-end!important}.align-items-center{align-items:center!important}.align-items-baseline{align-items:baseline!important}.align-items-stretch{align-items:stretch!important}.align-content-start{align-content:flex-start!important}.align-content-end{align-content:flex-end!important}.align-content-center{align-content:center!important}.align-content-between{align-content:space-between!important}.align-content-around{align-content:space-around!important}.align-content-stretch{align-content:stretch!important}.align-self-auto{-ms-grid-row-align:auto!important;align-self:auto!important}.align-self-start{align-self:flex-start!important}.align-self-end{align-self:flex-end!important}.align-self-center{-ms-grid-row-align:center!important;align-self:center!important}.align-self-baseline{align-self:baseline!important}.align-self-stretch{-ms-grid-row-align:stretch!important;align-self:stretch!important}@media (min-width:576px){.text-xs-left{text-align:left!important}.text-xs-right{text-align:right!important}.text-xs-center{text-align:center!important}.flex-xs-first{order:-1}.flex-xs-last{order:1}.flex-xs-unordered{order:0}.flex-xs-row{flex-direction:row!important}.flex-xs-column{flex-direction:column!important}.flex-xs-row-reverse{flex-direction:row-reverse!important}.flex-xs-column-reverse{flex-direction:column-reverse!important}.flex-xs-wrap{flex-wrap:wrap!important}.flex-xs-nowrap{flex-wrap:nowrap!important}.flex-xs-wrap-reverse{flex-wrap:wrap-reverse!important}.justify-content-xs-start{justify-content:flex-start!important}.justify-content-xs-end{justify-content:flex-end!important}.justify-content-xs-center{justify-content:center!important}.justify-content-xs-between{justify-content:space-between!important}.justify-content-xs-around{justify-content:space-around!important}.align-items-xs-start{align-items:flex-start!important}.align-items-xs-end{align-items:flex-end!important}.align-items-xs-center{align-items:center!important}.align-items-xs-baseline{align-items:baseline!important}.align-items-xs-stretch{align-items:stretch!important}.align-content-xs-start{align-content:flex-start!important}.align-content-xs-end{align-content:flex-end!important}.align-content-xs-center{align-content:center!important}.align-content-xs-between{align-content:space-between!important}.align-content-xs-around{align-content:space-around!important}.align-content-xs-stretch{align-content:stretch!important}.align-self-xs-auto{-ms-grid-row-align:auto!important;align-self:auto!important}.align-self-xs-start{align-self:flex-start!important}.align-self-xs-end{align-self:flex-end!important}.align-self-xs-center{-ms-grid-row-align:center!important;align-self:center!important}.align-self-xs-baseline{align-self:baseline!important}.align-self-xs-stretch{-ms-grid-row-align:stretch!important;align-self:stretch!important}}@media (min-width:768px){.text-sm-left{text-align:left!important}.text-sm-right{text-align:right!important}.text-sm-center{text-align:center!important}.flex-sm-first{order:-1}.flex-sm-last{order:1}.flex-sm-unordered{order:0}.flex-sm-row{flex-direction:row!important}.flex-sm-column{flex-direction:column!important}.flex-sm-row-reverse{flex-direction:row-reverse!important}.flex-sm-column-reverse{flex-direction:column-reverse!important}.flex-sm-wrap{flex-wrap:wrap!important}.flex-sm-nowrap{flex-wrap:nowrap!important}.flex-sm-wrap-reverse{flex-wrap:wrap-reverse!important}.justify-content-sm-start{justify-content:flex-start!important}.justify-content-sm-end{justify-content:flex-end!important}.justify-content-sm-center{justify-content:center!important}.justify-content-sm-between{justify-content:space-between!important}.justify-content-sm-around{justify-content:space-around!important}.align-items-sm-start{align-items:flex-start!important}.align-items-sm-end{align-items:flex-end!important}.align-items-sm-center{align-items:center!important}.align-items-sm-baseline{align-items:baseline!important}.align-items-sm-stretch{align-items:stretch!important}.align-content-sm-start{align-content:flex-start!important}.align-content-sm-end{align-content:flex-end!important}.align-content-sm-center{align-content:center!important}.align-content-sm-between{align-content:space-between!important}.align-content-sm-around{align-content:space-around!important}.align-content-sm-stretch{align-content:stretch!important}.align-self-sm-auto{-ms-grid-row-align:auto!important;align-self:auto!important}.align-self-sm-start{align-self:flex-start!important}.align-self-sm-end{align-self:flex-end!important}.align-self-sm-center{-ms-grid-row-align:center!important;align-self:center!important}.align-self-sm-baseline{align-self:baseline!important}.align-self-sm-stretch{-ms-grid-row-align:stretch!important;align-self:stretch!important}}@media (min-width:1080px){.text-md-left{text-align:left!important}.text-md-right{text-align:right!important}.text-md-center{text-align:center!important}.flex-md-first{order:-1}.flex-md-last{order:1}.flex-md-unordered{order:0}.flex-md-row{flex-direction:row!important}.flex-md-column{flex-direction:column!important}.flex-md-row-reverse{flex-direction:row-reverse!important}.flex-md-column-reverse{flex-direction:column-reverse!important}.flex-md-wrap{flex-wrap:wrap!important}.flex-md-nowrap{flex-wrap:nowrap!important}.flex-md-wrap-reverse{flex-wrap:wrap-reverse!important}.justify-content-md-start{justify-content:flex-start!important}.justify-content-md-end{justify-content:flex-end!important}.justify-content-md-center{justify-content:center!important}.justify-content-md-between{justify-content:space-between!important}.justify-content-md-around{justify-content:space-around!important}.align-items-md-start{align-items:flex-start!important}.align-items-md-end{align-items:flex-end!important}.align-items-md-center{align-items:center!important}.align-items-md-baseline{align-items:baseline!important}.align-items-md-stretch{align-items:stretch!important}.align-content-md-start{align-content:flex-start!important}.align-content-md-end{align-content:flex-end!important}.align-content-md-center{align-content:center!important}.align-content-md-between{align-content:space-between!important}.align-content-md-around{align-content:space-around!important}.align-content-md-stretch{align-content:stretch!important}.align-self-md-auto{-ms-grid-row-align:auto!important;align-self:auto!important}.align-self-md-start{align-self:flex-start!important}.align-self-md-end{align-self:flex-end!important}.align-self-md-center{-ms-grid-row-align:center!important;align-self:center!important}.align-self-md-baseline{align-self:baseline!important}.align-self-md-stretch{-ms-grid-row-align:stretch!important;align-self:stretch!important}}@media (min-width:1320px){.text-lg-left{text-align:left!important}.text-lg-right{text-align:right!important}.text-lg-center{text-align:center!important}.flex-lg-first{order:-1}.flex-lg-last{order:1}.flex-lg-unordered{order:0}.flex-lg-row{flex-direction:row!important}.flex-lg-column{flex-direction:column!important}.flex-lg-row-reverse{flex-direction:row-reverse!important}.flex-lg-column-reverse{flex-direction:column-reverse!important}.flex-lg-wrap{flex-wrap:wrap!important}.flex-lg-nowrap{flex-wrap:nowrap!important}.flex-lg-wrap-reverse{flex-wrap:wrap-reverse!important}.justify-content-lg-start{justify-content:flex-start!important}.justify-content-lg-end{justify-content:flex-end!important}.justify-content-lg-center{justify-content:center!important}.justify-content-lg-between{justify-content:space-between!important}.justify-content-lg-around{justify-content:space-around!important}.align-items-lg-start{align-items:flex-start!important}.align-items-lg-end{align-items:flex-end!important}.align-items-lg-center{align-items:center!important}.align-items-lg-baseline{align-items:baseline!important}.align-items-lg-stretch{align-items:stretch!important}.align-content-lg-start{align-content:flex-start!important}.align-content-lg-end{align-content:flex-end!important}.align-content-lg-center{align-content:center!important}.align-content-lg-between{align-content:space-between!important}.align-content-lg-around{align-content:space-around!important}.align-content-lg-stretch{align-content:stretch!important}.align-self-lg-auto{-ms-grid-row-align:auto!important;align-self:auto!important}.align-self-lg-start{align-self:flex-start!important}.align-self-lg-end{align-self:flex-end!important}.align-self-lg-center{-ms-grid-row-align:center!important;align-self:center!important}.align-self-lg-baseline{align-self:baseline!important}.align-self-lg-stretch{-ms-grid-row-align:stretch!important;align-self:stretch!important}}@media (min-width:1560px){.text-xlg-left{text-align:left!important}.text-xlg-right{text-align:right!important}.text-xlg-center{text-align:center!important}.flex-xlg-first{order:-1}.flex-xlg-last{order:1}.flex-xlg-unordered{order:0}.flex-xlg-row{flex-direction:row!important}.flex-xlg-column{flex-direction:column!important}.flex-xlg-row-reverse{flex-direction:row-reverse!important}.flex-xlg-column-reverse{flex-direction:column-reverse!important}.flex-xlg-wrap{flex-wrap:wrap!important}.flex-xlg-nowrap{flex-wrap:nowrap!important}.flex-xlg-wrap-reverse{flex-wrap:wrap-reverse!important}.justify-content-xlg-start{justify-content:flex-start!important}.justify-content-xlg-end{justify-content:flex-end!important}.justify-content-xlg-center{justify-content:center!important}.justify-content-xlg-between{justify-content:space-between!important}.justify-content-xlg-around{justify-content:space-around!important}.align-items-xlg-start{align-items:flex-start!important}.align-items-xlg-end{align-items:flex-end!important}.align-items-xlg-center{align-items:center!important}.align-items-xlg-baseline{align-items:baseline!important}.align-items-xlg-stretch{align-items:stretch!important}.align-content-xlg-start{align-content:flex-start!important}.align-content-xlg-end{align-content:flex-end!important}.align-content-xlg-center{align-content:center!important}.align-content-xlg-between{align-content:space-between!important}.align-content-xlg-around{align-content:space-around!important}.align-content-xlg-stretch{align-content:stretch!important}.align-self-xlg-auto{-ms-grid-row-align:auto!important;align-self:auto!important}.align-self-xlg-start{align-self:flex-start!important}.align-self-xlg-end{align-self:flex-end!important}.align-self-xlg-center{-ms-grid-row-align:center!important;align-self:center!important}.align-self-xlg-baseline{align-self:baseline!important}.align-self-xlg-stretch{-ms-grid-row-align:stretch!important;align-self:stretch!important}}.pagination{display:inline-block;list-style:none;padding:0;margin:0}.pagination li{margin:0 8px;float:left}.pagination li a{cursor:pointer;font-size:.875rem;line-height:22px;width:24px;height:24px;display:inline-block;color:#5c6570;float:left;text-decoration:none;border-radius:2px;text-align:center;border:1px solid transparent}.pagination li a.hover,.pagination li a:hover{border:1px solid #3fa9f5}.pagination li a.current{color:#fff;background:#3fa9f5}.pagination li svg{width:8px;height:12px;fill:#3fa9f5;transition:.25s}.pagination .pagination-next.hide,.pagination .pagination-prev.hide{display:none}.pagination .pagination-next.hover a,.pagination .pagination-next:hover a,.pagination .pagination-prev.hover a,.pagination .pagination-prev:hover a{border:0}.pagination .pagination-next.hover a svg,.pagination .pagination-next:hover a svg,.pagination .pagination-prev.hover a svg,.pagination .pagination-prev:hover a svg{width:16px;height:20px;fill:#3fa9f5}.pagination[dir=rtl] li{float:right}"]
                }] }
    ];
    /** @nocollapse */
    CmxPaginationComponent.ctorParameters = function () { return [
        { type: CmxTableComponent, decorators: [{ type: Host }] }
    ]; };
    CmxPaginationComponent.propDecorators = {
        pages: [{ type: Input }],
        currentPage: [{ type: Input }],
        pageChanged: [{ type: Output }]
    };
    return CmxPaginationComponent;
}());
export { CmxPaginationComponent };
if (false) {
    /** @type {?} */
    CmxPaginationComponent.prototype.pageChanged;
    /** @type {?} */
    CmxPaginationComponent.prototype.beginning;
    /** @type {?} */
    CmxPaginationComponent.prototype.end;
    /** @type {?} */
    CmxPaginationComponent.prototype.state;
    /** @type {?} */
    CmxPaginationComponent.prototype.isRtl;
    /**
     * @type {?}
     * @private
     */
    CmxPaginationComponent.prototype._pages;
    /**
     * @type {?}
     * @private
     */
    CmxPaginationComponent.prototype._currentPage;
    /**
     * @type {?}
     * @private
     */
    CmxPaginationComponent.prototype.range;
    /**
     * @type {?}
     * @private
     */
    CmxPaginationComponent.prototype.parent;
}
