// @cemex-core/angular-services-v7/esm2015/lib/analytics/countly.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class CountlyService {
    constructor() {
        this.defaultTrackings = [
            ['track_pageview'],
            ['track_clicks'],
            ['track_errors'],
            ['track_links'],
            ['track_forms'],
            ['collect_from_forms'] /*['track_sessions'] */
        ];
        this.Countly = ((/** @type {?} */ (window))).Countly;
    }
    /**
     * @return {?}
     */
    init() {
        this.addTracking('track_sessions');
        this.Countly.init();
    }
    /**
     * @param {?} flag
     * @return {?}
     */
    debug(flag) {
        this.Countly.debug = flag;
    }
    /**
     * @param {?} url
     * @param {?} appKey
     * @return {?}
     */
    startService(url, appKey) {
        if (this.Countly !== undefined) {
            this.Countly.url = url;
            this.Countly.app_key = appKey;
            this.Countly.q = this.defaultTrackings;
        }
    }
    /**
     * @param {?} key
     * @param {?=} value
     * @return {?}
     */
    addTracking(key, value) {
        if (!value) {
            this.Countly.q.push([key]);
        }
        else {
            this.Countly.q.push([key, value]);
        }
    }
    /**
     * @param {?} key
     * @return {?}
     */
    addTrackingPage(key) {
        this.Countly.q.push(['track_pageview', key]);
    }
    /**
     * @param {?} countryCode
     * @param {?} newId
     * @return {?}
     */
    changeId(countryCode, newId) {
        this.Countly.q.push(['change_id', countryCode + newId]);
    }
    /**
     * @param {?} userfields
     * @return {?}
     */
    identifyUser(userfields) {
        this.Countly.q.push(['user_details', userfields]);
    }
    /**
     * @return {?}
     */
    endSession() {
        this.Countly.q.push(['end_session']);
    }
    /**
     * @param {?} trackingjson
     * @return {?}
     */
    removeTracking(trackingjson) {
        this.Countly.q.forEach((/**
         * @param {?} element
         * @return {?}
         */
        element => {
            if (element.indexOf(trackingjson) !== -1) {
                element[0] = 'canceled';
            }
        }));
    }
}
if (false) {
    /** @type {?} */
    CountlyService.prototype.defaultTrackings;
    /** @type {?} */
    CountlyService.prototype.Countly;
}
