// @cemex-core/angular-services-v7/esm5/lib/analytics/countly.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var CountlyService = /** @class */ (function () {
    function CountlyService() {
        this.defaultTrackings = [
            ['track_pageview'],
            ['track_clicks'],
            ['track_errors'],
            ['track_links'],
            ['track_forms'],
            ['collect_from_forms'] /*['track_sessions'] */
        ];
        this.Countly = ((/** @type {?} */ (window))).Countly;
    }
    /**
     * @return {?}
     */
    CountlyService.prototype.init = /**
     * @return {?}
     */
    function () {
        this.addTracking('track_sessions');
        this.Countly.init();
    };
    /**
     * @param {?} flag
     * @return {?}
     */
    CountlyService.prototype.debug = /**
     * @param {?} flag
     * @return {?}
     */
    function (flag) {
        this.Countly.debug = flag;
    };
    /**
     * @param {?} url
     * @param {?} appKey
     * @return {?}
     */
    CountlyService.prototype.startService = /**
     * @param {?} url
     * @param {?} appKey
     * @return {?}
     */
    function (url, appKey) {
        if (this.Countly !== undefined) {
            this.Countly.url = url;
            this.Countly.app_key = appKey;
            this.Countly.q = this.defaultTrackings;
        }
    };
    /**
     * @param {?} key
     * @param {?=} value
     * @return {?}
     */
    CountlyService.prototype.addTracking = /**
     * @param {?} key
     * @param {?=} value
     * @return {?}
     */
    function (key, value) {
        if (!value) {
            this.Countly.q.push([key]);
        }
        else {
            this.Countly.q.push([key, value]);
        }
    };
    /**
     * @param {?} key
     * @return {?}
     */
    CountlyService.prototype.addTrackingPage = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        this.Countly.q.push(['track_pageview', key]);
    };
    /**
     * @param {?} countryCode
     * @param {?} newId
     * @return {?}
     */
    CountlyService.prototype.changeId = /**
     * @param {?} countryCode
     * @param {?} newId
     * @return {?}
     */
    function (countryCode, newId) {
        this.Countly.q.push(['change_id', countryCode + newId]);
    };
    /**
     * @param {?} userfields
     * @return {?}
     */
    CountlyService.prototype.identifyUser = /**
     * @param {?} userfields
     * @return {?}
     */
    function (userfields) {
        this.Countly.q.push(['user_details', userfields]);
    };
    /**
     * @return {?}
     */
    CountlyService.prototype.endSession = /**
     * @return {?}
     */
    function () {
        this.Countly.q.push(['end_session']);
    };
    /**
     * @param {?} trackingjson
     * @return {?}
     */
    CountlyService.prototype.removeTracking = /**
     * @param {?} trackingjson
     * @return {?}
     */
    function (trackingjson) {
        this.Countly.q.forEach((/**
         * @param {?} element
         * @return {?}
         */
        function (element) {
            if (element.indexOf(trackingjson) !== -1) {
                element[0] = 'canceled';
            }
        }));
    };
    return CountlyService;
}());
export { CountlyService };
if (false) {
    /** @type {?} */
    CountlyService.prototype.defaultTrackings;
    /** @type {?} */
    CountlyService.prototype.Countly;
}
