// @cemex-core/angular-services-v7/esm5/lib/analytics/ianalytics.interface.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function IAnalyticsService() { }
if (false) {
    /**
     * @param {?} url
     * @param {?} appKey
     * @return {?}
     */
    IAnalyticsService.prototype.startService = function (url, appKey) { };
    /**
     * @param {?} flag
     * @return {?}
     */
    IAnalyticsService.prototype.debug = function (flag) { };
    /**
     * @param {?} key
     * @param {?=} value
     * @return {?}
     */
    IAnalyticsService.prototype.addTracking = function (key, value) { };
    /**
     * @param {?} userfields
     * @return {?}
     */
    IAnalyticsService.prototype.identifyUser = function (userfields) { };
    /**
     * @param {?} countryCode
     * @param {?} newId
     * @return {?}
     */
    IAnalyticsService.prototype.changeId = function (countryCode, newId) { };
    /**
     * @param {?} page
     * @return {?}
     */
    IAnalyticsService.prototype.addTrackingPage = function (page) { };
    /**
     * @return {?}
     */
    IAnalyticsService.prototype.endSession = function () { };
    /**
     * @return {?}
     */
    IAnalyticsService.prototype.init = function () { };
}
/**
 * @record
 */
export function IAnalyticsConfiguration() { }
if (false) {
    /** @type {?} */
    IAnalyticsConfiguration.prototype.url;
    /** @type {?} */
    IAnalyticsConfiguration.prototype.key;
}
/**
 * @record
 */
export function IAnalyticsUser() { }
if (false) {
    /** @type {?} */
    IAnalyticsUser.prototype.userId;
    /** @type {?} */
    IAnalyticsUser.prototype.email;
    /** @type {?} */
    IAnalyticsUser.prototype.name;
    /** @type {?} */
    IAnalyticsUser.prototype.phone;
    /** @type {?} */
    IAnalyticsUser.prototype.username;
}
