// @cemex-core/events-v7/esm2015/lib/calendar.event.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
export class CalendarEvents {
    constructor() {
        this.datesSubject = new BehaviorSubject({
            start: this.moment(),
            end: this.moment()
        });
        // Observable items streams
        this.dates$ = this.datesSubject.asObservable();
    }
    /**
     * @param {?} dates
     * @return {?}
     */
    setDates(dates) {
        ((/** @type {?} */ (window))).localStorage.setItem('dates', JSON.stringify(dates));
        /** @type {?} */
        const currentDates = this.getCurrentValue();
        if (dates.start.diff(currentDates.start, 'days') !== 0 || dates.end.diff(currentDates.end, 'days') !== 0) {
            this.datesSubject.next(dates);
        }
    }
    /**
     * @return {?}
     */
    getDates() {
        return this.dates$;
    }
    /**
     * @return {?}
     */
    getCurrentValue() {
        return this.datesSubject.getValue();
    }
}
CalendarEvents.decorators = [
    { type: Injectable }
];
if (false) {
    /** @type {?} */
    CalendarEvents.prototype.moment;
    /**
     * @type {?}
     * @private
     */
    CalendarEvents.prototype.datesSubject;
    /**
     * @type {?}
     * @private
     */
    CalendarEvents.prototype.dates$;
}
