// @cemex-core/events-v7/esm2015/lib/order.event.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
export class OrderEvents {
    constructor() {
        // Observable items sources
        this.orderSelectedSubject = new BehaviorSubject(null);
        // Observable items streams
        this.order$ = this.orderSelectedSubject.asObservable();
    }
    /**
     * @param {?} order
     * @return {?}
     */
    selectOrder(order) {
        /** @type {?} */
        const currentOrder = this.orderSelectedSubject.getValue();
        // Fire only if customer differs from the selected one
        if (currentOrder !== order) {
            this.orderSelectedSubject.next(order);
        }
    }
    /**
     * @return {?}
     */
    getOrderSelected() {
        return this.order$;
    }
    /**
     * @return {?}
     */
    currentOrderSelected() {
        return this.orderSelectedSubject.getValue();
    }
}
OrderEvents.decorators = [
    { type: Injectable }
];
/** @nocollapse */
OrderEvents.ctorParameters = () => [];
if (false) {
    /**
     * @type {?}
     * @private
     */
    OrderEvents.prototype.orderSelectedSubject;
    /**
     * @type {?}
     * @private
     */
    OrderEvents.prototype.order$;
}
