// @cemex-core/events-v7/esm2015/lib/sidebar.event.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { BehaviorSubject } from 'rxjs';
export class SidebarEvents {
    /**
     * @param {?} router
     */
    constructor(router) {
        this.router = router;
        // Observable items sources
        this.jobsiteSubject = new BehaviorSubject(null);
        this.customerSubject = new BehaviorSubject(null);
        this.customersSubject = new BehaviorSubject(null);
        // Observable items streams
        this.jobsite$ = this.jobsiteSubject.asObservable();
        this.customer$ = this.customerSubject.asObservable();
    }
    /**
     * @param {?} jobsite
     * @param {?=} redirect
     * @param {?=} lock
     * @return {?}
     */
    selectJobsite(jobsite, redirect = false, lock = false) {
        /** @type {?} */
        const currentJobsite = this.jobsiteSubject.getValue();
        // Fire only if customer differs from the selected one
        if (currentJobsite !== jobsite) {
            if (!lock) {
                this.jobsiteSubject.next(jobsite);
            }
        }
        if (jobsite && redirect) {
            this.router.navigate(['app/jobsite', jobsite.jobsiteId]);
        }
    }
    /**
     * @param {?} customer
     * @param {?=} redirect
     * @return {?}
     */
    selectCustomer(customer, redirect = true) {
        /** @type {?} */
        const currentCustomer = this.customerSubject.getValue();
        // Fire only if customer differs from the selected one
        if (currentCustomer !== customer) {
            this.customerSubject.next(customer);
        }
        // If its not requesting an order directly
        if (redirect) {
            this.selectJobsite(null);
            this.router.navigate(['app/client-detail']);
        }
    }
    /**
     * @param {?} customers
     * @return {?}
     */
    setCustomers(customers) {
        this.customersSubject.next(customers);
    }
    /**
     * @return {?}
     */
    getJobsiteSelected() {
        return this.jobsite$;
    }
    /**
     * @return {?}
     */
    getCustomerSelected() {
        return this.customer$;
    }
    /**
     * @param {?} code
     * @param {?=} redirect
     * @return {?}
     */
    selectCustomerByCode(code, redirect = false) {
        for (const customer of this.customersSubject.getValue()) {
            if (customer.customerCode === code) {
                this.selectCustomer(customer, redirect);
            }
        }
    }
    /**
     * @return {?}
     */
    currentCustomer() {
        return this.customerSubject.getValue();
    }
    /**
     * @return {?}
     */
    currentJobsite() {
        return this.jobsiteSubject.getValue();
    }
}
SidebarEvents.decorators = [
    { type: Injectable }
];
/** @nocollapse */
SidebarEvents.ctorParameters = () => [
    { type: Router }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    SidebarEvents.prototype.jobsiteSubject;
    /**
     * @type {?}
     * @private
     */
    SidebarEvents.prototype.customerSubject;
    /**
     * @type {?}
     * @private
     */
    SidebarEvents.prototype.customersSubject;
    /**
     * @type {?}
     * @private
     */
    SidebarEvents.prototype.jobsite$;
    /** @type {?} */
    SidebarEvents.prototype.customer$;
    /**
     * @type {?}
     * @private
     */
    SidebarEvents.prototype.router;
}
