// @cemex/cmx-dialog-v7/esm5/lib/services/cmx-dialog.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var CmxDialogService = /** @class */ (function () {
    function CmxDialogService() {
        this.cmxDialogs = new Array();
    }
    /**
     * @param {?} cmxDialog
     * @return {?}
     */
    CmxDialogService.prototype.addDialog = /**
     * @param {?} cmxDialog
     * @return {?}
     */
    function (cmxDialog) {
        this.hideLastDialog();
        this.cmxDialogs.push(cmxDialog);
    };
    /**
     * @param {?} cmxDialog
     * @param {?} showLastDialog
     * @return {?}
     */
    CmxDialogService.prototype.removeDialog = /**
     * @param {?} cmxDialog
     * @param {?} showLastDialog
     * @return {?}
     */
    function (cmxDialog, showLastDialog) {
        this.cmxDialogs.splice(this.cmxDialogs.length - 1, 1);
        if (showLastDialog) {
            this.showLastDialog();
        }
    };
    /**
     * @return {?}
     */
    CmxDialogService.prototype.closeAllDialogs = /**
     * @return {?}
     */
    function () {
        for (var i = this.cmxDialogs.length - 1; i >= 0; i--) {
            this.cmxDialogs[i].close(false);
        }
    };
    /**
     * @private
     * @return {?}
     */
    CmxDialogService.prototype.showLastDialog = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.cmxDialogs.length > 0) {
            this.cmxDialogs[this.cmxDialogs.length - 1].show();
        }
    };
    /**
     * @private
     * @return {?}
     */
    CmxDialogService.prototype.hideLastDialog = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.cmxDialogs.length > 0) {
            this.cmxDialogs[this.cmxDialogs.length - 1].hide();
        }
    };
    CmxDialogService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root',
                },] }
    ];
    /** @nocollapse */
    CmxDialogService.ctorParameters = function () { return []; };
    /** @nocollapse */ CmxDialogService.ngInjectableDef = i0.defineInjectable({ factory: function CmxDialogService_Factory() { return new CmxDialogService(); }, token: CmxDialogService, providedIn: "root" });
    return CmxDialogService;
}());
export { CmxDialogService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    CmxDialogService.prototype.cmxDialogs;
}
