// @cemex/cmx-table-v7/esm2015/lib/classes/table.class.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class Table {
    constructor() {
        this.columns = [];
    }
    /**
     * @param {?} columns
     * @return {?}
     */
    setColumns(columns) {
        this.columns = columns;
    }
    /**
     * @param {?} column
     * @return {?}
     */
    addColumn(column) {
        this.columns.push(column);
    }
    /**
     * @param {?} key
     * @return {?}
     */
    getColumn(key) {
        /** @type {?} */
        let answer;
        this.columns.forEach((/**
         * @param {?} column
         * @return {?}
         */
        (column) => {
            if (column.getKey() === key) {
                answer = column;
            }
        }));
        return answer;
    }
    /**
     * @return {?}
     */
    getColumns() {
        return this.columns;
    }
    /**
     * @param {?} _customerCode
     * @return {?}
     */
    setCustomerCode(_customerCode) {
        this.customerCode = _customerCode;
    }
    /**
     * @param {?} _companyCode
     * @return {?}
     */
    setCompanyCode(_companyCode) {
        this.companyCode = _companyCode;
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    Table.prototype.customerCode;
    /**
     * @type {?}
     * @private
     */
    Table.prototype.companyCode;
    /**
     * @type {?}
     * @private
     */
    Table.prototype.columns;
}
