// @cemex-core/angular-services-v7/esm5/lib/push-notifications/push-notifications.constants.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
var PushNotificationsConstants = /** @class */ (function () {
    function PushNotificationsConstants() {
        // tslint:disable-next-line:no-string-literal
        this.appCode = window['APP_CODE']; // 'Foreman_App'
        // tslint:disable-next-line:no-string-literal
        this.apiClientId = window['CLIENT_ID']; // 'f624df7f-9b35-41c0-8a41-4854b86c84f1';
        // tslint:disable-next-line:no-string-literal
        this.apiHost = window['API_HOST']; // 'https://uscldcnxapmd01.azure-api.net/';
        // tslint:disable-next-line:no-string-literal
        this.appLanguage = window['LANGUAGE']; // 'es-MX';
        this.authToken = 'Bearer ' + sessionStorage.getItem('auth_token');
        this.jwt = sessionStorage.getItem('jwt');
        this.urlServiceWorker = 'service-worker/push_sw.js';
        this.deviceApi = this.apiHost + 'v5/cm/notifications/devices';
        this.sourcePush = 'postMessageSW';
        this.vapidPublicKey = '';
    }
    /**
     * @return {?}
     */
    PushNotificationsConstants.prototype.getAppCode = /**
     * @return {?}
     */
    function () {
        return this.appCode;
    };
    /**
     * @return {?}
     */
    PushNotificationsConstants.prototype.getClientId = /**
     * @return {?}
     */
    function () {
        return this.apiClientId;
    };
    /**
     * @return {?}
     */
    PushNotificationsConstants.prototype.getApiBasePath = /**
     * @return {?}
     */
    function () {
        return this.apiBasePath;
    };
    /**
     * @return {?}
     */
    PushNotificationsConstants.prototype.getAppLanguage = /**
     * @return {?}
     */
    function () {
        return this.appLanguage;
    };
    /**
     * @return {?}
     */
    PushNotificationsConstants.prototype.getAuthToken = /**
     * @return {?}
     */
    function () {
        return this.authToken;
    };
    /**
     * @return {?}
     */
    PushNotificationsConstants.prototype.getJwt = /**
     * @return {?}
     */
    function () {
        return this.jwt;
    };
    /**
     * @return {?}
     */
    PushNotificationsConstants.prototype.getUrlSW = /**
     * @return {?}
     */
    function () {
        return this.urlServiceWorker;
    };
    /**
     * @return {?}
     */
    PushNotificationsConstants.prototype.getDevicesPath = /**
     * @return {?}
     */
    function () {
        return this.deviceApi;
    };
    /**
     * @return {?}
     */
    PushNotificationsConstants.prototype.getSourcePush = /**
     * @return {?}
     */
    function () {
        return this.sourcePush;
    };
    /**
     * @return {?}
     */
    PushNotificationsConstants.prototype.getVapidPublicKey = /**
     * @return {?}
     */
    function () {
        return this.vapidPublicKey;
    };
    /**
     * @param {?} pVapidPublicKey
     * @return {?}
     */
    PushNotificationsConstants.prototype.setVapidPublicKey = /**
     * @param {?} pVapidPublicKey
     * @return {?}
     */
    function (pVapidPublicKey) {
        this.vapidPublicKey = pVapidPublicKey;
    };
    /**
     * @param {?} urlSW
     * @return {?}
     */
    PushNotificationsConstants.prototype.setUrlSW = /**
     * @param {?} urlSW
     * @return {?}
     */
    function (urlSW) {
        this.urlServiceWorker = urlSW;
    };
    PushNotificationsConstants.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    PushNotificationsConstants.ctorParameters = function () { return []; };
    return PushNotificationsConstants;
}());
export { PushNotificationsConstants };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PushNotificationsConstants.prototype.appCode;
    /**
     * @type {?}
     * @private
     */
    PushNotificationsConstants.prototype.apiClientId;
    /**
     * @type {?}
     * @private
     */
    PushNotificationsConstants.prototype.appLanguage;
    /**
     * @type {?}
     * @private
     */
    PushNotificationsConstants.prototype.urlServiceWorker;
    /**
     * @type {?}
     * @private
     */
    PushNotificationsConstants.prototype.deviceApi;
    /**
     * @type {?}
     * @private
     */
    PushNotificationsConstants.prototype.apiBasePath;
    /**
     * @type {?}
     * @private
     */
    PushNotificationsConstants.prototype.apiHost;
    /**
     * @type {?}
     * @private
     */
    PushNotificationsConstants.prototype.apiOrg;
    /**
     * @type {?}
     * @private
     */
    PushNotificationsConstants.prototype.apiEnv;
    /**
     * @type {?}
     * @private
     */
    PushNotificationsConstants.prototype.authToken;
    /**
     * @type {?}
     * @private
     */
    PushNotificationsConstants.prototype.jwt;
    /**
     * @type {?}
     * @private
     */
    PushNotificationsConstants.prototype.sourcePush;
    /**
     * @type {?}
     * @private
     */
    PushNotificationsConstants.prototype.vapidPublicKey;
}
