// @cemex-core/constants-v2/dist/project.constants.js
import { Injectable } from '@angular/core';
var ProjectConstants = /** @class */ (function () {
    function ProjectConstants() {
        /* tslint:disable:no-string-literal */
        this.appCode = window['APP_CODE'];
        this.clientId = window['CLIENT_ID'];
        this.apiBasePath = window['API_HOST'];
        this.apiOrg = window['API_ORG'];
        this.apiEnv = window['API_ENV'];
    }
    ProjectConstants.prototype.getAppCode = function () {
        return this.appCode;
    };
    ProjectConstants.prototype.getClientId = function () {
        return this.clientId;
    };
    ProjectConstants.prototype.getApiBasePath = function () {
        return this.apiBasePath;
    };
    ProjectConstants.prototype.getApiOrg = function () {
        return this.apiOrg;
    };
    ProjectConstants.prototype.getApiEnv = function () {
        return this.apiEnv;
    };
    ProjectConstants.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    ProjectConstants.ctorParameters = function () { return []; };
    return ProjectConstants;
}());
export { ProjectConstants };
