// @cemex/cmx-form-field-v4/dist/cmx-form-field.module.js
import { CommonModule } from "@angular/common";
import { NgModule, NO_ERRORS_SCHEMA, CUSTOM_ELEMENTS_SCHEMA } from "@angular/core";
import { FormsModule, ReactiveFormsModule } from "@angular/forms";
import { CmxDatepickerModule } from "@cemex/cmx-datepicker-v4";
import { CmxFormFieldComponent } from "./components/cmx-form-field";
import { CmxOptionComponent } from "./components/cmx-option";
import { CmxSelectComponent } from "./components/cmx-select";
import { DatepickerInputComponent } from "./components/datepicker-input";
import { NumberInputComponent } from "./components/number-input";
import { TimeInputComponent } from "./components/time-input";
import { ErrorMessageDirective } from "./directives";
import { DateFormatService } from "./services";
import { CmxCheckboxModule } from "@cemex/cmx-checkbox-v4";
import { FormatterService } from "./services/formatter.service";
import { CurrencyMaskModule } from "./components/number-input/currency-mask.module";
var CmxFormFieldModule = /** @class */ (function () {
    function CmxFormFieldModule() {
    }
    CmxFormFieldModule.forRoot = function (formatter) {
        return {
            ngModule: CmxFormFieldModule,
            providers: [
                { provide: FormatterService, useExisting: formatter },
            ],
        };
    };
    CmxFormFieldModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        CmxFormFieldComponent,
                        CmxOptionComponent,
                        CmxSelectComponent,
                        DatepickerInputComponent,
                        NumberInputComponent,
                        TimeInputComponent,
                        ErrorMessageDirective,
                    ],
                    exports: [
                        CmxFormFieldComponent,
                        CmxOptionComponent,
                        CmxSelectComponent,
                        DatepickerInputComponent,
                        NumberInputComponent,
                        TimeInputComponent,
                        ErrorMessageDirective,
                    ],
                    imports: [
                        CommonModule,
                        FormsModule,
                        ReactiveFormsModule,
                        CmxDatepickerModule,
                        CmxCheckboxModule,
                        CurrencyMaskModule,
                    ],
                    providers: [
                        DateFormatService,
                    ],
                    schemas: [
                        CUSTOM_ELEMENTS_SCHEMA,
                        NO_ERRORS_SCHEMA,
                    ],
                },] },
    ];
    /** @nocollapse */
    CmxFormFieldModule.ctorParameters = function () { return []; };
    return CmxFormFieldModule;
}());
export { CmxFormFieldModule };
