// @cemex/cmx-table-v7/esm5/lib/classes/table.class.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var Table = /** @class */ (function () {
    function Table() {
        this.columns = [];
    }
    /**
     * @param {?} columns
     * @return {?}
     */
    Table.prototype.setColumns = /**
     * @param {?} columns
     * @return {?}
     */
    function (columns) {
        this.columns = columns;
    };
    /**
     * @param {?} column
     * @return {?}
     */
    Table.prototype.addColumn = /**
     * @param {?} column
     * @return {?}
     */
    function (column) {
        this.columns.push(column);
    };
    /**
     * @param {?} key
     * @return {?}
     */
    Table.prototype.getColumn = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        /** @type {?} */
        var answer;
        this.columns.forEach((/**
         * @param {?} column
         * @return {?}
         */
        function (column) {
            if (column.getKey() === key) {
                answer = column;
            }
        }));
        return answer;
    };
    /**
     * @return {?}
     */
    Table.prototype.getColumns = /**
     * @return {?}
     */
    function () {
        return this.columns;
    };
    /**
     * @param {?} _customerCode
     * @return {?}
     */
    Table.prototype.setCustomerCode = /**
     * @param {?} _customerCode
     * @return {?}
     */
    function (_customerCode) {
        this.customerCode = _customerCode;
    };
    /**
     * @param {?} _companyCode
     * @return {?}
     */
    Table.prototype.setCompanyCode = /**
     * @param {?} _companyCode
     * @return {?}
     */
    function (_companyCode) {
        this.companyCode = _companyCode;
    };
    return Table;
}());
export { Table };
if (false) {
    /**
     * @type {?}
     * @private
     */
    Table.prototype.customerCode;
    /**
     * @type {?}
     * @private
     */
    Table.prototype.companyCode;
    /**
     * @type {?}
     * @private
     */
    Table.prototype.columns;
}
