// @cemex-core/angular-localization-v7/esm5/lib/pipes/cmx-number.pipe.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Pipe } from '@angular/core';
import { FormatterService } from './../services/formatter.service';
import { TranslationService } from './../services/translation.service';
var CmxNumberPipe = /** @class */ (function () {
    function CmxNumberPipe(translationService, formatterService) {
        this.translationService = translationService;
        this.formatterService = formatterService;
    }
    /**
     * @param {?} numberEntry
     * @return {?}
     */
    CmxNumberPipe.prototype.transform = /**
     * @param {?} numberEntry
     * @return {?}
     */
    function (numberEntry) {
        return this.formatterService.formatNumberTS(numberEntry, this.translationService.selectedLanguage.languageISO.replace('_', '-'), this.translationService.selectedLanguage.decimalNumbers);
    };
    CmxNumberPipe.decorators = [
        { type: Pipe, args: [{
                    name: 'cmxNumber',
                    pure: false
                },] }
    ];
    /** @nocollapse */
    CmxNumberPipe.ctorParameters = function () { return [
        { type: TranslationService },
        { type: FormatterService }
    ]; };
    return CmxNumberPipe;
}());
export { CmxNumberPipe };
if (false) {
    /**
     * @type {?}
     * @private
     */
    CmxNumberPipe.prototype.translationService;
    /**
     * @type {?}
     * @private
     */
    CmxNumberPipe.prototype.formatterService;
}
