// @cemex-core/angular-localization-v7/esm5/lib/pipes/cmx-date.pipe.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Pipe } from '@angular/core';
import { DateFormatterService } from './../services/date-formatter.service';
import { TranslationService } from './../services/translation.service';
var CmxDatePipe = /** @class */ (function () {
    function CmxDatePipe(translation, dateFormatter) {
        this.translation = translation;
        this.dateFormatter = dateFormatter;
        this.value = '';
        this.monthsNames = [];
        this.daysNames = [];
    }
    /**
     * @param {?} $value
     * @param {?=} $countryCode
     * @return {?}
     */
    CmxDatePipe.prototype.transform = /**
     * @param {?} $value
     * @param {?=} $countryCode
     * @return {?}
     */
    function ($value, $countryCode) {
        try {
            if (typeof $value === 'number') {
                $value = new Date($value);
            }
            if (typeof $value === 'string') {
                $value = this.dateFormatter.getDateObject($value);
            }
            if ($countryCode === undefined || $countryCode === null) {
                $countryCode = sessionStorage.getItem('country');
            }
            if ($value !== undefined && $value instanceof Date) {
                /** @type {?} */
                var $currentLanguage = localStorage.getItem('language');
                /** @type {?} */
                var $lang = this.translation.getLanguage($currentLanguage ? $currentLanguage : 'en_US');
                this.daysNames = $lang.dayNames.split(',');
                this.monthsNames = $lang.monthNames.split(',');
                this.value = this.dateFormatter.format($value, $lang.formatDate, this.daysNames, this.monthsNames);
            }
            else {
                console.error('Date trying to transform: ', $value);
                this.value = '';
            }
        }
        catch ($exception) {
            console.error('Date trying to transform: ', $value);
            console.error($exception);
            this.value = '';
        }
        return this.value;
    };
    CmxDatePipe.decorators = [
        { type: Pipe, args: [{
                    name: 'cmxDate',
                    pure: true
                },] }
    ];
    /** @nocollapse */
    CmxDatePipe.ctorParameters = function () { return [
        { type: TranslationService },
        { type: DateFormatterService }
    ]; };
    return CmxDatePipe;
}());
export { CmxDatePipe };
if (false) {
    /**
     * @type {?}
     * @private
     */
    CmxDatePipe.prototype.value;
    /**
     * @type {?}
     * @private
     */
    CmxDatePipe.prototype.monthsNames;
    /**
     * @type {?}
     * @private
     */
    CmxDatePipe.prototype.daysNames;
    /**
     * @type {?}
     * @private
     */
    CmxDatePipe.prototype.translation;
    /**
     * @type {?}
     * @private
     */
    CmxDatePipe.prototype.dateFormatter;
}
