// @cemex-core/angular-localization-v7/esm5/lib/pipes/cmx-date-time-moment.pipe.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Pipe } from '@angular/core';
import { DateTimeActions, FormatterService } from './../services/formatter.service';
import { TranslationService } from './../services/translation.service';
var CmxDateTimeMomentPipe = /** @class */ (function () {
    function CmxDateTimeMomentPipe(translationService, formatterService) {
        this.translationService = translationService;
        this.formatterService = formatterService;
    }
    /**
     * @param {?} dateTime
     * @param {?=} action
     * @return {?}
     */
    CmxDateTimeMomentPipe.prototype.transform = /**
     * @param {?} dateTime
     * @param {?=} action
     * @return {?}
     */
    function (dateTime, action) {
        if (action === void 0) { action = DateTimeActions.ConvertToLocal; }
        return this.formatterService.formatDateTimeMoment(dateTime, this.translationService.selectedLanguage.languageISO.replace('_', '-'), this.translationService.selectedLanguage.formatDate +
            ' ' +
            this.translationService.selectedLanguage.formatTime, action);
    };
    CmxDateTimeMomentPipe.decorators = [
        { type: Pipe, args: [{
                    name: 'cmxDateTimeMoment',
                    pure: false
                },] }
    ];
    /** @nocollapse */
    CmxDateTimeMomentPipe.ctorParameters = function () { return [
        { type: TranslationService },
        { type: FormatterService }
    ]; };
    return CmxDateTimeMomentPipe;
}());
export { CmxDateTimeMomentPipe };
if (false) {
    /**
     * @type {?}
     * @private
     */
    CmxDateTimeMomentPipe.prototype.translationService;
    /**
     * @type {?}
     * @private
     */
    CmxDateTimeMomentPipe.prototype.formatterService;
}
