// @cemex-core/angular-localization-v7/esm5/lib/pipes/cmx-date-moment.pipe.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Pipe } from '@angular/core';
import { FormatterService, DateTimeActions } from './../services/formatter.service';
import { TranslationService } from './../services/translation.service';
var CmxDateMomentPipe = /** @class */ (function () {
    function CmxDateMomentPipe(translationService, formatterService) {
        this.translationService = translationService;
        this.formatterService = formatterService;
    }
    /**
     * @param {?} dateTime
     * @param {?=} action
     * @return {?}
     */
    CmxDateMomentPipe.prototype.transform = /**
     * @param {?} dateTime
     * @param {?=} action
     * @return {?}
     */
    function (dateTime, action) {
        if (action === void 0) { action = DateTimeActions.ConvertToLocal; }
        return this.formatterService.formatDateTimeMoment(dateTime, this.translationService.selectedLanguage.languageISO.replace('_', '-'), this.translationService.selectedLanguage.formatDate, action);
    };
    CmxDateMomentPipe.decorators = [
        { type: Pipe, args: [{
                    name: 'cmxDateMoment',
                    pure: false
                },] }
    ];
    /** @nocollapse */
    CmxDateMomentPipe.ctorParameters = function () { return [
        { type: TranslationService },
        { type: FormatterService }
    ]; };
    return CmxDateMomentPipe;
}());
export { CmxDateMomentPipe };
if (false) {
    /**
     * @type {?}
     * @private
     */
    CmxDateMomentPipe.prototype.translationService;
    /**
     * @type {?}
     * @private
     */
    CmxDateMomentPipe.prototype.formatterService;
}
