// @cemex-core/angular-localization-v7/esm5/lib/pipes/cmx-currency.pipe.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { isPlatformBrowser } from '@angular/common';
import { Inject, Pipe, PLATFORM_ID } from '@angular/core';
import { FormatterService } from './../services/formatter.service';
import { TranslationService } from './../services/translation.service';
var CmxCurrencyPipe = /** @class */ (function () {
    function CmxCurrencyPipe(formatter, translation, platformId) {
        this.formatter = formatter;
        this.translation = translation;
        this.platformId = platformId;
        if (isPlatformBrowser(platformId)) {
            this.window = window;
        }
    }
    /**
     * @param {?} value
     * @param {?=} countryCode
     * @return {?}
     */
    CmxCurrencyPipe.prototype.transform = /**
     * @param {?} value
     * @param {?=} countryCode
     * @return {?}
     */
    function (value, countryCode) {
        if (value === '' || value == null || isNaN((/** @type {?} */ (value)))) {
            console.warn('The value is not a number');
            return '';
        }
        countryCode = countryCode || this.window.sessionStorage.getItem('country');
        // tslint:disable-next-line:triple-equals
        if (countryCode == undefined) {
            console.warn("Can't retrieve countryCode");
            return '';
        }
        /** @type {?} */
        var lang = this.translation.getLanguageByCountryCode(countryCode);
        /** @type {?} */
        var symbolPosition = (lang.currencySymbolFloat && lang.currencySymbolFloat.toLowerCase()) ||
            'left';
        /** @type {?} */
        var currencySymbol = lang.currencySymbol || '$';
        /** @type {?} */
        var formattedNumber = this.formatter.formatNumber(value, lang && lang.decimalNumbersMoney, lang && lang.thousandSeparator, lang && lang.decimalSeparator);
        /** @type {?} */
        var numberWithCurrency = symbolPosition === 'left'
            ? "" + currencySymbol + formattedNumber
            : "" + formattedNumber + currencySymbol;
        return '&lrm;' + numberWithCurrency;
    };
    CmxCurrencyPipe.decorators = [
        { type: Pipe, args: [{
                    name: 'cmxCurrency',
                    pure: true
                },] }
    ];
    /** @nocollapse */
    CmxCurrencyPipe.ctorParameters = function () { return [
        { type: FormatterService },
        { type: TranslationService },
        { type: undefined, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
    ]; };
    return CmxCurrencyPipe;
}());
export { CmxCurrencyPipe };
if (false) {
    /** @type {?} */
    CmxCurrencyPipe.prototype.window;
    /**
     * @type {?}
     * @private
     */
    CmxCurrencyPipe.prototype.formatter;
    /**
     * @type {?}
     * @private
     */
    CmxCurrencyPipe.prototype.translation;
    /**
     * @type {?}
     * @private
     */
    CmxCurrencyPipe.prototype.platformId;
}
