// @cemex-core/angular-services-v7/esm5/lib/country-config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Response } from '@angular/http';
import { BehaviorSubject, Observable } from 'rxjs';
import { map, catchError } from 'rxjs/operators';
import { HttpCemex } from './http.service';
import { ProjectSettings } from './project.settings';
import { SessionService } from './session.service';
var CountryConfigService = /** @class */ (function () {
    function CountryConfigService(http, projectEnv, sessionService) {
        var _this = this;
        this.http = http;
        this.projectEnv = projectEnv;
        this.sessionService = sessionService;
        this._countryConfig = new BehaviorSubject(null);
        this._countryConfigApiPath = '/v5/secm/console/countryconfig';
        if (!this._countryConfig.getValue()) {
            this.getCountryConfiguration().subscribe((/**
             * @param {?} data
             * @return {?}
             */
            function (data) {
                _this._countryConfig.next(data);
            }));
        }
    }
    Object.defineProperty(CountryConfigService.prototype, "countryConfig", {
        get: /**
         * @return {?}
         */
        function () {
            return this._countryConfig.asObservable();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    CountryConfigService.prototype.getCountryConfiguration = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var userCountry = this.sessionService.userCountry || 'US';
        /** @type {?} */
        var url = "" + this.projectEnv.getBaseOrgEnvPath() + this._countryConfigApiPath + "?countryCode=" + userCountry;
        return this.http.get(url).pipe(map((/**
         * @param {?} res
         * @return {?}
         */
        function (res) { return res.json(); })), catchError(this.handleError));
    };
    /**
     * @private
     * @param {?} error
     * @return {?}
     */
    CountryConfigService.prototype.handleError = /**
     * @private
     * @param {?} error
     * @return {?}
     */
    function (error) {
        // In a real world app, we might use a remote logging infrastructure
        /** @type {?} */
        var errMsg;
        if (error instanceof Response) {
            /** @type {?} */
            var body = error.json() || '';
            /** @type {?} */
            var err = body.error || JSON.stringify(body);
            errMsg = error.status + " - " + (error.statusText || '');
        }
        else {
            errMsg = error.message ? error.message : error.toString();
        }
        console.error(error);
        return Observable.throw(error);
    };
    CountryConfigService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    CountryConfigService.ctorParameters = function () { return [
        { type: HttpCemex },
        { type: ProjectSettings },
        { type: SessionService }
    ]; };
    return CountryConfigService;
}());
export { CountryConfigService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    CountryConfigService.prototype._countryConfig;
    /**
     * @type {?}
     * @private
     */
    CountryConfigService.prototype._countryConfigApiPath;
    /**
     * @type {?}
     * @private
     */
    CountryConfigService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    CountryConfigService.prototype.projectEnv;
    /**
     * @type {?}
     * @private
     */
    CountryConfigService.prototype.sessionService;
}
