// @cemex-core/angular-services-v7/esm5/lib/financialDocuments.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { map } from 'rxjs/operators';
import { EndpointsConstants } from '@cemex-core/constants-v2/dist';
import { HttpCemex } from './http.service';
var FinancialDocuments = /** @class */ (function () {
    function FinancialDocuments(http, endpoints) {
        this.http = http;
        this.endpoints = endpoints;
    }
    /**
     * @return {?}
     */
    FinancialDocuments.prototype.getLegalEntity = /**
     * @return {?}
     */
    function () {
        return this.http
            .get(this.getUrl('legalentity'))
            .pipe(map((/**
         * @param {?} response
         * @return {?}
         */
        function (response) {
            /** @type {?} */
            var aux = response.json();
            return aux;
        })))
            .toPromise()
            .then((/**
         * @param {?} data
         * @return {?}
         */
        function (data) { return data; }))
            .catch((/**
         * @param {?} error
         * @return {?}
         */
        function (error) { return error; }));
    };
    /**
     * @return {?}
     */
    FinancialDocuments.prototype.getAllLegalEntityCustomers = /**
     * @return {?}
     */
    function () {
        return this.http
            .get(this.getUrl('legalentitycustomers'))
            .pipe(map((/**
         * @param {?} response
         * @return {?}
         */
        function (response) {
            /** @type {?} */
            var aux = response.json();
            /** @type {?} */
            var snapshot = aux.responseLegalEntity.customer;
            return snapshot;
        })))
            .toPromise()
            .then((/**
         * @param {?} data
         * @return {?}
         */
        function (data) { return data; }))
            .catch((/**
         * @param {?} error
         * @return {?}
         */
        function (error) { return error; }));
    };
    /**
     * @param {?} customerCode
     * @return {?}
     */
    FinancialDocuments.prototype.getLegalEntityByCustomer = /**
     * @param {?} customerCode
     * @return {?}
     */
    function (customerCode) {
        return this.http
            .get(this.getUrl('legalentity?customerCode=' + customerCode))
            .pipe(map((/**
         * @param {?} response
         * @return {?}
         */
        function (response) {
            /** @type {?} */
            var aux = response.json();
            /** @type {?} */
            var snapshot = aux.responseLegalEntity.company;
            return snapshot;
        })))
            .toPromise()
            .then((/**
         * @param {?} data
         * @return {?}
         */
        function (data) { return data; }))
            .catch((/**
         * @param {?} error
         * @return {?}
         */
        function (error) { return error; }));
    };
    /**
     * @param {?} companyCode
     * @return {?}
     */
    FinancialDocuments.prototype.getLegalEntityByCompany = /**
     * @param {?} companyCode
     * @return {?}
     */
    function (companyCode) {
        return this.http
            .get(this.getUrl('legalentity?companyCode=' + companyCode))
            .pipe(map((/**
         * @param {?} response
         * @return {?}
         */
        function (response) {
            /** @type {?} */
            var aux = response.json();
            /** @type {?} */
            var snapshot = aux.responseLegalEntity.customer;
            return snapshot;
        })))
            .toPromise()
            .then((/**
         * @param {?} data
         * @return {?}
         */
        function (data) { return data; }))
            .catch((/**
         * @param {?} error
         * @return {?}
         */
        function (error) { return error; }));
    };
    /**
     * @param {?} customerCode
     * @param {?} companyCode
     * @return {?}
     */
    FinancialDocuments.prototype.getLegalEntityJobsites = /**
     * @param {?} customerCode
     * @param {?} companyCode
     * @return {?}
     */
    function (customerCode, companyCode) {
        /** @type {?} */
        var api = 'legalentityjobsites?companyCode=' +
            companyCode +
            '&customerCode=' +
            customerCode;
        return this.http
            .get(this.getUrl(api))
            .pipe(map((/**
         * @param {?} response
         * @return {?}
         */
        function (response) {
            /** @type {?} */
            var aux = response.json();
            /** @type {?} */
            var snapshot = aux.responseLegalEntity.jobSite;
            return snapshot;
        })))
            .toPromise()
            .then((/**
         * @param {?} data
         * @return {?}
         */
        function (data) { return data; }))
            .catch((/**
         * @param {?} error
         * @return {?}
         */
        function (error) { return error; }));
    };
    /**
     * @private
     * @param {?} api
     * @return {?}
     */
    FinancialDocuments.prototype.getUrl = /**
     * @private
     * @param {?} api
     * @return {?}
     */
    function (api) {
        /** @type {?} */
        var url = this.http.generateEndpoint(this.endpoints.V1 + this.endpoints.armCapability + api);
        return url;
    };
    FinancialDocuments.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    FinancialDocuments.ctorParameters = function () { return [
        { type: HttpCemex },
        { type: EndpointsConstants }
    ]; };
    return FinancialDocuments;
}());
export { FinancialDocuments };
if (false) {
    /**
     * @type {?}
     * @private
     */
    FinancialDocuments.prototype.http;
    /**
     * @type {?}
     * @private
     */
    FinancialDocuments.prototype.endpoints;
}
