// @cemex-core/angular-services-v7/esm5/lib/project.settings.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
var ProjectSettings = /** @class */ (function () {
    function ProjectSettings() {
        /** @type {?} */
        var APP_KEYS = {
            API_ENV_KEY: 'API_ENV',
            API_HOST_KEY: 'API_HOST',
            API_ORG_KEY: 'API_ORG',
            APP_CODE_KEY: 'APP_CODE',
            CLIENT_ID_KEY: 'CLIENT_ID'
        };
        this._appCode = ((/** @type {?} */ (global)))[APP_KEYS.APP_CODE_KEY];
        this._clientId = ((/** @type {?} */ (global)))[APP_KEYS.CLIENT_ID_KEY];
        this._apiBasePath = ((/** @type {?} */ (global)))[APP_KEYS.API_HOST_KEY];
        this._apiOrg = ((/** @type {?} */ (global)))[APP_KEYS.API_ORG_KEY];
        this._apiEnv = ((/** @type {?} */ (global)))[APP_KEYS.API_ENV_KEY];
        if (this._apiBasePath === undefined) {
            console.warn('environment variable \'API_HOST\' is missing !!');
        }
    }
    Object.defineProperty(ProjectSettings.prototype, "appCode", {
        // ============================================================================================
        // Getters
        // ============================================================================================
        get: 
        // ============================================================================================
        // Getters
        // ============================================================================================
        /**
         * @return {?}
         */
        function () {
            return this._appCode;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProjectSettings.prototype, "clientId", {
        get: /**
         * @return {?}
         */
        function () {
            return this._clientId;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProjectSettings.prototype, "apiBasePath", {
        get: /**
         * @return {?}
         */
        function () {
            return this._apiBasePath;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProjectSettings.prototype, "apiOrg", {
        get: /**
         * @return {?}
         */
        function () {
            return this._apiOrg;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProjectSettings.prototype, "apiEnv", {
        get: /**
         * @return {?}
         */
        function () {
            return this._apiEnv;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} apiEndpoint
     * @return {?}
     */
    ProjectSettings.prototype.generateEndpoint = /**
     * @param {?} apiEndpoint
     * @return {?}
     */
    function (apiEndpoint) {
        /** @type {?} */
        var url = "" + this._apiBasePath + this._apiOrg + this._apiEnv + apiEndpoint;
        return url;
    };
    /**
     * @return {?}
     */
    ProjectSettings.prototype.getBaseOrgEnvPath = /**
     * @return {?}
     */
    function () {
        return this._apiBasePath + this._apiOrg + this._apiEnv;
    };
    /**
     * @param {?} newKeys
     * @return {?}
     */
    ProjectSettings.prototype.setAppKeys = /**
     * @param {?} newKeys
     * @return {?}
     */
    function (newKeys) {
        this._appCode = newKeys.APP_CODE;
        this._clientId = newKeys.CLIENT_ID;
        if (this._apiBasePath === undefined) {
            console.warn('environment variable \'API_HOST\' is missing !!');
        }
    };
    ProjectSettings.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    ProjectSettings.ctorParameters = function () { return []; };
    return ProjectSettings;
}());
export { ProjectSettings };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ProjectSettings.prototype._appCode;
    /**
     * @type {?}
     * @private
     */
    ProjectSettings.prototype._clientId;
    /**
     * @type {?}
     * @private
     */
    ProjectSettings.prototype._apiBasePath;
    /**
     * @type {?}
     * @private
     */
    ProjectSettings.prototype._apiOrg;
    /**
     * @type {?}
     * @private
     */
    ProjectSettings.prototype._apiEnv;
}
