// @cemex/cmx-alert-v7/esm2015/lib/services/alert.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { AlertType } from '../models/alert';
import { AlertRef } from './alert-ref';
import { BehaviorSubject } from 'rxjs';
export class AlertService {
    constructor() {
        this.alerts$ = new BehaviorSubject([]);
        this.alerts = [];
    }
    /**
     * @param {?} alertMessage
     * @param {?} actionMessage
     * @param {?=} duration
     * @param {?=} alertType
     * @return {?}
     */
    openSuccess(alertMessage, actionMessage, duration = 5000, alertType = AlertType.Success) {
        return this.open(alertMessage, actionMessage, duration, alertType);
    }
    /**
     * @param {?} alertMessage
     * @param {?} actionMessage
     * @param {?=} duration
     * @param {?=} alertType
     * @return {?}
     */
    openError(alertMessage, actionMessage, duration = 5000, alertType = AlertType.Error) {
        return this.open(alertMessage, actionMessage, duration, alertType);
    }
    /**
     * @param {?} alertMessage
     * @param {?} actionMessage
     * @param {?=} duration
     * @param {?=} alertType
     * @return {?}
     */
    openWarning(alertMessage, actionMessage, duration = 5000, alertType = AlertType.Warning) {
        return this.open(alertMessage, actionMessage, duration, alertType);
    }
    /**
     * @param {?} alertMessage
     * @param {?} actionMessage
     * @param {?=} duration
     * @param {?=} alertType
     * @return {?}
     */
    openInfo(alertMessage, actionMessage, duration = 5000, alertType = AlertType.Info) {
        return this.open(alertMessage, actionMessage, duration, alertType);
    }
    /**
     * @private
     * @param {?} alertMessage
     * @param {?} actionMessage
     * @param {?} duration
     * @param {?} alertType
     * @return {?}
     */
    open(alertMessage, actionMessage, duration, alertType) {
        /** @type {?} */
        const alertRef = new AlertRef(alertMessage, actionMessage, duration, alertType);
        this.alerts.push(alertRef);
        this.alerts$.next(this.alerts);
        if (alertRef.duration > 0) {
            this._setTimeout(alertRef);
        }
        return alertRef;
    }
    /**
     * @private
     * @param {?} alert
     * @return {?}
     */
    _setTimeout(alert) {
        window.setTimeout((/**
         * @return {?}
         */
        () => {
            this.removeAlert(alert);
        }), alert.duration);
    }
    /**
     * @param {?} alert
     * @return {?}
     */
    removeAlert(alert) {
        this.alerts = this.alerts.filter((/**
         * @param {?} x
         * @return {?}
         */
        x => x !== alert));
        this.alerts$.next(this.alerts);
    }
}
AlertService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
AlertService.ctorParameters = () => [];
if (false) {
    /** @type {?} */
    AlertService.prototype.alerts$;
    /** @type {?} */
    AlertService.prototype.alerts;
}
