// @cemex/cmx-alert-v7/esm2015/lib/services/alert-ref.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Subject } from 'rxjs';
export class AlertRef {
    /**
     * @param {?} alertMessage
     * @param {?} actionMessage
     * @param {?} duration
     * @param {?} alertType
     */
    constructor(alertMessage, actionMessage, duration, alertType) {
        this.alertMessage = alertMessage;
        this.actionMessage = actionMessage;
        this.duration = duration;
        this.alertType = alertType;
        this._onAction = new Subject();
        this._onClose = new Subject();
    }
    /**
     * @return {?}
     */
    onAction() {
        this._onAction.next();
    }
    /**
     * @return {?}
     */
    onActionClicked() {
        return this._onAction.asObservable();
    }
    /**
     * @return {?}
     */
    onClose() {
        this._onClose.next();
    }
    /**
     * @return {?}
     */
    onCloseClicked() {
        return this._onClose.asObservable();
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    AlertRef.prototype._onAction;
    /**
     * @type {?}
     * @private
     */
    AlertRef.prototype._onClose;
    /** @type {?} */
    AlertRef.prototype.alertMessage;
    /** @type {?} */
    AlertRef.prototype.actionMessage;
    /** @type {?} */
    AlertRef.prototype.duration;
    /** @type {?} */
    AlertRef.prototype.alertType;
}
