// @cemex-core/angular-services-v7/esm5/lib/validators/number.validators.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var NumberValidators = /** @class */ (function () {
    function NumberValidators() {
    }
    /**
     * @param {?} c
     * @return {?}
     */
    NumberValidators.isNumeric = /**
     * @param {?} c
     * @return {?}
     */
    function (c) {
        if (!isNaN(parseFloat(c.value)) && isFinite(c.value)) {
            return null;
        }
        else {
            return {
                notNumeric: true
            };
        }
    };
    /**
     * @param {?} c
     * @return {?}
     */
    NumberValidators.onlyTwoDecimals = /**
     * @param {?} c
     * @return {?}
     */
    function (c) {
        /** @type {?} */
        var regexp = /^\d+(\.\d{1,2})?$/;
        if (regexp.test(c.value)) {
            return null;
        }
        else {
            return {
                patternInvalid: true
            };
        }
    };
    /**
     * @param {?} c
     * @return {?}
     */
    NumberValidators.greaterThanZero = /**
     * @param {?} c
     * @return {?}
     */
    function (c) {
        if (parseFloat(c.value) > 0) {
            return null;
        }
        else {
            return {
                lessThanZero: true
            };
        }
    };
    /**
     * @param {?} c
     * @return {?}
     */
    NumberValidators.isPositive = /**
     * @param {?} c
     * @return {?}
     */
    function (c) {
        if (parseFloat(c.value) >= 0) {
            return null;
        }
        else {
            return {
                isNegative: true
            };
        }
    };
    /**
     * @param {?} max
     * @return {?}
     */
    NumberValidators.lessThan = /**
     * @param {?} max
     * @return {?}
     */
    function (max) {
        /** @type {?} */
        var maxNumber = max;
        return (/**
         * @param {?} control
         * @return {?}
         */
        function (control) {
            /** @type {?} */
            var num = +control.value;
            if (isNaN(num) || num > maxNumber) {
                return {
                    isGreater: true
                };
            }
            return null;
        });
    };
    /**
     * @param {?} min
     * @return {?}
     */
    NumberValidators.greaterThan = /**
     * @param {?} min
     * @return {?}
     */
    function (min) {
        /** @type {?} */
        var minNumber = min;
        return (/**
         * @param {?} control
         * @return {?}
         */
        function (control) {
            /** @type {?} */
            var num = +control.value;
            if (isNaN(num) || num < minNumber) {
                return {
                    isLess: true
                };
            }
            return null;
        });
    };
    return NumberValidators;
}());
export { NumberValidators };
