// @cemex-core/angular-services-v7/esm2015/lib/validators/number.validators.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class NumberValidators {
    /**
     * @param {?} c
     * @return {?}
     */
    static isNumeric(c) {
        if (!isNaN(parseFloat(c.value)) && isFinite(c.value)) {
            return null;
        }
        else {
            return {
                notNumeric: true
            };
        }
    }
    /**
     * @param {?} c
     * @return {?}
     */
    static onlyTwoDecimals(c) {
        /** @type {?} */
        const regexp = /^\d+(\.\d{1,2})?$/;
        if (regexp.test(c.value)) {
            return null;
        }
        else {
            return {
                patternInvalid: true
            };
        }
    }
    /**
     * @param {?} c
     * @return {?}
     */
    static greaterThanZero(c) {
        if (parseFloat(c.value) > 0) {
            return null;
        }
        else {
            return {
                lessThanZero: true
            };
        }
    }
    /**
     * @param {?} c
     * @return {?}
     */
    static isPositive(c) {
        if (parseFloat(c.value) >= 0) {
            return null;
        }
        else {
            return {
                isNegative: true
            };
        }
    }
    /**
     * @param {?} max
     * @return {?}
     */
    static lessThan(max) {
        /** @type {?} */
        const maxNumber = max;
        return (/**
         * @param {?} control
         * @return {?}
         */
        (control) => {
            /** @type {?} */
            const num = +control.value;
            if (isNaN(num) || num > maxNumber) {
                return {
                    isGreater: true
                };
            }
            return null;
        });
    }
    /**
     * @param {?} min
     * @return {?}
     */
    static greaterThan(min) {
        /** @type {?} */
        const minNumber = min;
        return (/**
         * @param {?} control
         * @return {?}
         */
        (control) => {
            /** @type {?} */
            const num = +control.value;
            if (isNaN(num) || num < minNumber) {
                return {
                    isLess: true
                };
            }
            return null;
        });
    }
}
