// @cemex-core/angular-localization-v7/esm2015/lib/services/date-formatter.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
export class DateFormatterService {
    constructor() {
        //
    }
    /**
     * \@description Formats a date object to a string given a date format from a language
     * @param {?} $value
     * @param {?} $format
     * @param {?} $daysNames
     * @param {?} $monthsNames
     * @return {?} String of the formatted date
     */
    format($value, $format, $daysNames, $monthsNames) {
        /** @type {?} */
        let dateAsString = '';
        /** @type {?} */
        let $separator = '/';
        /** @type {?} */
        let day;
        /** @type {?} */
        let month;
        /** @type {?} */
        let year;
        /** @type {?} */
        let $formats = $format.split($separator);
        if ($formats.length === 1) {
            $separator = '-';
            $formats = $format.split($separator);
        }
        if ($formats.length === 1) {
            $separator = ' ';
            $formats = $format.split($separator);
        }
        if ($formats.length === 1) {
            $separator = '.';
            $formats = $format.split($separator);
        }
        $formats.forEach((/**
         * @param {?} $formatItem
         * @param {?} $index
         * @return {?}
         */
        ($formatItem, $index) => {
            /** @type {?} */
            const $auxValue = $value;
            /** @type {?} */
            const $formatCode = $formatItem.replace(',', '');
            if ($formatItem.includes('d') || $formatItem.includes('D')) {
                day = this.getDay($auxValue, $formatCode, $daysNames);
                $formatItem = $formatItem.replace($formatCode, day);
            }
            else if ($formatItem.includes('m') || $formatItem.includes('M')) {
                month = this.getMonth($auxValue, $formatCode, $monthsNames);
                $formatItem = $formatItem.replace($formatCode, month);
            }
            else if ($formatItem.includes('y') || $formatItem.includes('Y')) {
                year = this.getYear($auxValue, $formatCode);
                $formatItem = $formatItem.replace($formatCode, year);
            }
            if ($index !== $formats.length - 1) {
                dateAsString += $formatItem + $separator;
            }
            else {
                dateAsString += $formatItem;
            }
        }));
        return dateAsString;
    }
    /**
     * \@description Transforms a valid string to a date object
     * @param {?} date The date as string
     * @return {?} A javascript date object
     */
    getDateObject(date) {
        try {
            /** @type {?} */
            const auxDate = date.replace(/-g/, '/');
            /** @type {?} */
            const bits = auxDate.split(/\D/);
            /** @type {?} */
            const month = parseInt(bits[1], undefined) - 1;
            if (bits.length === 3) {
                return new Date(bits[0], month, bits[2]);
            }
            else {
                return new Date(bits[0], month, bits[2], bits[3], bits[4], bits[5]);
            }
        }
        catch ($exception) {
            return undefined;
        }
    }
    /**
     * @private
     * @param {?} $number
     * @return {?}
     */
    addZero($number) {
        return $number < 10 ? '0' + $number : $number.toString();
    }
    /**
     * @private
     * @param {?} $date
     * @param {?} $format
     * @param {?} $daysNames
     * @return {?}
     */
    getDay($date, $format, $daysNames) {
        /** @type {?} */
        let day;
        if ($format.length === 4) {
            // requires the complete name of the day
            day = $daysNames[$date.getDay()];
        }
        else if ($format.length === 3) {
            // requires an abbreviation of the day's name
            day = $daysNames[$date.getDay()].substring(0, 3);
        }
        else if ($format.length === 2 || $format.length === 1) {
            // requires the value of the day
            day = this.addZero($date.getDate());
        }
        // check if names should use lower or uppercase
        if ($format === $format.toLowerCase()) {
            // the format uses d
            day = day.toLowerCase();
        }
        return day;
    }
    /**
     * @private
     * @param {?} $date
     * @param {?} $format
     * @param {?} $monthsNames
     * @return {?}
     */
    getMonth($date, $format, $monthsNames) {
        /** @type {?} */
        let month;
        if ($format.length === 4) {
            // requires the complete name of the month
            month = $monthsNames[$date.getMonth()];
        }
        else if ($format.length === 3) {
            // requires an abbreviation of the day's name
            month = $monthsNames[$date.getMonth()].substring(0, 3);
        }
        else if ($format.length === 2) {
            // requires the value of the day
            month = this.addZero($date.getMonth() + 1);
        }
        // check if names should use lower or uppercase
        if ($format === $format.toLowerCase()) {
            // the format uses d
            month = month.toLowerCase();
        }
        return month;
    }
    /**
     * @private
     * @param {?} $date
     * @param {?} $format
     * @return {?}
     */
    getYear($date, $format) {
        /** @type {?} */
        let year;
        if ($format.length === 4) {
            // requires the complete year
            year = '' + $date.getFullYear();
        }
        else {
            // requires only the last 2 numbers of the year value
            year = '' + $date.getFullYear();
            year = year.substring(2);
        }
        return year;
    }
}
DateFormatterService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
DateFormatterService.ctorParameters = () => [];
