// @cemex-core/angular-localization-v7/esm2015/lib/services/direction.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import { TranslationService } from './translation.service';
export class DirectionService {
    /**
     * @param {?} translation
     */
    constructor(translation) {
        this.translation = translation;
        this.direction$ = new Subject();
        this.rtl$ = new Subject();
        this.translation.getSelectedLanguage().subscribe((/**
         * @param {?} language
         * @return {?}
         */
        (language) => {
            this.direction$.next(language.direction);
            this.rtl$.next(language.rtl);
        }));
    }
    /**
     * @return {?}
     */
    getDirection() {
        return this.direction$.asObservable();
    }
    /**
     * @return {?}
     */
    getRTL() {
        return this.rtl$.asObservable();
    }
    /**
     * @param {?} languageCode
     * @return {?}
     */
    getDirectionFor(languageCode) {
        /** @type {?} */
        const language = this.translation.getLanguage(languageCode.split(/[-_]/)[0]);
        this.direction$.next(language.dir);
    }
    /**
     * @param {?} languageCode
     * @return {?}
     */
    getRtlFor(languageCode) {
        /** @type {?} */
        const language = this.translation.getLanguage(languageCode.split(/[-_]/)[0]);
        this.rtl$.next(language.rtl);
    }
}
DirectionService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
DirectionService.ctorParameters = () => [
    { type: TranslationService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    DirectionService.prototype.direction$;
    /**
     * @type {?}
     * @private
     */
    DirectionService.prototype.rtl$;
    /**
     * @type {?}
     * @private
     */
    DirectionService.prototype.translation;
}
