// @cemex-core/angular-services-v7/esm2015/lib/country-config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Response } from '@angular/http';
import { BehaviorSubject, Observable } from 'rxjs';
import { map, catchError } from 'rxjs/operators';
import { HttpCemex } from './http.service';
import { ProjectSettings } from './project.settings';
import { SessionService } from './session.service';
export class CountryConfigService {
    /**
     * @param {?} http
     * @param {?} projectEnv
     * @param {?} sessionService
     */
    constructor(http, projectEnv, sessionService) {
        this.http = http;
        this.projectEnv = projectEnv;
        this.sessionService = sessionService;
        this._countryConfig = new BehaviorSubject(null);
        this._countryConfigApiPath = '/v5/secm/console/countryconfig';
        if (!this._countryConfig.getValue()) {
            this.getCountryConfiguration().subscribe((/**
             * @param {?} data
             * @return {?}
             */
            (data) => {
                this._countryConfig.next(data);
            }));
        }
    }
    /**
     * @return {?}
     */
    get countryConfig() {
        return this._countryConfig.asObservable();
    }
    /**
     * @return {?}
     */
    getCountryConfiguration() {
        /** @type {?} */
        const userCountry = this.sessionService.userCountry || 'US';
        /** @type {?} */
        const url = `${this.projectEnv.getBaseOrgEnvPath()}${this._countryConfigApiPath}?countryCode=${userCountry}`;
        return this.http.get(url).pipe(map((/**
         * @param {?} res
         * @return {?}
         */
        (res) => res.json())), catchError(this.handleError));
    }
    /**
     * @private
     * @param {?} error
     * @return {?}
     */
    handleError(error) {
        // In a real world app, we might use a remote logging infrastructure
        /** @type {?} */
        let errMsg;
        if (error instanceof Response) {
            /** @type {?} */
            const body = error.json() || '';
            /** @type {?} */
            const err = body.error || JSON.stringify(body);
            errMsg = `${error.status} - ${error.statusText || ''}`;
        }
        else {
            errMsg = error.message ? error.message : error.toString();
        }
        console.error(error);
        return Observable.throw(error);
    }
}
CountryConfigService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
CountryConfigService.ctorParameters = () => [
    { type: HttpCemex },
    { type: ProjectSettings },
    { type: SessionService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    CountryConfigService.prototype._countryConfig;
    /**
     * @type {?}
     * @private
     */
    CountryConfigService.prototype._countryConfigApiPath;
    /**
     * @type {?}
     * @private
     */
    CountryConfigService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    CountryConfigService.prototype.projectEnv;
    /**
     * @type {?}
     * @private
     */
    CountryConfigService.prototype.sessionService;
}
