// @cemex-core/angular-services-v7/esm2015/lib/financialDocuments.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { map } from 'rxjs/operators';
import { EndpointsConstants } from '@cemex-core/constants-v2/dist';
import { HttpCemex } from './http.service';
export class FinancialDocuments {
    /**
     * @param {?} http
     * @param {?} endpoints
     */
    constructor(http, endpoints) {
        this.http = http;
        this.endpoints = endpoints;
    }
    /**
     * @return {?}
     */
    getLegalEntity() {
        return this.http
            .get(this.getUrl('legalentity'))
            .pipe(map((/**
         * @param {?} response
         * @return {?}
         */
        response => {
            /** @type {?} */
            const aux = response.json();
            return aux;
        })))
            .toPromise()
            .then((/**
         * @param {?} data
         * @return {?}
         */
        data => data))
            .catch((/**
         * @param {?} error
         * @return {?}
         */
        error => error));
    }
    /**
     * @return {?}
     */
    getAllLegalEntityCustomers() {
        return this.http
            .get(this.getUrl('legalentitycustomers'))
            .pipe(map((/**
         * @param {?} response
         * @return {?}
         */
        response => {
            /** @type {?} */
            const aux = response.json();
            /** @type {?} */
            const snapshot = aux.responseLegalEntity.customer;
            return snapshot;
        })))
            .toPromise()
            .then((/**
         * @param {?} data
         * @return {?}
         */
        data => data))
            .catch((/**
         * @param {?} error
         * @return {?}
         */
        error => error));
    }
    /**
     * @param {?} customerCode
     * @return {?}
     */
    getLegalEntityByCustomer(customerCode) {
        return this.http
            .get(this.getUrl('legalentity?customerCode=' + customerCode))
            .pipe(map((/**
         * @param {?} response
         * @return {?}
         */
        response => {
            /** @type {?} */
            const aux = response.json();
            /** @type {?} */
            const snapshot = aux.responseLegalEntity.company;
            return snapshot;
        })))
            .toPromise()
            .then((/**
         * @param {?} data
         * @return {?}
         */
        data => data))
            .catch((/**
         * @param {?} error
         * @return {?}
         */
        error => error));
    }
    /**
     * @param {?} companyCode
     * @return {?}
     */
    getLegalEntityByCompany(companyCode) {
        return this.http
            .get(this.getUrl('legalentity?companyCode=' + companyCode))
            .pipe(map((/**
         * @param {?} response
         * @return {?}
         */
        response => {
            /** @type {?} */
            const aux = response.json();
            /** @type {?} */
            const snapshot = aux.responseLegalEntity.customer;
            return snapshot;
        })))
            .toPromise()
            .then((/**
         * @param {?} data
         * @return {?}
         */
        data => data))
            .catch((/**
         * @param {?} error
         * @return {?}
         */
        error => error));
    }
    /**
     * @param {?} customerCode
     * @param {?} companyCode
     * @return {?}
     */
    getLegalEntityJobsites(customerCode, companyCode) {
        /** @type {?} */
        const api = 'legalentityjobsites?companyCode=' +
            companyCode +
            '&customerCode=' +
            customerCode;
        return this.http
            .get(this.getUrl(api))
            .pipe(map((/**
         * @param {?} response
         * @return {?}
         */
        response => {
            /** @type {?} */
            const aux = response.json();
            /** @type {?} */
            const snapshot = aux.responseLegalEntity.jobSite;
            return snapshot;
        })))
            .toPromise()
            .then((/**
         * @param {?} data
         * @return {?}
         */
        data => data))
            .catch((/**
         * @param {?} error
         * @return {?}
         */
        error => error));
    }
    /**
     * @private
     * @param {?} api
     * @return {?}
     */
    getUrl(api) {
        /** @type {?} */
        const url = this.http.generateEndpoint(this.endpoints.V1 + this.endpoints.armCapability + api);
        return url;
    }
}
FinancialDocuments.decorators = [
    { type: Injectable }
];
/** @nocollapse */
FinancialDocuments.ctorParameters = () => [
    { type: HttpCemex },
    { type: EndpointsConstants }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    FinancialDocuments.prototype.http;
    /**
     * @type {?}
     * @private
     */
    FinancialDocuments.prototype.endpoints;
}
