// @cemex-core/angular-services-v7/esm2015/lib/project.settings.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
export class ProjectSettings {
    constructor() {
        /** @type {?} */
        const APP_KEYS = {
            API_ENV_KEY: 'API_ENV',
            API_HOST_KEY: 'API_HOST',
            API_ORG_KEY: 'API_ORG',
            APP_CODE_KEY: 'APP_CODE',
            CLIENT_ID_KEY: 'CLIENT_ID'
        };
        this._appCode = ((/** @type {?} */ (global)))[APP_KEYS.APP_CODE_KEY];
        this._clientId = ((/** @type {?} */ (global)))[APP_KEYS.CLIENT_ID_KEY];
        this._apiBasePath = ((/** @type {?} */ (global)))[APP_KEYS.API_HOST_KEY];
        this._apiOrg = ((/** @type {?} */ (global)))[APP_KEYS.API_ORG_KEY];
        this._apiEnv = ((/** @type {?} */ (global)))[APP_KEYS.API_ENV_KEY];
        if (this._apiBasePath === undefined) {
            console.warn('environment variable \'API_HOST\' is missing !!');
        }
    }
    // ============================================================================================
    // Getters
    // ============================================================================================
    /**
     * @return {?}
     */
    get appCode() {
        return this._appCode;
    }
    /**
     * @return {?}
     */
    get clientId() {
        return this._clientId;
    }
    /**
     * @return {?}
     */
    get apiBasePath() {
        return this._apiBasePath;
    }
    /**
     * @return {?}
     */
    get apiOrg() {
        return this._apiOrg;
    }
    /**
     * @return {?}
     */
    get apiEnv() {
        return this._apiEnv;
    }
    /**
     * @param {?} apiEndpoint
     * @return {?}
     */
    generateEndpoint(apiEndpoint) {
        /** @type {?} */
        const url = `${this._apiBasePath}${this._apiOrg}${this._apiEnv}${apiEndpoint}`;
        return url;
    }
    /**
     * @return {?}
     */
    getBaseOrgEnvPath() {
        return this._apiBasePath + this._apiOrg + this._apiEnv;
    }
    /**
     * @param {?} newKeys
     * @return {?}
     */
    setAppKeys(newKeys) {
        this._appCode = newKeys.APP_CODE;
        this._clientId = newKeys.CLIENT_ID;
        if (this._apiBasePath === undefined) {
            console.warn('environment variable \'API_HOST\' is missing !!');
        }
    }
}
ProjectSettings.decorators = [
    { type: Injectable }
];
/** @nocollapse */
ProjectSettings.ctorParameters = () => [];
if (false) {
    /**
     * @type {?}
     * @private
     */
    ProjectSettings.prototype._appCode;
    /**
     * @type {?}
     * @private
     */
    ProjectSettings.prototype._clientId;
    /**
     * @type {?}
     * @private
     */
    ProjectSettings.prototype._apiBasePath;
    /**
     * @type {?}
     * @private
     */
    ProjectSettings.prototype._apiOrg;
    /**
     * @type {?}
     * @private
     */
    ProjectSettings.prototype._apiEnv;
}
