// @cemex-core/angular-localization-v7/fesm5/cemex-core-angular-localization-v7.js
import * as moment_ from 'moment';
import { __values } from 'tslib';
import { HttpClient } from '@angular/common/http';
import { isPlatformBrowser, CommonModule } from '@angular/common';
import { ReplaySubject, Subject } from 'rxjs';
import { Inject, Injectable, Optional, Pipe, PLATFORM_ID, NgModule } from '@angular/core';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var TranslationService = /** @class */ (function () {
    // we can change 'languageLoaded' to 'selectedLanguageSubject' in translation resolver  ???
    function TranslationService(httpClient, productPath, loadedLanguages, defaultLanguageISO // e.g.: en_US from america region || en_UK from europe region.
    ) {
        this.httpClient = httpClient;
        this.productPath = productPath;
        this.loadedLanguages = loadedLanguages;
        this.defaultLanguageISO = defaultLanguageISO;
        this.languageLoaded = new ReplaySubject(1);
        this.selectedLanguageSubject = new ReplaySubject(1);
        this.languagesSubject = new ReplaySubject(1);
        if (!this.loadedLanguages) {
            console.warn('The Translation Service can\'t init without Injected Languages.');
            console.warn('Inject the languages from your application: > ' +
                '{ provide: \'TRANSLATION_LANGUAGES, useValue: window[\'CMX_LANGUAGES\'] },\'');
            return;
        }
        if (!defaultLanguageISO) {
            console.warn('The Translation Service need defaultLanguageISO.');
        }
        TranslationService.languages = this.loadedLanguages;
        /** @type {?} */
        var localStorageLanguage = ((/** @type {?} */ (global))).localStorage.getItem('language');
        /** @type {?} */
        var allKeys = Object.keys(TranslationService.translations);
        if (allKeys.length === 0) {
            this.setLanguage(localStorageLanguage || defaultLanguageISO);
        }
        else {
            this.languageLoaded.next(true);
            // we can change 'languageLoaded' to 'selectedLanguageSubject' in translation resolver?
        }
    }
    /**
     * @description Gets the labels data from the server.
     * @param $lang Language code for the file
     */
    /**
     * \@description Gets the labels data from the server.
     * @param {?} $lang Language code for the file
     * @return {?}
     */
    TranslationService.prototype.getTranslations = /**
     * \@description Gets the labels data from the server.
     * @param {?} $lang Language code for the file
     * @return {?}
     */
    function ($lang) {
        /** @type {?} */
        var endpoint = "/translate/translate/" + this.productPath + "/" + $lang;
        return this.httpClient.get(endpoint);
    };
    /**
     * @description Observable to gets the languages that exist in the server
     */
    /**
     * \@description Observable to gets the languages that exist in the server
     * @return {?}
     */
    TranslationService.prototype.getLanguages = /**
     * \@description Observable to gets the languages that exist in the server
     * @return {?}
     */
    function () {
        return this.languagesSubject.asObservable();
    };
    /**
     * @return {?}
     */
    TranslationService.prototype.getSelectedLanguage = /**
     * @return {?}
     */
    function () {
        return this.selectedLanguageSubject.asObservable();
    };
    /**
     * @description Retrieve the public translation of a label
     * @param $textId Label id as written in the json file
     */
    /**
     * \@description Retrieve the public translation of a label
     * @param {?} $textId Label id as written in the json file
     * @return {?}
     */
    TranslationService.prototype.pt = /**
     * \@description Retrieve the public translation of a label
     * @param {?} $textId Label id as written in the json file
     * @return {?}
     */
    function ($textId) {
        /** @type {?} */
        var textValue = TranslationService.translations[$textId];
        if (!textValue || textValue === undefined) {
            return 'NOT:' + $textId;
        }
        return textValue;
    };
    /**
     * @description Retrieve the public translation of a label
     * @param $textId Label id as written in the json file
     */
    /**
     * \@description Retrieve the public translation of a label
     * @param {?} $textId Label id as written in the json file
     * @return {?}
     */
    TranslationService.prototype.getLabel = /**
     * \@description Retrieve the public translation of a label
     * @param {?} $textId Label id as written in the json file
     * @return {?}
     */
    function ($textId) {
        /** @type {?} */
        var textValue = TranslationService.translations[$textId];
        if (!textValue || textValue === undefined) {
            return 'NOT:' + $textId;
        }
        return textValue;
    };
    /**
     * @description Retrieve language published on the server.
     * @param $languageISO Can be the country code or the language iso combination
     */
    /**
     * \@description Retrieve language published on the server.
     * @param {?} $languageISO Can be the country code or the language iso combination
     * @return {?}
     */
    TranslationService.prototype.getLanguage = /**
     * \@description Retrieve language published on the server.
     * @param {?} $languageISO Can be the country code or the language iso combination
     * @return {?}
     */
    function ($languageISO) {
        if ($languageISO !== undefined && $languageISO !== null) {
            /** @type {?} */
            var langValue = void 0;
            langValue = TranslationService.languages.find((/**
             * @param {?} language
             * @return {?}
             */
            function (language) {
                return (language.languageISO === $languageISO ||
                    language.countryCode.toLowerCase() === $languageISO.toLowerCase() ||
                    language.languageISO.includes($languageISO));
            }));
            // if the language cannot reached, go for default
            if (langValue === undefined) {
                console.warn('The Language cannot reached', $languageISO);
                langValue = this.getDefaultLanguage();
            }
            langValue.momentConfig = this.getMomentConfig(langValue);
            return langValue;
        }
        else {
            console.warn('$languageIso param is undefined.');
            return this.getDefaultLanguage();
        }
    };
    /**
     * @description Retrieve language published on the server.
     * @param $countryCode Can be the country code or the language iso combination
     */
    /**
     * \@description Retrieve language published on the server.
     * @param {?} $countryCode Can be the country code or the language iso combination
     * @return {?}
     */
    TranslationService.prototype.getLanguageByCountryCode = /**
     * \@description Retrieve language published on the server.
     * @param {?} $countryCode Can be the country code or the language iso combination
     * @return {?}
     */
    function ($countryCode) {
        if ($countryCode !== undefined && $countryCode !== null) {
            /** @type {?} */
            var langValue = void 0;
            /** @type {?} */
            var countryLan = TranslationService.languages.filter((/**
             * @param {?} $language
             * @return {?}
             */
            function ($language) {
                return ($language.countryCode.toLowerCase() === $countryCode.toLowerCase());
            }));
            if (countryLan.length > 1) {
                /** @type {?} */
                var languageCode_1 = sessionStorage.getItem('language');
                langValue = countryLan.find((/**
                 * @param {?} $language
                 * @return {?}
                 */
                function ($language) {
                    return ($language.languageISO.toLowerCase() === languageCode_1.toLowerCase());
                }));
            }
            else {
                langValue =
                    countryLan && countryLan.length > 0 ? countryLan[0] : undefined;
            }
            // if the language cannot reached, go for default
            if (langValue === undefined) {
                console.warn('The Language cannot reached', $countryCode);
                langValue = this.getDefaultLanguage();
            }
            langValue.momentConfig = this.getMomentConfig(langValue);
            return langValue;
        }
        else {
            console.warn('$countryCode param is undefined.');
            return this.getDefaultLanguage();
        }
    };
    /**
     * @description Set language by languageISO or CountryCode.
     * @param $languageISO Can be the country code or the language iso combination
     */
    /**
     * \@description Set language by languageISO or CountryCode.
     * @param {?} $languageISO Can be the country code or the language iso combination
     * @return {?}
     */
    TranslationService.prototype.setLanguage = /**
     * \@description Set language by languageISO or CountryCode.
     * @param {?} $languageISO Can be the country code or the language iso combination
     * @return {?}
     */
    function ($languageISO) {
        var _this = this;
        var e_1, _a;
        this.selectedLanguage = this.getLanguage($languageISO);
        try {
            for (var _b = __values(TranslationService.languages), _c = _b.next(); !_c.done; _c = _b.next()) {
                var lang = _c.value;
                lang.isSelected = lang.languageISO === this.selectedLanguage.languageISO;
                lang.dir = lang.textFloat === 'right' ? 'rtl' : 'ltr';
                lang.rtl = lang.dir === 'rtl' ? true : false;
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        this.getTranslations(this.selectedLanguage.languageISO).subscribe((/**
         * @param {?} translations
         * @return {?}
         */
        function (translations) { return _this.populateTranslation(translations); }));
    };
    /**
     * @param {?} $result
     * @return {?}
     */
    TranslationService.prototype.populateTranslation = /**
     * @param {?} $result
     * @return {?}
     */
    function ($result) {
        TranslationService.translations = $result;
        // emit changes until populate is completed
        localStorage.setItem('language', this.selectedLanguage.languageISO);
        sessionStorage.setItem('language', this.selectedLanguage.languageISO);
        this.languageLoaded.next(true);
        // we can change 'languageLoaded' to 'selectedLanguageSubject' in translation resolver?
        this.selectedLanguageSubject.next(this.selectedLanguage);
        this.languagesSubject.next(TranslationService.languages);
    };
    Object.defineProperty(TranslationService.prototype, "selectedLanguage", {
        get: /**
         * @return {?}
         */
        function () {
            return TranslationService.currentLanguage;
        },
        set: /**
         * @param {?} language
         * @return {?}
         */
        function (language) {
            TranslationService.currentLanguage = language;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    TranslationService.prototype.getDefaultLanguage = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var defaultLangValue;
        defaultLangValue = TranslationService.languages.find((/**
         * @param {?} language
         * @return {?}
         */
        function (language) {
            return (language.languageISO === _this.defaultLanguageISO ||
                language.countryCode.toLowerCase() ===
                    _this.defaultLanguageISO.toLowerCase() ||
                language.languageISO.includes(_this.defaultLanguageISO));
        }));
        if (defaultLangValue === undefined) {
            console.warn('The Default Language cannot reached', this.defaultLanguageISO);
            defaultLangValue = TranslationService.languages[0];
            console.warn('Trying get the first language', defaultLangValue.languageISO);
        }
        else {
            console.warn('The Default Language is', defaultLangValue.languageISO);
        }
        return defaultLangValue;
    };
    /**
     * @param {?} countryCode
     * @return {?}
     */
    TranslationService.prototype.getLanguagesByCountryCode = /**
     * @param {?} countryCode
     * @return {?}
     */
    function (countryCode) {
        /** @type {?} */
        var languagesByContryCode = TranslationService.languages.filter((/**
         * @param {?} language
         * @return {?}
         */
        function (language) {
            return language.countryCode.toLowerCase() === countryCode.toLowerCase();
        }));
        return languagesByContryCode;
    };
    /**
     * @private
     * @param {?} language
     * @return {?}
     */
    TranslationService.prototype.getMomentConfig = /**
     * @private
     * @param {?} language
     * @return {?}
     */
    function (language) {
        return {
            months: language.monthNames.split(','),
            monthsShort: language.shortDayMonths.split(','),
            weekdays: language.dayNames.split(','),
            weekdaysMin: language.shortDayNames2.split(','),
            weekdaysShort: language.shortDayNames.split(',')
        };
    };
    TranslationService.TRANSLATION_PRODUCT_PATH = 'TRANSLATION_PRODUCT_PATH';
    TranslationService.TRANSLATION_LANGUAGES = 'TRANSLATION_LANGUAGES';
    TranslationService.DEFAULT_LANGUAGE_ISO = 'DEFAULT_LANGUAGE_ISO';
    TranslationService.currentLanguage = undefined;
    TranslationService.translations = {};
    TranslationService.languages = [];
    TranslationService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    TranslationService.ctorParameters = function () { return [
        { type: HttpClient },
        { type: String, decorators: [{ type: Inject, args: [TranslationService.TRANSLATION_PRODUCT_PATH,] }, { type: Optional }] },
        { type: undefined, decorators: [{ type: Inject, args: [TranslationService.TRANSLATION_LANGUAGES,] }, { type: Optional }] },
        { type: undefined, decorators: [{ type: Inject, args: [TranslationService.DEFAULT_LANGUAGE_ISO,] }, { type: Optional }] }
    ]; };
    return TranslationService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * From https://github.com/shrpne/from-exponential
 * Can't install the package beacause it is ES6, UglifyJS on its version compatible with Webpack 3 can't do ES6
 * The code needs to be copied locally so Typescript could compile it to ES5
 * or just upgrade to Webpack 4 and use the webpack-uglify-plugin
 */
/**
 * Return two parts array of exponential number
 * @param {?} num - number or array of its parts
 * @return {?}
 */
function getExponentialParts(num) {
    return Array.isArray(num) ? num : String(num).split(/[eE]/);
}
/**
 *
 * @param {?} num - number or array of its parts
 * @return {?}
 */
function isExponential(num) {
    /** @type {?} */
    var eParts = getExponentialParts(num);
    return !Number.isNaN(Number(eParts[1]));
}
/**
 * Converts exponential notation to a human readable string
 * @param {?} num - number or array of its parts
 * @return {?}
 */
function fromExponential(num) {
    /** @type {?} */
    var eParts = getExponentialParts(num);
    if (!isExponential(eParts)) {
        return eParts[0];
    }
    /** @type {?} */
    var sign = eParts[0][0] === '-' ? '-' : '';
    /** @type {?} */
    var digits = eParts[0].replace(/^-/, '');
    /** @type {?} */
    var digitsParts = digits.split('.');
    /** @type {?} */
    var wholeDigits = digitsParts[0];
    /** @type {?} */
    var fractionDigits = digitsParts[1] || '';
    /** @type {?} */
    var e = Number(eParts[1]);
    if (e === 0) {
        return sign + wholeDigits + "." + fractionDigits;
    }
    else if (e < 0) {
        // move dot to the left
        /** @type {?} */
        var countWholeAfterTransform = wholeDigits.length + e;
        if (countWholeAfterTransform > 0) {
            // transform whole to fraction
            /** @type {?} */
            var wholeDigitsAfterTransform = wholeDigits.substr(0, countWholeAfterTransform);
            /** @type {?} */
            var wholeDigitsTransformedToFracton = wholeDigits.substr(countWholeAfterTransform);
            return sign +
                wholeDigitsAfterTransform + "." + wholeDigitsTransformedToFracton + fractionDigits;
        }
        else {
            // not enough whole digits: prepend with fractional zeros
            // first e goes to dotted zero
            /** @type {?} */
            var zeros = '0.';
            e += 1;
            while (e) {
                zeros += '0';
                e += 1;
            }
            return sign + zeros + wholeDigits + fractionDigits;
        }
    }
    else {
        // move dot to the right
        /** @type {?} */
        var countFractionAfterTransform = fractionDigits.length - e;
        if (countFractionAfterTransform > 0) {
            // transform fraction to whole
            // countTransformedFractionToWhole = e
            /** @type {?} */
            var fractionDigitsAfterTransform = fractionDigits.substr(e);
            /** @type {?} */
            var fractionDigitsTransformedToWhole = fractionDigits.substr(0, e);
            return sign +
                wholeDigits +
                fractionDigitsTransformedToWhole + "." + fractionDigitsAfterTransform;
        }
        else {
            // not enough fractions: append whole zeros
            /** @type {?} */
            var zerosCount = -countFractionAfterTransform;
            /** @type {?} */
            var zeros = '';
            while (zerosCount) {
                zeros += '0';
                zerosCount -= 1;
            }
            return sign + wholeDigits + fractionDigits + zeros;
        }
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var moment = moment_;
/** @enum {number} */
var DateTimeActions = {
    ConvertToLocal: 0,
    NoConvertToLocal: 1,
};
DateTimeActions[DateTimeActions.ConvertToLocal] = 'ConvertToLocal';
DateTimeActions[DateTimeActions.NoConvertToLocal] = 'NoConvertToLocal';
var FormatterService = /** @class */ (function () {
    function FormatterService() {
    }
    // DATES
    /**
     * @description Transforms a valid string to a date object
     * @param date The date as string
     * @returns A javascript date object
     */
    // DATES
    /**
     * \@description Transforms a valid string to a date object
     * @param {?} date The date as string
     * @return {?} A javascript date object
     */
    FormatterService.prototype.getDateObject = 
    // DATES
    /**
     * \@description Transforms a valid string to a date object
     * @param {?} date The date as string
     * @return {?} A javascript date object
     */
    function (date) {
        try {
            /** @type {?} */
            var auxDate = date.replace(/-g/, '/');
            /** @type {?} */
            var bits = auxDate.split(/\D/);
            /** @type {?} */
            var month = parseInt(bits[1], undefined) - 1;
            if (bits.length === 3) {
                return new Date(bits[0], month, bits[2]);
            }
            else {
                return new Date(bits[0], month, bits[2], bits[3], bits[4], bits[5]);
            }
        }
        catch ($exception) {
            return undefined;
        }
    };
    /**
     * @param {?} date
     * @param {?} languageISO
     * @param {?} format
     * @param {?=} action
     * @return {?}
     */
    FormatterService.prototype.formatDateTimeMoment = /**
     * @param {?} date
     * @param {?} languageISO
     * @param {?} format
     * @param {?=} action
     * @return {?}
     */
    function (date, languageISO, format, action) {
        if (action === void 0) { action = DateTimeActions.ConvertToLocal; }
        if (TranslationService.currentLanguage) {
            moment.updateLocale(TranslationService.currentLanguage.languageISO.substr(0, 2), TranslationService.currentLanguage.momentConfig);
        }
        if (action === DateTimeActions.NoConvertToLocal) {
            return moment(date)
                .locale(languageISO.substr(0, 2))
                .format(format);
        }
        return this.utcToLocal(date)
            .locale(languageISO.substr(0, 2))
            .format(format);
    };
    /**
     * @description Format a number by TypeScript Number function
     * The static value '-u-nu-latn' is for always returns commons numbers '1,2,3,4,5,6,7,8,9,0'
     * @param numberEntry: value to format
     * @param languageISO: specifict language to format
     * @param decimalNumbers: quantity decimals to set in the format
     * @returns A value formatted
     */
    /**
     * \@description Format a number by TypeScript Number function
     * The static value '-u-nu-latn' is for always returns commons numbers '1,2,3,4,5,6,7,8,9,0'
     * @param {?} numberEntry
     * @param {?} languageISO
     * @param {?} decimalNumbers
     * @return {?} A value formatted
     */
    FormatterService.prototype.formatNumberTS = /**
     * \@description Format a number by TypeScript Number function
     * The static value '-u-nu-latn' is for always returns commons numbers '1,2,3,4,5,6,7,8,9,0'
     * @param {?} numberEntry
     * @param {?} languageISO
     * @param {?} decimalNumbers
     * @return {?} A value formatted
     */
    function (numberEntry, languageISO, decimalNumbers) {
        if (isNaN(numberEntry) || numberEntry === null) {
            return 0;
        }
        return Number(this.round(numberEntry, decimalNumbers).toFixed(decimalNumbers)).toLocaleString(languageISO + '-u-nu-latn', {
            minimumFractionDigits: numberEntry % 1 === 0 ? 0 : decimalNumbers
        });
    };
    /**
     * @param {?} dateEntry
     * @return {?}
     */
    FormatterService.prototype.utcToLocal = /**
     * @param {?} dateEntry
     * @return {?}
     */
    function (dateEntry) {
        return moment.utc(dateEntry).local();
    };
    /**
     * @param {?} date
     * @param {?=} format
     * @return {?}
     */
    FormatterService.prototype.localToUtc = /**
     * @param {?} date
     * @param {?=} format
     * @return {?}
     */
    function (date, format) {
        if (format !== undefined) {
            return moment(date, 'YYYY-MM-DD THH:mm:ssZ')
                .utc()
                .format(format);
        }
        else {
            return moment(date, 'YYYY-MM-DD THH:mm:ssZ')
                .utc()
                .format();
        }
    };
    // expected time '00:00:00'
    // expected time '00:00:00'
    /**
     * @param {?} time
     * @return {?}
     */
    FormatterService.prototype.timeToMinutes = 
    // expected time '00:00:00'
    /**
     * @param {?} time
     * @return {?}
     */
    function (time) {
        /** @type {?} */
        var timeSplit = time.split(':');
        /** @type {?} */
        var hours = +timeSplit[0];
        /** @type {?} */
        var minutes = +timeSplit[1];
        /** @type {?} */
        var seconds = +timeSplit[2];
        return hours * 60 + minutes + this.roundToDown(seconds / 60);
    };
    // NUMBERS
    /**
     * @description Formats a number as a currency preserving its precision.
     * In order the keep the precision the number must be passed as a string
     * @param value The number value to format
     * @param numberDecimals The number of decimals to use
     * @param thousandSeparator The string character to use when separating numbers
     * @param decimalSeparator The string character to use when separating the decimal and integer part
     * @returns The number as a currency string
     */
    // NUMBERS
    /**
     * \@description Formats a number as a currency preserving its precision.
     * In order the keep the precision the number must be passed as a string
     * @param {?} value The number value to format
     * @param {?=} numberDecimals The number of decimals to use
     * @param {?=} thousandSeparator The string character to use when separating numbers
     * @param {?=} decimalSeparator The string character to use when separating the decimal and integer part
     * @return {?} The number as a currency string
     */
    FormatterService.prototype.formatNumber = 
    // NUMBERS
    /**
     * \@description Formats a number as a currency preserving its precision.
     * In order the keep the precision the number must be passed as a string
     * @param {?} value The number value to format
     * @param {?=} numberDecimals The number of decimals to use
     * @param {?=} thousandSeparator The string character to use when separating numbers
     * @param {?=} decimalSeparator The string character to use when separating the decimal and integer part
     * @return {?} The number as a currency string
     */
    function (value, numberDecimals, thousandSeparator, decimalSeparator) {
        // tslint:disable-next-line: triple-equals
        if (value == undefined || value.toString() == '' || isNaN((/** @type {?} */ (value)))) {
            console.warn('Not a valid number to format');
            return '';
        }
        // tslint:disable-next-line:triple-equals
        numberDecimals = numberDecimals != undefined ? numberDecimals : 2;
        // tslint:disable-next-line:triple-equals
        thousandSeparator =
            thousandSeparator !== undefined ? thousandSeparator : ',';
        // tslint:disable-next-line:triple-equals
        decimalSeparator = decimalSeparator !== undefined ? decimalSeparator : '.';
        /** @type {?} */
        var completeNumber = fromExponential(value);
        /** @type {?} */
        var sign = value < 0 ? '-' : '';
        /** @type {?} */
        var absoluteValue = value < 0 ? completeNumber.split('-')[1] : completeNumber;
        /** @type {?} */
        var truncatedDecimalValue = absoluteValue.split('.')[1] &&
            absoluteValue.split('.')[1].slice(0, numberDecimals);
        /** @type {?} */
        var integerPart = absoluteValue.split('.')[0];
        /** @type {?} */
        var floatingPart = this.rightPad(truncatedDecimalValue, '0', numberDecimals);
        /** @type {?} */
        var formattedInteger = integerPart;
        if (Number(integerPart) >= 1000) {
            formattedInteger = integerPart.replace(/./g, (/**
             * @param {?} c
             * @param {?} i
             * @param {?} a
             * @return {?}
             */
            function (c, i, a) {
                return i && c !== decimalSeparator && (a.length - i) % 3 === 0
                    ? thousandSeparator + c
                    : c;
            }));
        }
        // tslint:disable-next-line:max-line-length
        /** @type {?} */
        var formattedNumber = "" + sign + formattedInteger + (floatingPart.length > 0 ? decimalSeparator : '') + floatingPart;
        return formattedNumber;
    };
    /**
     * @param {?} value
     * @param {?} decimals
     * @return {?}
     */
    FormatterService.prototype.round = /**
     * @param {?} value
     * @param {?} decimals
     * @return {?}
     */
    function (value, decimals) {
        /** @type {?} */
        var type = '';
        type = 'round';
        /** @type {?} */
        var valueDecimals = value % 1;
        if (valueDecimals !== 0.5) {
            return Number(Math[type](value + 'e' + decimals) + 'e-' + decimals);
        }
        else {
            return Number(value);
        }
    };
    /**
     * @param {?} value
     * @return {?}
     */
    FormatterService.prototype.roundToDown = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return Math.floor(value);
    };
    // STRINGS
    // STRINGS
    /**
     * @param {?} value
     * @return {?}
     */
    FormatterService.prototype.removeFirstZeros = 
    // STRINGS
    /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return value.replace(/^0+/, '');
    };
    /**
     * @param {?} text
     * @param {?} wildcard
     * @param {?} length
     * @return {?}
     */
    FormatterService.prototype.pad = /**
     * @param {?} text
     * @param {?} wildcard
     * @param {?} length
     * @return {?}
     */
    function (text, wildcard, length) {
        /** @type {?} */
        var result = text;
        while (result.length < length) {
            result = wildcard + result;
        }
        return result;
    };
    /**
     * @description Fills up a string with the specified character appending it to the right till lenght is achieved
     * @param text String to add padding
     * @param wildcard Character that's gonne be repeated
     * @param lenght The desired length
     */
    /**
     * \@description Fills up a string with the specified character appending it to the right till lenght is achieved
     * @param {?} text String to add padding
     * @param {?} wildcard Character that's gonne be repeated
     * @param {?} length
     * @return {?}
     */
    FormatterService.prototype.rightPad = /**
     * \@description Fills up a string with the specified character appending it to the right till lenght is achieved
     * @param {?} text String to add padding
     * @param {?} wildcard Character that's gonne be repeated
     * @param {?} length
     * @return {?}
     */
    function (text, wildcard, length) {
        if (text === undefined || text === null || text === '') {
            return wildcard.repeat(length);
        }
        /** @type {?} */
        var str = text.toString();
        while (str.length < length) {
            str += wildcard;
        }
        return str;
    };
    /**
     * @param {?} measureUnit
     * @return {?}
     */
    FormatterService.prototype.formatMeasureUnit = /**
     * @param {?} measureUnit
     * @return {?}
     */
    function (measureUnit) {
        switch (measureUnit) {
            case 'YD3':
                return 'Yd³';
            case 'M3':
                return 'M³';
            default:
                return measureUnit;
        }
    };
    FormatterService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    FormatterService.ctorParameters = function () { return []; };
    return FormatterService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var CmxCurrencyCustomPipe = /** @class */ (function () {
    function CmxCurrencyCustomPipe(formatter) {
        this.formatter = formatter;
        this.value = '';
    }
    /**
     * @param {?} $value
     * @param {?} $currencySymbol
     * @param {?} $decimalNumbers
     * @param {?} $thousandSeparator
     * @param {?} $decimalSeparator
     * @param {?=} $floatCurrencySymbolRight
     * @return {?}
     */
    CmxCurrencyCustomPipe.prototype.transform = /**
     * @param {?} $value
     * @param {?} $currencySymbol
     * @param {?} $decimalNumbers
     * @param {?} $thousandSeparator
     * @param {?} $decimalSeparator
     * @param {?=} $floatCurrencySymbolRight
     * @return {?}
     */
    function ($value, $currencySymbol, $decimalNumbers, $thousandSeparator, $decimalSeparator, $floatCurrencySymbolRight) {
        try {
            if ($value !== undefined && isNaN($value) === false) {
                this.value = $value.toString();
                $decimalNumbers = parseInt($decimalNumbers.toString(), undefined);
                if ($value < 0) {
                    // value is negative
                    /** @type {?} */
                    var v = parseFloat($value.toString().replace(/\-/g, ''));
                    /** @type {?} */
                    var $aux = this.formatter.formatNumber(v, $decimalNumbers, $thousandSeparator, $decimalSeparator);
                    if ($floatCurrencySymbolRight) {
                        this.value = '-' + $aux + $currencySymbol;
                    }
                    else {
                        this.value = '-' + $currencySymbol + $aux;
                    }
                }
                else {
                    this.value =
                        $currencySymbol +
                            this.formatter.formatNumber($value, $decimalNumbers, $thousandSeparator, $decimalSeparator);
                }
            }
            else {
                this.value = '';
            }
        }
        catch ($exception) {
            console.error('Value trying to transform: ', $value);
            console.error($exception);
            this.value = '';
        }
        return this.value;
    };
    CmxCurrencyCustomPipe.decorators = [
        { type: Pipe, args: [{
                    name: 'cmxCurrencyCustom',
                    pure: true
                },] }
    ];
    /** @nocollapse */
    CmxCurrencyCustomPipe.ctorParameters = function () { return [
        { type: FormatterService }
    ]; };
    return CmxCurrencyCustomPipe;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var CmxCurrencyPipe = /** @class */ (function () {
    function CmxCurrencyPipe(formatter, translation, platformId) {
        this.formatter = formatter;
        this.translation = translation;
        this.platformId = platformId;
        if (isPlatformBrowser(platformId)) {
            this.window = window;
        }
    }
    /**
     * @param {?} value
     * @param {?=} countryCode
     * @return {?}
     */
    CmxCurrencyPipe.prototype.transform = /**
     * @param {?} value
     * @param {?=} countryCode
     * @return {?}
     */
    function (value, countryCode) {
        if (value === '' || value == null || isNaN((/** @type {?} */ (value)))) {
            console.warn('The value is not a number');
            return '';
        }
        countryCode = countryCode || this.window.sessionStorage.getItem('country');
        // tslint:disable-next-line:triple-equals
        if (countryCode == undefined) {
            console.warn("Can't retrieve countryCode");
            return '';
        }
        /** @type {?} */
        var lang = this.translation.getLanguageByCountryCode(countryCode);
        /** @type {?} */
        var symbolPosition = (lang.currencySymbolFloat && lang.currencySymbolFloat.toLowerCase()) ||
            'left';
        /** @type {?} */
        var currencySymbol = lang.currencySymbol || '$';
        /** @type {?} */
        var formattedNumber = this.formatter.formatNumber(value, lang && lang.decimalNumbersMoney, lang && lang.thousandSeparator, lang && lang.decimalSeparator);
        /** @type {?} */
        var numberWithCurrency = symbolPosition === 'left'
            ? "" + currencySymbol + formattedNumber
            : "" + formattedNumber + currencySymbol;
        return '&lrm;' + numberWithCurrency;
    };
    CmxCurrencyPipe.decorators = [
        { type: Pipe, args: [{
                    name: 'cmxCurrency',
                    pure: true
                },] }
    ];
    /** @nocollapse */
    CmxCurrencyPipe.ctorParameters = function () { return [
        { type: FormatterService },
        { type: TranslationService },
        { type: undefined, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
    ]; };
    return CmxCurrencyPipe;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var CmxDateMomentPipe = /** @class */ (function () {
    function CmxDateMomentPipe(translationService, formatterService) {
        this.translationService = translationService;
        this.formatterService = formatterService;
    }
    /**
     * @param {?} dateTime
     * @param {?=} action
     * @return {?}
     */
    CmxDateMomentPipe.prototype.transform = /**
     * @param {?} dateTime
     * @param {?=} action
     * @return {?}
     */
    function (dateTime, action) {
        if (action === void 0) { action = DateTimeActions.ConvertToLocal; }
        return this.formatterService.formatDateTimeMoment(dateTime, this.translationService.selectedLanguage.languageISO.replace('_', '-'), this.translationService.selectedLanguage.formatDate, action);
    };
    CmxDateMomentPipe.decorators = [
        { type: Pipe, args: [{
                    name: 'cmxDateMoment',
                    pure: false
                },] }
    ];
    /** @nocollapse */
    CmxDateMomentPipe.ctorParameters = function () { return [
        { type: TranslationService },
        { type: FormatterService }
    ]; };
    return CmxDateMomentPipe;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var DateFormatterService = /** @class */ (function () {
    function DateFormatterService() {
        //
    }
    /**
     * @description Formats a date object to a string given a date format from a language
     * @param value The date value to format
     * @param format The format to comply, e.x. DD MM, YYYY
     * @param daysNames String array of the weekdays' names
     * @param monthsNames String array of the months' names
     * @returns String of the formatted date
     */
    /**
     * \@description Formats a date object to a string given a date format from a language
     * @param {?} $value
     * @param {?} $format
     * @param {?} $daysNames
     * @param {?} $monthsNames
     * @return {?} String of the formatted date
     */
    DateFormatterService.prototype.format = /**
     * \@description Formats a date object to a string given a date format from a language
     * @param {?} $value
     * @param {?} $format
     * @param {?} $daysNames
     * @param {?} $monthsNames
     * @return {?} String of the formatted date
     */
    function ($value, $format, $daysNames, $monthsNames) {
        var _this = this;
        /** @type {?} */
        var dateAsString = '';
        /** @type {?} */
        var $separator = '/';
        /** @type {?} */
        var day;
        /** @type {?} */
        var month;
        /** @type {?} */
        var year;
        /** @type {?} */
        var $formats = $format.split($separator);
        if ($formats.length === 1) {
            $separator = '-';
            $formats = $format.split($separator);
        }
        if ($formats.length === 1) {
            $separator = ' ';
            $formats = $format.split($separator);
        }
        if ($formats.length === 1) {
            $separator = '.';
            $formats = $format.split($separator);
        }
        $formats.forEach((/**
         * @param {?} $formatItem
         * @param {?} $index
         * @return {?}
         */
        function ($formatItem, $index) {
            /** @type {?} */
            var $auxValue = $value;
            /** @type {?} */
            var $formatCode = $formatItem.replace(',', '');
            if ($formatItem.includes('d') || $formatItem.includes('D')) {
                day = _this.getDay($auxValue, $formatCode, $daysNames);
                $formatItem = $formatItem.replace($formatCode, day);
            }
            else if ($formatItem.includes('m') || $formatItem.includes('M')) {
                month = _this.getMonth($auxValue, $formatCode, $monthsNames);
                $formatItem = $formatItem.replace($formatCode, month);
            }
            else if ($formatItem.includes('y') || $formatItem.includes('Y')) {
                year = _this.getYear($auxValue, $formatCode);
                $formatItem = $formatItem.replace($formatCode, year);
            }
            if ($index !== $formats.length - 1) {
                dateAsString += $formatItem + $separator;
            }
            else {
                dateAsString += $formatItem;
            }
        }));
        return dateAsString;
    };
    /**
     * @description Transforms a valid string to a date object
     * @param date The date as string
     * @returns A javascript date object
     */
    /**
     * \@description Transforms a valid string to a date object
     * @param {?} date The date as string
     * @return {?} A javascript date object
     */
    DateFormatterService.prototype.getDateObject = /**
     * \@description Transforms a valid string to a date object
     * @param {?} date The date as string
     * @return {?} A javascript date object
     */
    function (date) {
        try {
            /** @type {?} */
            var auxDate = date.replace(/-g/, '/');
            /** @type {?} */
            var bits = auxDate.split(/\D/);
            /** @type {?} */
            var month = parseInt(bits[1], undefined) - 1;
            if (bits.length === 3) {
                return new Date(bits[0], month, bits[2]);
            }
            else {
                return new Date(bits[0], month, bits[2], bits[3], bits[4], bits[5]);
            }
        }
        catch ($exception) {
            return undefined;
        }
    };
    /**
     * @private
     * @param {?} $number
     * @return {?}
     */
    DateFormatterService.prototype.addZero = /**
     * @private
     * @param {?} $number
     * @return {?}
     */
    function ($number) {
        return $number < 10 ? '0' + $number : $number.toString();
    };
    /**
     * @private
     * @param {?} $date
     * @param {?} $format
     * @param {?} $daysNames
     * @return {?}
     */
    DateFormatterService.prototype.getDay = /**
     * @private
     * @param {?} $date
     * @param {?} $format
     * @param {?} $daysNames
     * @return {?}
     */
    function ($date, $format, $daysNames) {
        /** @type {?} */
        var day;
        if ($format.length === 4) {
            // requires the complete name of the day
            day = $daysNames[$date.getDay()];
        }
        else if ($format.length === 3) {
            // requires an abbreviation of the day's name
            day = $daysNames[$date.getDay()].substring(0, 3);
        }
        else if ($format.length === 2 || $format.length === 1) {
            // requires the value of the day
            day = this.addZero($date.getDate());
        }
        // check if names should use lower or uppercase
        if ($format === $format.toLowerCase()) {
            // the format uses d
            day = day.toLowerCase();
        }
        return day;
    };
    /**
     * @private
     * @param {?} $date
     * @param {?} $format
     * @param {?} $monthsNames
     * @return {?}
     */
    DateFormatterService.prototype.getMonth = /**
     * @private
     * @param {?} $date
     * @param {?} $format
     * @param {?} $monthsNames
     * @return {?}
     */
    function ($date, $format, $monthsNames) {
        /** @type {?} */
        var month;
        if ($format.length === 4) {
            // requires the complete name of the month
            month = $monthsNames[$date.getMonth()];
        }
        else if ($format.length === 3) {
            // requires an abbreviation of the day's name
            month = $monthsNames[$date.getMonth()].substring(0, 3);
        }
        else if ($format.length === 2) {
            // requires the value of the day
            month = this.addZero($date.getMonth() + 1);
        }
        // check if names should use lower or uppercase
        if ($format === $format.toLowerCase()) {
            // the format uses d
            month = month.toLowerCase();
        }
        return month;
    };
    /**
     * @private
     * @param {?} $date
     * @param {?} $format
     * @return {?}
     */
    DateFormatterService.prototype.getYear = /**
     * @private
     * @param {?} $date
     * @param {?} $format
     * @return {?}
     */
    function ($date, $format) {
        /** @type {?} */
        var year;
        if ($format.length === 4) {
            // requires the complete year
            year = '' + $date.getFullYear();
        }
        else {
            // requires only the last 2 numbers of the year value
            year = '' + $date.getFullYear();
            year = year.substring(2);
        }
        return year;
    };
    DateFormatterService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    DateFormatterService.ctorParameters = function () { return []; };
    return DateFormatterService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var CmxDatePipe = /** @class */ (function () {
    function CmxDatePipe(translation, dateFormatter) {
        this.translation = translation;
        this.dateFormatter = dateFormatter;
        this.value = '';
        this.monthsNames = [];
        this.daysNames = [];
    }
    /**
     * @param {?} $value
     * @param {?=} $countryCode
     * @return {?}
     */
    CmxDatePipe.prototype.transform = /**
     * @param {?} $value
     * @param {?=} $countryCode
     * @return {?}
     */
    function ($value, $countryCode) {
        try {
            if (typeof $value === 'number') {
                $value = new Date($value);
            }
            if (typeof $value === 'string') {
                $value = this.dateFormatter.getDateObject($value);
            }
            if ($countryCode === undefined || $countryCode === null) {
                $countryCode = sessionStorage.getItem('country');
            }
            if ($value !== undefined && $value instanceof Date) {
                /** @type {?} */
                var $currentLanguage = localStorage.getItem('language');
                /** @type {?} */
                var $lang = this.translation.getLanguage($currentLanguage ? $currentLanguage : 'en_US');
                this.daysNames = $lang.dayNames.split(',');
                this.monthsNames = $lang.monthNames.split(',');
                this.value = this.dateFormatter.format($value, $lang.formatDate, this.daysNames, this.monthsNames);
            }
            else {
                console.error('Date trying to transform: ', $value);
                this.value = '';
            }
        }
        catch ($exception) {
            console.error('Date trying to transform: ', $value);
            console.error($exception);
            this.value = '';
        }
        return this.value;
    };
    CmxDatePipe.decorators = [
        { type: Pipe, args: [{
                    name: 'cmxDate',
                    pure: true
                },] }
    ];
    /** @nocollapse */
    CmxDatePipe.ctorParameters = function () { return [
        { type: TranslationService },
        { type: DateFormatterService }
    ]; };
    return CmxDatePipe;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var CmxDateTimeMomentPipe = /** @class */ (function () {
    function CmxDateTimeMomentPipe(translationService, formatterService) {
        this.translationService = translationService;
        this.formatterService = formatterService;
    }
    /**
     * @param {?} dateTime
     * @param {?=} action
     * @return {?}
     */
    CmxDateTimeMomentPipe.prototype.transform = /**
     * @param {?} dateTime
     * @param {?=} action
     * @return {?}
     */
    function (dateTime, action) {
        if (action === void 0) { action = DateTimeActions.ConvertToLocal; }
        return this.formatterService.formatDateTimeMoment(dateTime, this.translationService.selectedLanguage.languageISO.replace('_', '-'), this.translationService.selectedLanguage.formatDate +
            ' ' +
            this.translationService.selectedLanguage.formatTime, action);
    };
    CmxDateTimeMomentPipe.decorators = [
        { type: Pipe, args: [{
                    name: 'cmxDateTimeMoment',
                    pure: false
                },] }
    ];
    /** @nocollapse */
    CmxDateTimeMomentPipe.ctorParameters = function () { return [
        { type: TranslationService },
        { type: FormatterService }
    ]; };
    return CmxDateTimeMomentPipe;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var CmxNumberPipe = /** @class */ (function () {
    function CmxNumberPipe(translationService, formatterService) {
        this.translationService = translationService;
        this.formatterService = formatterService;
    }
    /**
     * @param {?} numberEntry
     * @return {?}
     */
    CmxNumberPipe.prototype.transform = /**
     * @param {?} numberEntry
     * @return {?}
     */
    function (numberEntry) {
        return this.formatterService.formatNumberTS(numberEntry, this.translationService.selectedLanguage.languageISO.replace('_', '-'), this.translationService.selectedLanguage.decimalNumbers);
    };
    CmxNumberPipe.decorators = [
        { type: Pipe, args: [{
                    name: 'cmxNumber',
                    pure: false
                },] }
    ];
    /** @nocollapse */
    CmxNumberPipe.ctorParameters = function () { return [
        { type: TranslationService },
        { type: FormatterService }
    ]; };
    return CmxNumberPipe;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var CmxTimeMomentPipe = /** @class */ (function () {
    function CmxTimeMomentPipe(translationService, formatterService) {
        this.translationService = translationService;
        this.formatterService = formatterService;
    }
    /**
     * @param {?} dateTime
     * @param {?=} action
     * @return {?}
     */
    CmxTimeMomentPipe.prototype.transform = /**
     * @param {?} dateTime
     * @param {?=} action
     * @return {?}
     */
    function (dateTime, action) {
        if (action === void 0) { action = DateTimeActions.ConvertToLocal; }
        return this.formatterService.formatDateTimeMoment(dateTime, this.translationService.selectedLanguage.languageISO.replace('_', '-'), this.translationService.selectedLanguage.formatTime, action);
    };
    CmxTimeMomentPipe.decorators = [
        { type: Pipe, args: [{
                    name: 'cmxTimeMoment',
                    pure: false
                },] }
    ];
    /** @nocollapse */
    CmxTimeMomentPipe.ctorParameters = function () { return [
        { type: TranslationService },
        { type: FormatterService }
    ]; };
    return CmxTimeMomentPipe;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var AngularLocalizationModule = /** @class */ (function () {
    function AngularLocalizationModule() {
    }
    AngularLocalizationModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        CmxCurrencyCustomPipe,
                        CmxCurrencyPipe,
                        CmxDateMomentPipe,
                        CmxDatePipe,
                        CmxDateTimeMomentPipe,
                        CmxNumberPipe,
                        CmxTimeMomentPipe
                    ],
                    exports: [
                        CmxCurrencyCustomPipe,
                        CmxCurrencyPipe,
                        CmxDateMomentPipe,
                        CmxDatePipe,
                        CmxDateTimeMomentPipe,
                        CmxNumberPipe,
                        CmxTimeMomentPipe
                    ],
                    imports: [CommonModule]
                },] }
    ];
    return AngularLocalizationModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var DirectionService = /** @class */ (function () {
    function DirectionService(translation) {
        var _this = this;
        this.translation = translation;
        this.direction$ = new Subject();
        this.rtl$ = new Subject();
        this.translation.getSelectedLanguage().subscribe((/**
         * @param {?} language
         * @return {?}
         */
        function (language) {
            _this.direction$.next(language.direction);
            _this.rtl$.next(language.rtl);
        }));
    }
    /**
     * @return {?}
     */
    DirectionService.prototype.getDirection = /**
     * @return {?}
     */
    function () {
        return this.direction$.asObservable();
    };
    /**
     * @return {?}
     */
    DirectionService.prototype.getRTL = /**
     * @return {?}
     */
    function () {
        return this.rtl$.asObservable();
    };
    /**
     * @param {?} languageCode
     * @return {?}
     */
    DirectionService.prototype.getDirectionFor = /**
     * @param {?} languageCode
     * @return {?}
     */
    function (languageCode) {
        /** @type {?} */
        var language = this.translation.getLanguage(languageCode.split(/[-_]/)[0]);
        this.direction$.next(language.dir);
    };
    /**
     * @param {?} languageCode
     * @return {?}
     */
    DirectionService.prototype.getRtlFor = /**
     * @param {?} languageCode
     * @return {?}
     */
    function (languageCode) {
        /** @type {?} */
        var language = this.translation.getLanguage(languageCode.split(/[-_]/)[0]);
        this.rtl$.next(language.rtl);
    };
    DirectionService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    DirectionService.ctorParameters = function () { return [
        { type: TranslationService }
    ]; };
    return DirectionService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var LocaleService = /** @class */ (function () {
    function LocaleService() {
        // empty
    }
    /**
     * @param {?} languageISO
     * @return {?}
     */
    LocaleService.prototype.getConfiguration = /**
     * @param {?} languageISO
     * @return {?}
     */
    function (languageISO) {
        switch (languageISO) {
            case 'en':
                return {
                    hourShort: 'hr',
                    minuteShort: 'min',
                    monthsShort: 'Jan,Feb,Mar,Apr,May,Jun,Jul,Aug,Sep,Oct,Nov,Dec',
                    secondShort: 'sec',
                    weekdaysMin: 'Su,Mo,Tu,We,Th,Fr,Sa',
                    weekdaysShort: 'Sun,Mon,Tue,Wed,Thu,Fri,Sat'
                };
            case 'es':
                return {
                    hourShort: 'hr',
                    minuteShort: 'min',
                    monthsShort: 'Ene,Feb,Mar,Abr,May,Jun,Jul,Ago,Sep,Oct,Nov,Dic',
                    secondShort: 'seg',
                    weekdaysMin: 'Do,Lu,Ma,Mi,Ju,Vi,Sá',
                    weekdaysShort: 'Dom,Lun,Mar,Mié,Jue,Vie,Sáb'
                };
            case 'fr':
                return {
                    hourShort: 'h',
                    minuteShort: 'min',
                    monthsShort: 'Janv,Févr,Mars,Avr,Mai,Juin,Juil,Août,Sept,Oct,Nov,Déc',
                    secondShort: 'sec',
                    weekdaysMin: 'Di,Lu,Ma,Me,Je,Ve,Sa',
                    weekdaysShort: 'Dim,Lun,Mar,Mer,Jeu,Ven,Sam'
                };
            case 'de':
                return {
                    hourShort: 'Std.',
                    minuteShort: 'Min.',
                    monthsShort: 'Jan,Febr,Mrz,Apr,Mai,Jun,Jul,Aug,Sept,Okt,Nov,Dez',
                    secondShort: 'Sekunde',
                    weekdaysMin: 'So,Mo,Di,Mi,Do,Fr,Sa',
                    weekdaysShort: 'So,Mo,Di,Mi,Do,Fr,Sa'
                };
            case 'he':
                return {
                    hourShort: 'שעה',
                    minuteShort: 'דקות',
                    monthsShort: 'ינו׳,פבר׳,מרץ,אפר׳,מאי,יוני,יולי,אוג׳,ספט׳,אוק׳,נוב׳,דצמ׳',
                    secondShort: 'שנייה',
                    weekdaysMin: 'א,ב,ג,ד,ה,ו,ש',
                    weekdaysShort: 'א׳,ב׳,ג׳,ד׳,ה׳,ו׳,ש׳'
                };
            case 'pl':
                return {
                    hourShort: 'h',
                    minuteShort: 'min',
                    monthsShort: 'Sty,Lut,Mar,Kwi,Maj,Cze,Lip,Sie,Wrz,Paź,Lis,Gru',
                    secondShort: 's',
                    weekdaysMin: 'Nd,Pn,Wt,Śr,Cz,Pt,So',
                    weekdaysShort: 'Ndz,Pon,Wt,Śr,Czw,Pt,Sob'
                };
            default:
                return {
                    hourShort: 'hr',
                    minuteShort: 'min',
                    monthsShort: 'Jan,Feb,Mar,Apr,May,Jun,Jul,Aug,Sep,Oct,Nov,Dec',
                    secondShort: 'sec',
                    weekdaysMin: 'Su,Mo,Tu,We,Th,Fr,Sa',
                    weekdaysShort: 'Sun,Mon,Tue,Wed,Thu,Fri,Sat'
                };
        }
    };
    LocaleService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    LocaleService.ctorParameters = function () { return []; };
    return LocaleService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var TranslationResolver = /** @class */ (function () {
    function TranslationResolver(translationService) {
        this.translationService = translationService;
    }
    /**
     * @param {?} route
     * @param {?} state
     * @return {?}
     */
    TranslationResolver.prototype.resolve = /**
     * @param {?} route
     * @param {?} state
     * @return {?}
     */
    function (route, state) {
        return this.translationService.languageLoaded;
    };
    TranslationResolver.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    TranslationResolver.ctorParameters = function () { return [
        { type: TranslationService }
    ]; };
    return TranslationResolver;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { AngularLocalizationModule, DateTimeActions, FormatterService, TranslationService, DirectionService, LocaleService, DateFormatterService, CmxCurrencyCustomPipe, CmxCurrencyPipe, CmxDatePipe, CmxDateMomentPipe, CmxTimeMomentPipe, CmxDateTimeMomentPipe, CmxNumberPipe, TranslationResolver };
