// @cemex-core/angular-services-v7/esm2015/lib/push-notifications/push-notifications.constants.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
export class PushNotificationsConstants {
    constructor() {
        // tslint:disable-next-line:no-string-literal
        this.appCode = window['APP_CODE']; // 'Foreman_App'
        // tslint:disable-next-line:no-string-literal
        this.apiClientId = window['CLIENT_ID']; // 'f624df7f-9b35-41c0-8a41-4854b86c84f1';
        // tslint:disable-next-line:no-string-literal
        this.apiHost = window['API_HOST']; // 'https://uscldcnxapmd01.azure-api.net/';
        // tslint:disable-next-line:no-string-literal
        this.appLanguage = window['LANGUAGE']; // 'es-MX';
        this.authToken = 'Bearer ' + sessionStorage.getItem('auth_token');
        this.jwt = sessionStorage.getItem('jwt');
        this.urlServiceWorker = 'service-worker/push_sw.js';
        this.deviceApi = this.apiHost + 'v5/cm/notifications/devices';
        this.sourcePush = 'postMessageSW';
        this.vapidPublicKey = '';
    }
    /**
     * @return {?}
     */
    getAppCode() {
        return this.appCode;
    }
    /**
     * @return {?}
     */
    getClientId() {
        return this.apiClientId;
    }
    /**
     * @return {?}
     */
    getApiBasePath() {
        return this.apiBasePath;
    }
    /**
     * @return {?}
     */
    getAppLanguage() {
        return this.appLanguage;
    }
    /**
     * @return {?}
     */
    getAuthToken() {
        return this.authToken;
    }
    /**
     * @return {?}
     */
    getJwt() {
        return this.jwt;
    }
    /**
     * @return {?}
     */
    getUrlSW() {
        return this.urlServiceWorker;
    }
    /**
     * @return {?}
     */
    getDevicesPath() {
        return this.deviceApi;
    }
    /**
     * @return {?}
     */
    getSourcePush() {
        return this.sourcePush;
    }
    /**
     * @return {?}
     */
    getVapidPublicKey() {
        return this.vapidPublicKey;
    }
    /**
     * @param {?} pVapidPublicKey
     * @return {?}
     */
    setVapidPublicKey(pVapidPublicKey) {
        this.vapidPublicKey = pVapidPublicKey;
    }
    /**
     * @param {?} urlSW
     * @return {?}
     */
    setUrlSW(urlSW) {
        this.urlServiceWorker = urlSW;
    }
}
PushNotificationsConstants.decorators = [
    { type: Injectable }
];
/** @nocollapse */
PushNotificationsConstants.ctorParameters = () => [];
if (false) {
    /**
     * @type {?}
     * @private
     */
    PushNotificationsConstants.prototype.appCode;
    /**
     * @type {?}
     * @private
     */
    PushNotificationsConstants.prototype.apiClientId;
    /**
     * @type {?}
     * @private
     */
    PushNotificationsConstants.prototype.appLanguage;
    /**
     * @type {?}
     * @private
     */
    PushNotificationsConstants.prototype.urlServiceWorker;
    /**
     * @type {?}
     * @private
     */
    PushNotificationsConstants.prototype.deviceApi;
    /**
     * @type {?}
     * @private
     */
    PushNotificationsConstants.prototype.apiBasePath;
    /**
     * @type {?}
     * @private
     */
    PushNotificationsConstants.prototype.apiHost;
    /**
     * @type {?}
     * @private
     */
    PushNotificationsConstants.prototype.apiOrg;
    /**
     * @type {?}
     * @private
     */
    PushNotificationsConstants.prototype.apiEnv;
    /**
     * @type {?}
     * @private
     */
    PushNotificationsConstants.prototype.authToken;
    /**
     * @type {?}
     * @private
     */
    PushNotificationsConstants.prototype.jwt;
    /**
     * @type {?}
     * @private
     */
    PushNotificationsConstants.prototype.sourcePush;
    /**
     * @type {?}
     * @private
     */
    PushNotificationsConstants.prototype.vapidPublicKey;
}
